---
title: "Population Genomics Lab Notebook "
output: 
  html_document:
    code_folding: show
    theme:
      bg: "black"
      fg: "azure"
      primary: "greenyellow"
      secondary: "darkolivegreen"
      base_font:
        google: Prompt
      heading_font:
        google: Proza Libre
      
---

```{r setup, include=FALSE}
if (requireNamespace("thematic")) 
  thematic::thematic_rmd(font = "auto")
```

## Author: Maryam Nouri-Aiin
#### Affiliation: PSS
#### E-mail contact: [mnouriai\@uvm.edu](mailto:mnouriai@uvm.edu){.email}
##### Start Date: 09-11-2023
##### End Date: TBD

## Project Descriptions
Notebook to document the bioinformatic of [Population Genomics](https://pespenilab.github.io/Ecological-Genomics/Tutorials.html)

## Entries {.tabset .tabset-pills}


### Entry 1: 2023-09-11. 
  
- Get background on the study system (Red spruce, Picea rubens), and the experimental design of the exome capture data
- Understand the general work flow or “pipeline” for processing and analyzing the exome capture sequence data
- Visualize and interpret Illumina data quality (what is a fastq file; what are Phred (Q) scores?).
- Learn how to make/write a bash script, and how to use bash commands to process files in batches
- Trim the reads based on base quality scores in preparation for mapping to the reference genome
- Red spruce study system and exome capture file
- Structure of fastq file (DNA sequences, plus the Qscores)
- Using program FastQC, analyzed the sequencing run for one file
 

### Entry 2: 2023-09-13

- FastQC results discussion, good qulaity sequence data for most reads
- Initial 5 bp or so had more variable base frequncies, The very end of reads had slightly lower @-score
- Based on this we set up an analysis to trim the reads, using 'fastq' program
- bash script run  'fast.sh'
- looked at html files produced by 'fastq' and compared pre- and post-trimming -- things looked good!
- We ended the day setting up our read mapping of the trimmed and cleaned read using 'tmux'


### Entry 3: 2023-09-18

- Review visualizing QC on our sequence reads
- Trim reads using fastp and assess pre- and post trimming
- Start mapping (a.k.a. aligning) each set of cleaned reads to a reference genome
- Visualize sequence alignment files
- Calculate mapping statistics to assess quality of the result

### Entry 4: 2023-09-20

- Finish calculating mapping statistics to assess quality of the result
- Introduce use of genotype-likelihoods for analyzing diversity in low coverage sequences
- Use the ‘ANGSD’ program to calculate nucleotide diversity (thetas) and neutrality stats

### Entry 5: 2023-09-25

- Calculate diversity stats for our focal pops (SFS, theta-W, theta-Pi, Tajima’s D)
- Summarize the results in R and share to google doc
- Introduce Fst in ANGSD using genotype probabilities

### Entry 6: 2023-09-27

- Review the diversity stats for our focal pops on the google doc
- Estimate genetic differentiation (Fst) in ANGSD between our focal red spruce pops and black spruce
- Visualize population structure using PCA and Admixture

### Entry 7: 2023-10-02

- Review the population structure results
- Perform a scan for selection and identify contigs with outlier loci
- Identify and visualize outlier loci


### Entry 8: 2023-10-04

- Revisit pcANGSD selection scan to output genetic PCs and outlier loci list
- Extract climate data for red spruce localities and summarize with PCA
- Run genotype-environment association (GEA) analysis running

### Entry 9: 2023-10-09

- Articulate the purpose of and experimental considerations for transcriptomics studies.
- Understand the general work flow or “pipeline” for processing and analyzing RNAseq data.
- Review Acartia hudsonica ecology and biogeography and the experimental evolution/transcriptomics experimental desig

### Entry 9: 2023-10-11

- Review Acartia hudsonica ecology and biogeography and the experimental evolution/transcriptomics experimental design.
- Develop questions that can be addressed and hypotheses that can be tested with the A. hudsonica experiment.
- Understand the general work flow or “pipeline” for processing and analyzing RNAseq data.
- Visualize and interpret the quality of our Illumina data.
- Assess our previously assembled de novo transcriptome assembly using Trinity.
- Start mapping reads and quantifying abundance simultaneously using Salmon.

### Entry 9: 2023-10-16

- Review the quality of our Illumina data from our google sheet.
- Assess our previously assembled de novo transcriptome assembly using scripts provided through Trinity.
- Map reads and quantify abundance simultaneously using Salmon.
- Prepare data for import into DESeq2.
- Start analyzing the gene expression data using DESeq2

### Entry 9: 2023-10-18

- Assemble the quant.sf files generated by salmon into a matrix to import into DESeq2.
- Start analyzing the gene expression data using DESeq2

### Entry 9: 2023-10-23

- Review results from filtering of the de novo assembly and transfer the new gene expression matrix.
- Explore patterns in the gene expression data using DESeq2: explore data, plot PCA, test for DGE, make a Venn diagram.

### Entry 9: 2023-10-25
- Transfer a data file that includes additional trait information.
- Work step by step through correlation (or association) analyses using WGCNA: install packages, import and filter the data, look for outliers, normalize, construct networks, explore module eigengenes, associate module eigengenes with traits, and visualize the data along the way!!!

### Entry 9: 2023-10-30

- Discuss “what’s missing” in hands-on coding sessions - to better set you up for the future (so you’re not surprised)
- Continue working through the correlation (or association) analyses using WGCNA: associate module eigengenes with traits and visualize the data!
- Perform Gene Ontology (GO) functional enrichment analyses using GOMWU with our DESeq2 results.

### Entry 9: 2023-11-01

- Review Homework Assignment #2.
- Perform Gene Ontology (GO) functional enrichment analyses using GOMWU with our DESeq2 results.

### Entry 9: 2023-11-06

- Structural varients