<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Maryam Nouri-Aiin" />


<title>Pop Gen Lectures</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>





<style type="text/css">@font-face{font-family:"Open Sans";font-style:normal;font-weight:400;src:local("Open Sans"),local("OpenSans"),url(data:font/woff;base64,d09GRgABAAAAAE8YABIAAAAAhWwAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABlAAAABYAAAAWABAA3UdQT1MAAAGsAAAADAAAAAwAFQAKR1NVQgAAAbgAAABZAAAAdN3O3ptPUy8yAAACFAAAAF8AAABgoT6eyWNtYXAAAAJ0AAAAmAAAAMyvDbOdY3Z0IAAAAwwAAABZAAAAog9NGKRmcGdtAAADaAAABJsAAAe0fmG2EWdhc3AAAAgEAAAAEAAAABAAFQAjZ2x5ZgAACBQAADWFAABReBn1yj5oZWFkAAA9nAAAADYAAAA293bipmhoZWEAAD3UAAAAHwAAACQNzAapaG10eAAAPfQAAAIIAAADbLTLWYhrZXJuAAA//AAAChcAAB6Qo+uk42xvY2EAAEoUAAABuQAAAbz3ewp/bWF4cAAAS9AAAAAgAAAAIAJ2AgpuYW1lAABL8AAAAKwAAAEyFNwvSnBvc3QAAEycAAABhgAAAiiYDmoRcHJlcAAATiQAAADyAAABCUO3lqQAAQAAAAwAAAAAAAAAAgABAAAA3AABAAAAAQAAAAoACgAKAAB4AR3HNcJBAQDA8d+rLzDatEXOrqDd4S2ayUX1beTyDwEyyrqCbXrY+xPD8ylAsF0tUn/4nlj89Z9A7+tETl5RXdNNZGDm+vXYXWjgLDRzEhoLBAYv0/0NHAAAAHgBY2Bm2cY4gYGVgYN1FqsxAwOjPIRmvsiQxviRg4mJm42NmZWFiYnlAQPTewcGhWgGBgYNBiAwdAx2ZgAK/P/LJv9PhKGFo5cpQoGBcT5IjsWDdRuQUmBgBgD40BA5AHgBY2BgYGRgBmIGBh4GFoYDQFqHQYGBBcjzYPBkqGM4zXCe4T+jIWMw0zGmW0x3FEQUpBTkFJQU1BSsFFwUShTWKAn9/w/UpQBU7cWwgOEMwwWg6iCoamEFCQUZsGpLhOr/jxn6/z/6f5CB9//e/z3/c/7++vv877MHGx6sfbDmwcoHyx5MedD9IOGByr39QHeRAABARzfieAFjE2EQZ/Bj3QYkS1m3sZ5lQAEsHgwiDBMZGP6/AfEQ5D8REAnUJfxnyv+3/1r/v/q3Eigi8W8PA1mAA0J1MzQy3GWYwdDP0Mcwk6GDoZGRn6ELAE09H/8AAAB4AXVUR3fbxhPfhRqr/6Cr3h8pi4wpN9K9V4QEYCrq7b2F0gC1R+XkS3rjKWXlfJeBfaF88jH1M6TfoqNzdWaXxZ0NM7/ftJ2ZpXfzzeVILi0uzM/NzkxPTU68Md64GQZ+vfa6d+P6tatXLl+6eOH8uVMnTxyvVg4fGisfhNfcV0f3luz/7Srmc9nMyPDQ4IDFWUUgjwMcKItSmEAASaNaEcFo069WAghjFIlAegyOQaNhIEhQxALHEqIeg2P0yHLjKUuvY+n1LbktrrKrOgUI/MUH0ebLc5Lk73yIBO4YeUrL5GGUIimuSx6mKl2tCDD8oKmCmGrkaT5Xh/p6rlphaS5PYp4kPAy3Un74OjeCdTi4nFosU6Qg+qRBsoazczLwHdeNqpVx3AW+oVjdhMThOo6YkGJTl862RFq5r263bbYSHyuswVrylsSBhHzVQKDU11g6hkfAxyOf/DVKJ1/HCvgBHtNRJ+b7eSYepeQ4VLZBqAeMjgM7/zyJJF1kuGw/YFpEq458Xrr65YTUa6VCEKGKVdJ+2FoBYYNKCwV1K6B2s1mJnPB7Ww6GtyO04ya/HHWPHs5P4J65NyVa5VA0E0LocwPci45b6tvMvohm1BYc1h12Xd2GrbbHVkjB1pzs6IKtOHeYd+JYhFasmfs9Zt+SZlo9pu8eg0utWZAKB8vjaxBQx7cSbK3Qdr2nBwM27vrXcUHtLolLJyJjK3CAbDcFDo3hsPZ63IH2RrsoWyskdB47jiKitFtcAgqj4wQQxN3PB81RCiCo0Y1jnUVYlOj5JHhJd2JBevIEeSQxDWzTN8PEE3AL90KtP11dVrC5II1L1w331pHFq10vPBGYeyUCFRvB7PAEzMltdubhb+lZ4dw9w86yyNfG++u0ZWOBkmsb+GrsrKGIN4R0XPQimnAEcj3CI6ZDR35zzHJEZlcW5cQCTMwty4umkB5B4ajHwVNhQDqdMLSAmClnhLScgYgMbQJESALUrtIvjpQz9LVxuIPSiYgQkjusZ01l4BERrPtdO9KfDErKQLne6EUbJlXHqTccNzL163tuES26ickjo5va6FIkCyIyaFEYA+lejuqlFxLWIYKmQG9W0tlMe0yXu80wPe/OavEJrd8srSFziSal30wMj5H2mH7T6H218RQ93qOFysDEgtLBoRuQUeXjyPQKexdLjoa4vtAQJiBsEXYutEo9T1/m5mUdBMbXFCzIq8Z6Yl5+7nyic+1mE3xisVatpBarpcC/mUs9/s3Csty2GRPfLMo7FrfqcS1KDxIntwVjnkEtjRJoFKEVHWmelIyxd7Y9xlqGHTSA0VfbnBks08M4W21bHczuJBrTiYixiBnsMF7PepCwTAdrGcy8UqZb5uWGvIyX9QpW0XJSrqE7hNzjjGU5u1vgRe6k5DVv4DZvpVnP6Vi0yMKLOhUvPUq9tCzvFhi5mV9KVNMvWpfRJg1bggjEml6Uz6KmiiN92dh+Gg19OHK4TmOC61TIcAFzsF7DPNQ0fkPjNzr4sMZHaEX5fk7uLZr9LHK9AW9KF2wU///BUfaOnlREfyrK/rv6Hyn3ISkAAAEAAwAIAAoADQAH//8AD3gBhXwHfFRV1vg5974yvZdMQspkSIYkQkgmhdAyIIQQWsSADCLSpajUiMgiAkuJNGmhKyJGDCyybCiyiGBHRGQtyLIuf2UX19UPy7oWyFz+972ZBxOE72N+L2+Yd+be0+5p99wBAscBBIN4ACjI4D4oUJEIVAbIL8wPYX4oP1TQ3um3+0v5dZz2bj44nsyKLhYPXKkaL1wCAhuuXcQ69dsWyAu7qF5PBMFqQzQRkzQgYvIQCuXleXYHlCXl2x1YZg+F7HxMDNAQLQoVetwuKZCZjRUTQqc/f7RjebisqAeuEQJXmpZUdA/3KgcgsJA2kL1xDNPDZqCyQAWdXiIy5YOHThUq4/KB1XFpgPr5heVtJuSQvJzxOeKB6HfEplzKWCEA4Sc+Vgqkw8bwIF16K7fg0ttNJr3DajEKBqfT5UlNkwXJKyD4hCRRlFySwU+TvTTJkJTh1wkms6l/pBWa08Fmt/WP+Nz2AWYcYEez3WwXvU5qECE/VB5ylJXl5993Hyc3zw6hkHaPoerldxVjh7eMX/F3hYWxu0KF382pcKpXsV+9QlS93Mj/Sz/ujinsVE1dDTszcEk1u4LpPdjXmDdw6UAsqFlUg7rmf2J+d3aGLmC757GBuEe55mHNXGxifZVrLtuNNUBhwbU6wSQ5IAOyoS2MCxcH7VmpXkHIdZlFP4BPtOvFdvlZZsncL0Kl1pZcS99Iam5eK1erfhFvrkviL9HDKc5X6OV/ChUq7aGEvw5U6QuFVCbEhOSSZHegODM7WOzxhOzZ2cVFJaXFIbfHK2cH7WlELuK3EnR5vHZJEkzvHZw35S933n0ucur5ky/MO7SraN2mrVuqGiNPnIt+NnTy6HF4fMkfvf+6EEjfkpWPh7rtXrJgp+NAk9hzQScj6194/+yxlZE72Ow0KvcdloMLbPcBiDD+2jdSW/Ek6MENfk55AfQMtwabaPC0aZWZ2a6Nob1NKgxRc3qemb/aF0jtk3xZPtkpc4Xjr3KVXE7WDfpi+sfVJ1RotwUyJVFVbE4ZV3JUPi0pLsq++XMM4A9Vd+/YcXcVvrtx7bLN61av2oINVTU11dU1NVV4cuPaFRvXrV7xDGPNH6+heQJpbMQaHLiz8R9fXb5w8dLl5vO7XnzhD7uef37Xxa8u//3ipa9pxpUqrt5AYeq1b8QPxVNg5BQWw13h9k4PpEqB3Lx2eW0DlmxfqkdfUhoy9Y6EnNZgW0t7MZ/6smlubka+I0NfFckQoDwPkjih+d4yrpTleTdRqoinJE6Ts7AULcTt8mRxQbYjMeLcXMpYwucgMgaCkrrMn668Z97YBwZHJm/+/hnWZ/KwOzazl5c2DerS+o2Xth9eshXXd7jTu7NHHeb98+VHfqw/+z/Cmp5zhvSZe3e/kSOubt2EO3tExnWrrbsy/51x94+aWFa/84V1k/bfx2Z1fWE0+2It+2zfxGEfAaBiMbBctRiug0CpIBLFUpyK2R+OumYgYrZB+cZAdoT4+TfM0CpsksEggGCxGoNUsV4J5sVpc5SGJE6pwxvIJgM3r97+1Kq1S7et2UQKUI/v7znOCn/8jpW80ohvKaN24aOatFEFAx8XLFYDFYItR0UbkQMljuIiEgx5HMS0efW2pWtXPbVdGZb9yjruPIInv/sR3z/+EisAhMFkrmCRXGCB9uEUKgoomw16o95qEwxoJiaT2cDtl84CUP5G4XWJOTBmWLK8olOmNOjMKhUpWZWHK5LZgl9279229we2OBUX50kuVjv5QDo7PBwnsvrhWJF+YDIuVagZDxeFHOF1MEKbsBMEQS+KJjOVdXJ1BKw61EH+feqSTzTz3I7ZA3Zuv+whshy3sDFL2TjctJR6n2SDsfFJ3A0I5ewXfAgugw7s+0XQG0SAfFVWHOEsr6TyphSHW5NHFc9J6Wa+7B3Dfp42HguHAUINniPlZCpQ/l0CogDIrW/8u85iv7sGv8ZzGzYAxjwV/MCxTwobJQCTWU8HRPQeruaaXpRqestVdUOXso7dupeF7px4Z8+ed3arKFc44AIg51W9ch4kIIiUEocmSk4sBpCcj15oUDRJXYYExl37RmirrkIv55rLASYJJF+S3t0nopeptU+E+mLrLK+lPgQyid3mCBU6UP1rVz8R2n770zc/Xf7x8s/Nn9fvaFi3rmFHPfmMLWRP4lycho/jNPY4W82Os88wiJ34K4tdAIQjAOQkx8YArcM2PaAOjSZBL8uolzAJFFvGDXd8ej67P2AvKpUkOYghcnK7zl300RBcsExwzJ/hbrd7GuYBwhgAIYtbTx/3+d4klJ3gtKCQnGIz9InYZEzqG8EkjSzNavCB/cXYlcQshhyMsZrI6PYLWc3lOG/vlA4rHr/3uTFD3r38/r+3fMKOke9W4oJ9G566u7au84CpOz/ct5R99wF7W6dIYjjnawrHIAh3hlungFOWgXoyzVKbHOr1eD19Il6vISsrrU8kSzbY+0QMGpdjgYh60zDTHJKHoyP4404pw27zB4o1o62gq+BLL299am8j+zv774zj995/dgTOZsOfWr3rnTWPj2h8qGbo1/M//kYYvmxfms7TtPrM54E7ns4vwBw0rFy/aNJjRRVTet31OgCBPABhongUDOCAzuE0h6gnxChToCJ1ulB0iH0jeqvscFBZotflk+hMQ5oJDqhrC/l//FxmAUlGYeK5Z6Jl5MDec2yJQdc+l5ViNduL1avoZ805eGll04jy6COKheT8S+U6kQwdw+lW6nPpXF4qtEoBziwAye3mMnRLkqlPRLqZdQlsKxTcLghkqhzjrLL5M+WgUwldSkjbL1HPLrCf51d8MHbv66zu/mcGl5Kz0YNZ0+mcf759kbEB29qGGrZiYWop2b2R9fYqnKnlWOVzqXqgNfQIB5LtRr8fQLLT7CyT0ZLaL2K0WFzU5e0TcfmojkckcgvcyhJ4pNlr8Bd63VyEhIbiGhfIBFGTq8R9lqcWB2Dl1G79Rn/9i8n08OU3L/760UX2E369YuvqVUPrI9VryFR8CXc5V/rYefbW7svv/YNdxUHv/OnFVQ1V8yse2Dde0UcAIY/zU4L0sA1FEQg3jJT0jVAJFBlqbOOrALk1dCOmkuHNF+mpaKOYunHhldNAlZhEyFGpz4R20C+c47Vmu+6gqXo9lewuq5TfXrLnZORk9Ink5JjAlNwvYvJBoF8E5N8qd9nN3jrmj7mOx8OPLDXqolpgwv0zZkpuzaeTynf+vWjNvnr22b+bsfDJR7+e+cL6dQ1bXlu3CDvOWfHIMytnrhJPHt7x4L7eg/48+8C5U0euLuu/f8ozr1xteHTRssdGru8V3kwfeHTMsN937/zksLEzFdlO5NQpNsMLWdAtnJlizzQYAAQu26AljUvWZbEQlyuJi1Ymcr8Iaal2jjKNg5qJ9Ctqx02jMyDFKHJw8TpUIvjHKhXZQlZ0/Iwe1eO++6/RVHpg2mv/uPbBuguPMtfKLU+tuXfjkIFraEVzg2tlMuZg6O57/vXBP1C3kZ3H9od2PPV81RMVE/aNAy3HEcaokRS34Ta+LAA8XotzQMRiizkRDVfN87X0JXae6NzkVR6Znehb6J8XL+Y3IKovXMjn0oEDMrkmmc2iXu9yGm0DIkab6hgTZklwj/T6FDccpXsmn6Rjlxv+knyrTFMR8+U/cF9+DiRwh/UCiChwdeXD58cDhSwsRjeikNNcTo83/0AtP2DDKLywji1nhxSezMTjgo9eVHOy3LBbJgIQ0OsEsToiIFRHrIjI4wHOlfxEz6a4ZOTXTLq9eTjdTofW1bEH6up+g5GIBDhGEr2BkRNVlMZTa/P3HKVyrMMKrF3H/KPYUAWjlGsXaRnXrxTIhrJwqp/bMtnphFYWIdgGoLWtddqASGuPzdA7YhNaqFZLvVJSEa48LZwUd4YSN4mJ+aq/ctSSXgtmD6gf2emV91/9KNj38bHd9l3PX0tq19dMnzFw3OSsgsWjj+zqPXn0w4On3e9nZ+NJLYFZ1yqkQ2ITFEM5zzwyA+1KLJ1kVwpAjsvSTgx3S+rQQeiisxv5Ky+9kGbnqUmllmSFEhOP6/G4ug6C2nJQUPdSt0td36R1IFMgbsUalrqlQAbw4KK1v1BwIH/udKqm8NCQbeMHP2LUtVk3rv7Fb4712N3Tt/DeaWvZt3+8wA7swe6Y/5cvjv3I1rHJn+AyhLM44ODVn14/7bBUDpq/hpxb8c388XfdM+rU3veu+Tws17Pv7O79aFvzMnvxc3aaHRq8sAZX4jgUsP7CfvYntoNhGYquJiAAAKJNPAIyWLjk0ojFqENR0SwqyILNaiG9I0bRYhFECoKD518xh6iplZYz+5W8H0OIlBsz/tURB6IHmnaT7itJORvb6A94cnbjGZYvHrnSg0zENwfPGTGddQIKJwCEo9xyW8ALGdA7nO0UUg1Wn89iEGQLjwd01iRrUlXEarWAxVcVsTjAWxUBevt4QnM9/gxBMbluwe4SAjxpj/mcgN0ef3cCt2IAhVVLsR/7+TIjjZjU9PTeY1ew4I9/Ovhn8cCeI/Nf9BnK2Pk3/kZ7TF00+6HoquhndauXPAGAMIdb09Oqr8gOu6jFpbdQb5IDekccglHi/HK2DL+4emRymUNIE3+Ro3WokKfbtNP37Cs0/7rxjQ0X2Cvs2Rex/NNLuysbxBB7lX3FPmdvl64rwyU44QusOVSzuj8AUTgmDuEc04FdsYcWQQ8COJyiuSoiUsFSFREct4ppwc9rSBlA+ZuAPZTBx2Az2Uo2CY/hIHysic/1z59PI/dU5CtWz+aJB9gi9gKmYebVKZgHgMq89Bc+r1GJWSSDAQXQoWAyS/reEUlCQsTeEUKRr3B03DZmUZBwxy/6S/MZmh+dTYZHt5OF4oH1LKc+eilhJj0UhpMlAKQ6pAbjTRPxSW45Q0CbAac3asPzwaNfrY9LTuyi2ilOhUvnI8SSohNapUJK7wiAaDLZe0dMgujtHRGdt4+8/HaphRyV9+rq5lT1xe9nfPc0a2IrDuKQL//9bve3DrL/so/Qj0kbVrGXCYuWZWXjUhzzD7xn/+D6GvYau8Q+Ze8H8LUY7WK6yuVQ2KdHBJ0giCCaTTraO6LTiQaJoshJV81RgnG/Qbydi5f/DYnpjc2ssZGSRrI3Ws1z7dXkYQC8NoLNxfFqVpwaNht1OotVT4GzFDJj9GrpGI15+JJiPpxLMg0v6dVv9AONx9jclFWuR6fyFGvI0TNxvRC+UjHmnkjBViRGg4Ix0Yn6RGzLWkgJZRVRDKHw1TvRrzc2NpL1J6JN5M0l0dc5snnk4+jCBF0QIT1soQCCJCMFzgtw3EBXxTekkO0+0aio0pV/bIp9V+KIgpPrUZJOFCUev/JSmsuNBjuVjDK1gKQgp2DnLbuZlRjwuJUAn2MY4nce4COtZjadZSsCntbhh6zRomMm0bbpo+bh4oGrVQLPOume7Uev/BCXo1IDsUG7sFsvcaytVpDB7jBS2aqjKCdypaUI4xPzabNJKZdj+WvNn+tsW4/RVB2xkGeEk582NR/nE3ZMwaxy2guAqFp99FZ5bu+IXqDW3hHqvLVNiOltBiTmueJRtpW9oZgjHIE9sBOOujo9+v1/fvn5h/9Eeb77LHuYa+94HIt1bArbxs6yU1iIuRjEAnYqZp+E8erqdUBRONnA+c75DE6XQaiKGAySLDuqIjKVEtavhpXmSgW/mlplYChutYXx7Ay7tLsRZ5PWUePGL949euKoYPr7t1HOh2jK6mdXrVC5wHaoXLBCCp+Zp8MeAIEa+OqmZtns6x0xC7KTL2yZM+MtlRs3J6I2pViG8q258sX7OOxndrH0tpz5ki3rzuqxivyf/DnN+WMCN1SGs8yIxKS3y0aDQdYTwePVm8EMVRGzmVDK5UepkSi6cntnp2Ku8ktw20SOf5bGNm4BcRXyGdhfcfkJ9jQ7/VXTzl2vfEZGRLeJB94/zf4+LjqZjFi9cuWqJwDVHIFw29ha4V6a0wSQ5BSFrGxTGvV4uH30CFSfoEoJiY4mt0CGlozy8D+o5jgx+6jmBbwy4BEI+9d3rHnZ0I/GN+7usnL1ey+xM389WLx/1+INHRbWXfoDLjz+6Z07su+YN73vyIFFvd959sV3qtf2nfFA35F3FQw8AoDgABCGcv7JvJ7iABSRUp1epgK3CYLmFeJ5qGYSi7k3IEsbWYFQyQrE9PWqJzjM14yPj2OHrLDdhgYZZafDrqOCmQ8UpzGUuFzsLkUnVHMYs4uij/2F/cJfFxrfee3ld8QDzf2vsC8wo5nuaa44+Mabh+ghQAAA4XW1/pMcNqJgMuooCJQqiPLlrxWvQhjgF8//SgXTwej3O6M/NmF1x8zWHdVaFh/5uU3bnwXkmg1yXz6aT6km+QwpyW6LRdQn2Q0U9TGTotqUGOKqNclWAjJldKcyenwSZ0h8cyc75y5CT3v2xU42u+nL9p6UYpSa0Nne7yy+1EQ/7PaW6/dbm0N88llHNx18ic5qnrv59RXv0YUK93QAQr1q9QNhhyCJ3ORLiskXFJMvtDT5KhocAz63Yu7rj/PIY0oTXmKdjuAkfHg/60QWROeQZnI4+gq5M9oX4lybrUY5GWGrIBJRpnoDiChTUeOcJmE+qKL+GCJdcNEhlrSb+Q6T8+R887zoCZJPFyv1ZQBBscZ6pWKmQyqDLKBgMIoCNwcUdUrMcuuKmVot8AvlzU6qi9roq82/0LSFwoaNC69OAIQGdoRMVnSRY2mRUFAYoxcJlTDIOdBSfeJRD5nMSvEEu4B+dkS6svyKX6HWC0A+i1c2Kd5c2XRy3h0mgYbo/4spg/KNEDuCzdrMFFACSacHOUgFevPMXj5rMb9CfMoLfOrSA+KF5b9KyigFJCgExOMgQVJYD1TWiQQEwrO+G5rpVFUTC3DfaPxsA1vG9pEg3dQ8jnwV9QJea2Zv0k3XKtUKsJLHIlEqwBgjmU/LQUfRp9mbCwCxTjhHHZIf9OA8AILRID2BkJ+s1ZoxwDW1OMStBHU83G1fm5MZ0+4QzhUdK3f33F8MRKk50lPCUEXzoVc4K1NnTEvz+Rw6yqMpYkzrFSFGI7jd1ooIt4LJFRHRA24o/98LVH4tX7NllapJZ7zS6LZn8QVeLKsVKjrQrxv43GPPvUychyc/VveH0F3HR77xCrNs/mPDWy89tOWB3js3Y1+b1GPe7Jq5dxTuORZ11TZuHC3LD00fOhwI7OVWtVZygRPSeVUt0+D1Wq2mVGqiGX4zmNwOu8HOhccRljzgqoiArYV5DSXF1SDB1sddEk825YBijeRQiVcrvHAqyJ5Pv/3+k0l/7GwKzGzQ6Wa811i/qXFjfb0wlJ1jP/DXxwMGLpdcbNHcsTuWvv7ll29fOPPJXwAQpnMOLxWGxbIaK6VuPU3ySmaOmQ0cHDPPzVmNGM9qlJ1DHgNzu6hmOGTcZXYV9f8d8HTbUOn8QrbvuW11Tz3swiw0oRPvyPQu96Sywe9+2mlNGRBlVqGU88fB+dM97E+VvGCx2CV7ht/htgIgmqhez9mjt1FnRYR6bscerSYTkLTqvTcUDPLPA6osi+JOiG7ST//n2W+/++TCTLMsNCxmTzdu3Ny4evOmNS9gNlr5647tA/rh0V+/mfny+4Gv3r54+i+fxLF0cN44IRk6hdOTDF4jpdzqtkrxGit4uRskyaUyyqIw6paZQyiRZQ632++JsUuivNbh53Kb+x/2JYp/e/+7qFl8eecf/zBk65bfb7WQLstc2AZl1GMH9v3fJxx/p2pttp/+c/eGrS8oUksFoBYpHVxK3cVlMjkJ4UaSuj0GvhQMgKIsVkScspUqq0GtY98IAxWmOZS1p2QNgeJSXkPW3DX3mE+zrxreeANH3lObN6LH8KHopW83l9G3+3TugmsDC9PnPNkLgEKQuYQCzplcKIVu8HC4a56vQ5YpvYtY4ESnSHIzW6Vn+Qzd72xlLbYWV0R0nXpFDJm6XKvOqvPk5pJekVxrm/JekTY2T7teEU9KnHUa+zj/8pXd+rzbxD1uragaVBdAqDC+jaAUkrJv/OXKcGMXmJOnbhQXF/F3QsHJVnf87VhB3sSqoa/te5X9jf3r7FdPzMgtC/ccNOnTtwb3ZPb6ZWdOPLzh7amPD50/4z8/1T4uVE5ICkzt9ewxXYdBbfPqVx54ddvqMauTndXFnYfmBnY+2PS66ypEhs2ZFOn5IO08/ZFvfn4cEPYCCD24nnuUzM5i0nFz7dF7vEkWvcMhVEQcNgOA3q0Y7xjlCatesVT2mALbtRUfM1P06cfm/+GZhgadoWD/jBMnyJuLfn/kk+jrfHXnDOow4N5XP4gWAxDYDoDjxAtAwcr9tZ3PJCDa7Ga5MmImVlQ04/3EwqZSIqAJJVQc3NDQ1CG3TceObXI7CJWYU1Zc0qFDaSkAubaKudSxTZAEd4Q9TqPRrNP5kj22yognrLcC1z6ISzW5xSTOhATTljhb3v2det7Zv/eNGZnLt9g16B6h+aqNHZHv0yaP8TSV89QGJTzetxgMRqNOEkSdYHeYAGw2nY7KRje1xiKGfD5zeUyFyuJsRTUiQi0bdclYkzcER73JeuD5E2zOnB07dKSgy2icydpGlxLpQTZOcjW/XTo9NjcO5nNT4GQCoiASQHfca2tMVBjHYVRo6SRfJQGoCAfcdruDiz+gdwRo66xWHrfb4RPMPm5p0302p1UPDkUPuCLEt534Igi1bHVIVIgEzfAqepHh1bRDypryyOa1DVNmblnVsDhFl79rIuIAXcHhmYdfJicWLNj3cnSLcv/zx9HjQmV99dDDg8e8+heuMZq2cnxdUBBOApeiri69x23S22xcWW02g/V2ytpSV72Jmrp7m4JG6NDUt95RNPXwJ+q8d0XUSWM2dhSfU9EknsU6wSyDnOwzeLgds1GbYvxvmcVylSHFilGFxE4PYRT74fKaf/wOTZcvobX5lZ3PPffii88/10Cy2I/swyeR/AFNmMfeZ1f/8rfzH545p1j5vdyW1apU+6E8nOEzCrKsS3foHJkBwQhWq7siYrXprboUaHXDzMdZ0GLBqpaeO2hPAhMUr62Y+gRHrThpU8Niry7c+PBf/+f7yzvryabGFc8+6xowcMRg1kUqqh9azT5h/1GcNr14+GTWl29fevfUeYVXHNNSlVexqMKW6qHJyT6bL8OfnOK1pqalecxOp8wtv80MFRHz/+Y2VT5yJ1l63Ul6r3vQ0njtQyL9GzaIW15cvXnjnI8uf/fJ57P0SQsajObpM/d9mHXp3YunT59birloRDO2a6z/9T38eEzFCzE9okGOpw1ywy6zXm8wEF4DsZrB4FYtg03rc2nRkaE5IY15ZEfvjt4eRQtfaahz6rrsFoaZNlk/fTbaJFSenDQjlrnS6XyW1twOtIplrqLzeuZaEfHYJKq/rj/5t8pdueG5kbsG25Hfpq50+j/e/+tjA/bXzF82+dmN88r/evSPL3Z6ftEjj7Yds+J13jSzsaHnpjbt7h4Uvrdr2aAH+yzaXLm4R1W3O7p2KO71FCCkX/uG7BQrwKPWJlwu3jPioEKS1+C0OXtFLGGbVeaCkj1xU3kqIVjV5ONWqo52xVGXhtxKNuHyEMcdA5NSJuSy17ZurRiBXdlrw2vN8lyzHQeQZdU9/83mRWePngiAsIOvrjKhElx8fh86ZZPJ4DS4PSaz2aZzWdVV7TFqEbMS/4daVmW0rJcrhBY127EvX9TPNNQl6UP7Z7zztlAZLeMO6GMSvnpozV2Dj54hp7RcjgiVau+HAQ0ms6hHK6jhiJZl+NX0NFTicIYQt7ER+76ptuiMte/tYyP4oI/8o0cx9iPtrx6K5UpSgI/Winsblz4lNc3rsZipYBZ0yQ7ubnTuxCyYK7c2A1U2Z2Rlk8LhUHSq1BmbsoRPKeSfcBbp2qSdPsY+3jNxsk5nLHCcaHqjg0snBF7dzc6QBZ3OvHR/dK5QyUaz6j5l+4tJbXTp7trW9eRvHClACAIIOpXGzLBdFiVAUWlxQZ3RLaD1pnQ4ngmjmhUfYgteQT9m/JktwFVH2Cn27hFSQLxsGO6IfhU9jUdYD0AgfL1LfHw3z/sVMqnHK5jB7OBLO0UHfIJCVam1GRJo46KKOdrSUrLvuwFOnfnuS/tYTsWfl/StKu2xq3cXzuCVn9wf+pn87mrGy5vtC03HtkAsZ6YPCZW3yJl7RUQr6npF0P2/5cz0oeZ/ksHR0+TL6D5y31Q6eN685sPxrixetlPl5/YlJxu9AFbZRbmnpqlpTq09K3F7TdV/bpXcPJZTfEtxCddDvj7d3EK4ZLfHjedrpx794PFH58/49MClCxdM44aRZaRxE+aPjywnw0Zg4ebdS6Xj7NzZoCl4FhAvMxuZrfluorSo0RSABN+tlHzx8nKeJv3cDAiV7Ijaw5Oq4OwWDQ4H8UFqqsXiE2laujso0QScEzYFFXSDxYr7U7DPVNCV5Dj2pcRw4eKhDx+Z/9jjp45OnvHwVFIePIvB49LSPRvZ+yPvJcsjvOq5cRenZNg4zJn2qEvdpyXVQg6tAS/XAzu1JvkcpuoIdVglCaojEuTngS3pjfw38rSkOlOZT8nQVNOmbD9lKoU5HFg8t2TMUz2mRrqPyi95omTcisrHK/sMJSfuLFn/UKvsVinhsvqH/RkZSeoOPFuKdcJwrcuYCALV8343AGpSu4xtNPOWXcZcCQNO1/Xt0PNKk/Gszp3Ly0IVZPfVC2Lfxb3C5ZVhQDjK7fd5dVemazjNozNTahCARxo62irVJxKnwUz4SzDKgg+07k9ljt9sw2apra1KOJCldLR6NAOuqD89OWHNwpPHcdniPisKChY+tHv7My8sX/FdifTO+xlov4LNXXfvoH7vstCH5z462QkQypUYSDzBpV4Zzk5y6s3mZI+dGD1OMS3dlORL6h/R+3xOcNr6RpxJIPa5uRWkRdPQzZ6Nm29lf5Lfinl2ypuduEqQxqONXTatnD0HG9jQblU05erVU2+99f/EEzUL+/1uGTs397MxS+7YtDz/xwtzsfO+U4psZqMkeIVtnHNByAibW0GmBSxtctLd7iwZeNSYn1gJchaVBku9il8r9co82Ja9clCxDnKwNLs0IXQ6VLV4+OLx8+eOq7t/UVXVgmF14+YuGrN42MKqeVtnzHh627QZW8mHj01aNmxh794Lhz059ZEFD/CHvfj7JZN+N2XbM1Onbd8BiscDEJT9Fw8MDrdzWGSj0WYS9URPTS6LW/YmGSwW2So5HBScbqsz3UmsTqvThG7JlATlWg+33RHrzL7lpjuGUOGj1uaovjBEKnH2HjYCJfY6dmGv72BvYGd+ARu7j1wgZ5vZ3Ma57Ec08RslQBKsgaxUVYkkUR726QUqUDlmFjgmiYqtbgjFLYRiI5p/YebmnxVpXPuF1kupUABdeGdcdiE4pdy0Dj5fmkmCgNS13E07lbRqK/n1/mCviN+tt/WK6OGGznh/s4t9I39VVFmLztSUlwuwZdCiRC2l/Kk33lG0dHD/qprTbw5/ZmTxqMV9Z8yYvelw/cCqjf/+6K9P9H9t4KLl7R+cvmJR99W/f6Ggbs3LPQbRnMF1WW0mD5q1NDW4IJjSKdy5prTH+klDl+fctXrZxm5rs9r27dWuY8e8oqHTRvWb0MVZPfnuKWXOMUCwWLTQ8eKH6u5TWpiTanKAI8lnpW495N90QCAhzctKeI/FxVnZpaXZWcU4pzgrq7Q0K6tYnFrUrl1RYUFBYfwOQGEM7xzvEdt5hxKeSwWDXmrNT0936a1esbSDZAKH1ZRuIuCwOYjJYXKk5AWcoRQByhNPBdhblgFRMxHuG90bnN2obu8KDjc3eYHM1py5DiFU2NqhNXTQOXMWz10weE77sRWvffDZq0880vHB5vXv4PB3les1tv2D02z76xP2YNvdezD3pT3s7N497JOXhMCeTTu3t/2dq9X3n575qfMjIXZI/Q7b/u6brOGD0zj0rT+wD/+wB3P2xr8GQKCCushU8W1OdzqUhlt5pRQDokeJazP8rQwGh88D1EYJNTvSOakf3feGku9qVGpqG4xTV8ojfbXWGSt18iYUtdZJXEnDlt0/edPztWvHjM+btnB+HauecmLUlAeov2bk6HHjJkhCcGFoRIcJs1jnI2OaCgRBqd8NhFraSI+CBGbICTupxI21YNTrBbMkWKwmUYegHGS5WbPRiyhjVuw2EAfPVEriM1kjLsUhtexzTK9lO0kQ1/dk29mzvXB9yo23qh9EHfeDXhAhJWwiKKAki0J1RCSQr20nattixUJOXfM71Bv9Hhc+CdeuaV3LRAIbAAjXdUoX16r7wqGgF3iOLui5Zpn1JodXKu1gsnFoi9Pi0DmtjnQHAR63E4fT4bythikCCP22ZKVVoUS+hp0Bqm51Fnr+L2UjHz5YPXLwfRNx36B+l3eeXrwWxYbNVy/8n+pGrtwd7tNtSfXsNFaLo9jTdPZ89ub/pXB47YrkEiRpzW3r+oJ09UfBJLnmAoG5dBi5LJ5U83Z/2GIGp7L7nGwzHPNQhS3J7yWaAKe27LkytvA6c/fPn39g4Oqa+fun195VPX3qwLunC2vmH9i/oGZlTdOCgdOm3l0zdZoiv/GASic8yQYLAMhwBiA6Q93NqCLLub9OUmpcstOLaHGCwAsItnQvZqjyadHEUVx6cz+0JMt+sjy645vIQH91edGont0XbPj9msiaPXiIVI2/NHhk35IePbMLh0yeP6V6/ZPPA4KflKlzBqAsnGkVRaCONIPUOstxn/MhJ+nrRKMzxUmcTl2yP92s88eVhKvIfTe2KDHRmKtlyd/2PpPpA3vsPbRzw4w1sz/8snbmA6Or7+w+pUPP8mXDl2wVvqx+wJu//YmVHWb32L5q0oAeXXrkBYa2LZl5056LnkfvwhP6xD0X5YAIN3pyAOvaT85494494cnCD133dnN3O1oEqNZDegiV4IHicLJoMOhs4HS6dC6+LeC2ulLMRKks6LWkMWHX6XqfaELKyMnTOhsGs13PNCxJNkz+Z/0Qg6GhAeewK698pKaNLwyr2caOScrsU1mzMEJygRWCYYcgIoBopDa7TidSq4jaQa/8RJkG7MortqVTEvILI6Z9PL1rzacn//ov0pY1S3t/raYhx5WrKDBA2ED6Yh0dqvitsEECMJuofkCEQsyAJOqq2jzatUOseZR82L1nz+7xMwlZzIVNAOBQIge7xQhgUfrILXa7jtog/71CzQq3qDNoZYbSkOzBpo31obZtOw24a8BDQx4ubWIXRk7UT9S1Kckrtu+bHgSEvqQKP1d3kPleHwFKDSZuX2mGBGlK3sc5EGO7FpnEzw8MXLlQ8pQsvpNv4K4ld9471NP2/hFAoDt1kaPi26q3zgo7lONnEnBvHfMfbr3iP964r4XTTjgzJSYsWHJ0V/3qF3eu3/B8lN07fsKwYRMeGCZM3nHw8LPP7T+w/TH+b/YjjwCBau4hdsY9BF+ZRr1AgMrEoJdu5R/4fBhELEUxdqM72c5aTGef1+IQVnvjPTGxCb3wfhzek01IufGW24c+AOIZzq8gnCYLACAbHrsGKMNHNDV6EPR/osTBA8ziYuCw7Tjs+ThseQz2CwV2Ou3PYeV9xMZBVchkAMkvnuAQM34FFf4CxEZ9KD5qXmxUIBBiM2mNMBxSoY3Sba1zpQWwlbVVwCXk5EIqmmhqKj93lzEgkm2zG3tH7IEWecP9w+9rGZ4ohslCYnXDUm9MGF2J0ihbnJBfkf59Rs7q4vv9Y9X1ozq9+dbRTwPhSMnYbk2zOnXtXqqkXKHH1tZM7NOvw5ip2e0XjzjcWDEhMjB/yIz70jFvcU/eGRvmVKrdoPJ0bltbq9R1v/YaDgTdn4hNzIa84ltA1MLCGETS7SCOQSAGkdoSIv86xGsg3HKMrOsQE6CUQxiaKGmtgtyAkWIwIMNxKIN5QK4xAIk3MIIVnNA/fAdPM+wIOhPaRNEtuvROycm7kHm7iMHM7wabASUqOtByowkglmHm5an5G8bOiYau9y/SAF7vYVQ2zqR5UUeUXdxLDtMT0SMkNXqR9Lhag0cfURpetbZG/AvZr2jRHOZSOkc5ztkqzrMIAf55rM9N5VmbON8PqhxBs8aRmyFqoTwG4b4dxLFrV2MQyS0hsq5DTACHylWC/hhXgUA+gFip9id54Z5wod3t1glmAKcgCUk+rogS11erXC6/JJ+WL8jcIsuyoNfbqiJ6Kri17tNEXW55EDWhHZV7uVhLarxnM5QhVqpNqbM3bcJ9eBf+bn/07S9xNlt4lIyKtaWSunqyntWxHSQcba5nhhhNYrmqS+3jurSmJdWx7jiVLwUx3sKsmLb5bgdRi4YYhP92EMegKQaR3RIiX4PgeGy65RhZ1yEmwMdxnW4b5z7CQrQJJmEDGMEX1st6ino0mXXgy0+0x2rMHLeOu0ewbTh8BHua7RiLw9m2MThS2DCa/3fbaLyfPTsaR+CIsWwrAOXzv877434CJ6RAQFkZnnRvmsAPExtcAA6rqFMCF0+a32f2945YHTpRoDazQHnjnES1lrm3+Fq4+YgL/ygm0lglwc7fxSoM1BZEj3qKzovZ1zsLv1479tEH9ykddGe2jnx04rGmh6Mjpu/9zy/NwbFk68SdWpPhmOUDNr2FDyl9dMMXV699l61D26bmvgOVZjp2ZRN9qTc7xVdOrI9LlUxpXLoVMfk7Nb7fDFELp2MQKbeDOAZzYhAZLSGyrkNMgA3xlRNMtEfCbHWUTvF5CmKjOFSQeO/frHjvH9+pMOtFUbKDBB6vWeALiC8fs96sl2LdkZoVarkRrHVH8v9lCDcaJGexM+zzQ42NZ9GHnuYrO3mL5LvvUdvFy4zXWq/B6ei/V+5Y9yQAqv0oW6R0aK94ppxcMTUAXpMJUu25YkGhw5Hbrl12RaQd5LrV3S5tj+vm0xpaZCBL2vZIQjWCo6Q2/2lnOTKUqE/1UYJv5ZAOKb36Lxv32p+OTCrfUnn27ofnjujZq094yVz2TcPf/v7+58IPi6dX3OnPyC0L3b917LZdPTcF8w/0mVQxcHZN+cTisqHF1YMuXO0r7Nv3562c52pXkOTnPL8TACXovgLUVWlXOH6L57V56vN2t3t+7FP1eajFc/Gz689fe+UW3xc/vP58whegruiOKsCNGRZehzj+cwyiTQwCqAIhKbtXOVDENWdkOJQLre3tedlIaF+WlJTe3ghi5y4pbYNtKyK+AqGgV6RD66BdECyZQU+xzqKriLgsNtBaO9R97viBxZsNL1corarUot3Jy/+qHSkOv7bLFExMz5TiAMaaVIb/wg7NmPnUc0VVb4+a/3xO8a6Hj/0reqcOO967tWbwurHswpy73lz03Mt7Jg1ZtfPpwzvoK7OWGon8BOY/+yddrEUqp/ie+4eMYP/9+yRWGwjyVpav5k5sXH9/5MVNo2XdQ6Sw4ektO5V1zXc4lW4kzreeMU+JFaqnVDtxVIn1ikl8vyqRVppEbn5e21993vp2z4/9rD7PafGcS1R7PsEQk1d7TaLX/gqAo9URXolZHHYXKGOgqI3xIgApTICovZYRgzDHIa79iUMMSoA4xl6IQTg0iG84RDrHQ4OYwA4CqBbHZ9d89VRlx1zyq6euqsJ5fsnUqhXwYN5jsTttkj7YRp9eETFSj91nsfLIR0+9LqSttY3QmLJw6/3b430QyITiIlAqxdlBMcj/lHpUk+6gRVqnV4kwil39+e/sK5T/9sUYXdkp9n3vr4YN77ll3OW+pzc8v7NpC3vppe0vPUtC7Ev2FzR/cQmlWcInr25+cGHXgtrefZ6cNHMlm8b+taaRbXjh4Aku21jXgbraqmOrzaLyJC1RNqNUrt0Vk/1HquySb/e8drD6PPN2z4+p45Ngi+d8fu35a9/f4vtcJtrzCSkx3Wh3fS2Ph2YhR9gJVO1CD4WTPAaDTSACKjsZTifKZjMqJ/QQ8tX1yhOfG8nPjUN6iccXE96Pp8ejezqVFHXsFCrqot3J8iefZP/q3KW8Y1m4nPwYfwOUY3tEGCUsjvv7PvxEa3orl8vQ6iZn76u47uxt1M+b2Kjnf3P2ZWVxBdGcfXw7QXSpTl4Si1SnX6L2X2yaUjNt+Dw0Xd40o6Z25NzmV4rxTJ9pvAljfYjl95r63Iuxboyetf0XbEBQGjL6zuy7cMOvu8aRRcWffLRjTHRO6DzXjNjutSq5e2KSf0PVDI8mmZuf107VNOfWz4851OeBFs+5ZLXnE/yxtZarrfrYDqw6wr2xGWIjpKsAWu+I2t+VyXex0jOkFJfNZpfsrQMOsKeYPHqqT+NdjB7q5euvRZPnb3oYUWsXUUomXo/W9JUVbx7J4HugOKR748Sz333/yd8fMwk63mSElTs38OYRzF9LmyID2Efsvwpjn83sV86KdcDaFQ1NOXQi58u3ce/ZMxo1nF6Nmgn7Y/TmxejV+puEyuv9TaJArLfsb+Iw6gkU6UvxFLggHe4Ot0uSrE5nKpjtqZKY4bc6eDxpBaOR51hGGj+Vwg8UUAc4b5zk4det2ia1fWVJO2TlvZF9aafq7NnSl1EYN4y9zJ7BYRgeN5RaonxdR8+Rfs09fmXXEH+ecs89LqzDiTgeF3ljSZmwlZ1m55QTGn6hNi32qy1yujAU0iAXCmBQuG26zkI8nqx8t7tVlk4oDOW1Mbbh0RHvSCKixdiunWg32pIyxcyKCIieFj7YoVjVRAeseV9R9a0q5rdyvYktTFkxnyvWs/Nzup6pu8B+ROnrBae6djz2+InL0aAOq4Y/e8+QDVf9G154buPm5xvWCb3mrjKRjN+7vp4xEwtQh3q8Y+a0KbPYz19MYDO5tw1mkLIPz3985rOPP/10x9NP7wBEE68Q7pH8YFF6wGWwWXmN0KJs3CSfKkwsE/Igzx1QzhIE0DR3nLfB89CcmUMWLuFF2u+WPJGTu3C+t3TBoiIAgpP5iG2lhdp+kEMyxSpMejflw753u9KSrHUfcfpp29njxj46a8zY3z3YPRTq3rmsqJu4b9TM2lGjps8c3qFLlw78AkQdn+k78TN1N5wPn+Szg2gC/nKrZc73En4mKLYb3o4vKU6BwvQ0olRTQpJEXXkDB/TOLAxZRpmn39tucP/KjIL21tHmqcL5rLZZnbvMquO3Tl1n1aldEci5Ff/FEyCCePMvngykw+K/eMIh5f8VUtYgffQ49lB7+R0HUNTpQenhP6WBBkscHEs5y+QZ1WF29yx63DMUTVyicNM3RdTpRZly061Rq55Od5RisXIk/bGKDPGARzmLjqmfcouq/e4LkcAKAEQZizSpY1khOWwS0KwXbHbQUZP2M1+x3pUgbyrhA/vjeGG9tcNjs9M6maNnb2B4FnXTeR1Tw7TF6DZldL0ZRcHuMIs2WRn9LW10DWe/ei9JQJ4ELUkjOsxJ7m6+QYbnXvbTY2Ow6D6FHh/7lTTBZZSVLOtqB8g4iCCHzeZK+dC1Y38ymWJ3vb5SBnteXszG7cAfyXB6EYzgPBD/URrIP3Wr6u+OqQ9OmDF94qRp5JtZj/9u9sx5C/icym8TiHvgB8gGOwAEwU4c/M4nELJA1RaoJelK5ZPTbBAIlYikk0WuCInpvPM3e2CJ+16ASv2UpGqjUBAIkMRRWhRNSeqtK6QAyGYBkJXxUyYgEkE7ZYLxAQJIVjbPWkkXx4+ZIJRzr1gnnuT0TQ2Xp3rTPZ5kI5Hl5NZ2wZDslYJtjN4kb/+ILklMTUvtHyFp1rT0tPw0qqdJaUlpzsxM6BvJlJ0W3iDhg5ZN3bwwdMsfKruRW2ZQbuRlt9evdcorVpPyolGwuJT/dUDsCHUKOz4AWfRHQvA065Z1snHLxtW7/oddaNewgZANO4LY+n9OPN+rQSxmD80rC7ed1/Rm9/puaEacl3tH9TwUsfXIpYPVzprl6o4iBXdYT0AUtDAtYc3y+EuJtrjkUwGEVlI650ylKvE+5ABA/HNTwuf9lc+BgItUcf0/AgZwQedwuks0ypTyaYjSqY+iqLe60l3E5aIWOZ1mxPuV70toergeGwR4g0v8V2eKi0otVJZJ05xV7GHcsHQO+0ESk9LSjDup6913x/KzVKdeX9THFGzb1v5TDDfpQ45bECoJ9+43cBcf0nCXXr/F8/43notvxJ6rVEnqc1TWG05X9cp+AAQRKWiHl2Knck80KgqljCAC4Aq1QvJpPHP6XaxCImp1FiUv6pwAUXstt2Ud9NrbHGJCAsQx9ufEKktsFtJBzroOMYF9EK/V+GK1mv8PflNJUQAAAAABAAAAARmahXJJOF8PPPUACQgAAAAAAMk1MYsAAAAAyehMTPua/dUJoghiAAAACQACAAAAAAAAeAFjYGRg4Oj9u4KBgXPN71n/qjkXAUVQwU0Ap6sHhAB4AW2SA6wYQRRF786+2d3atm3b9ldQ27atsG6D2mFt2zaC2ra2d/YbSU7u6C3OG7mIowAgGQFlKIBldiXM1CVQQRZiurMEffRtDLVOYqbqhBBSS/ohgnt9rG+ooxYiTOXDMvUBGbnWixwgPUgnUoLMJCOj5n1IP3Oe1ImajzZpD0YOtxzG6rSALoOzOiUm6ps4K8NJPs6vc/4cZ1UBv4u85FoRnHWr4azjkRqYKFej8hP3eqCfDER61uyT44DbBzlkBTwZD8h8/sMabOD3ZmFWkAiUs5f4f2SFNZfv6iTPscW+jOHynEzEcLULuaQbivCdW5SDNcrx50uFYLzFHYotZl1umvNM1tgNWX+V/3gdebi3ThTgVEMWKYci4kHZhxBie3TYx3rHbGr+Pdo7x4dIHTKe5DFn+O/j+W2VnE3ooW6isf0LIUENvZs1gf/LHojJwdpplCP5gn/5gi26FoYa19ZVFOJ6Sxuoz/q2Ti20IKVJdnqvYJwnhfPH/2f6YHoQF30aZaK9J8T026RxH5fA/WPW/8IW4zkpnIfoFLifGB86v0ffm5nbyRs5iaHR3hNBD0HSfTzoPugRM+hdN0x052KoHLBS0tdgpidAiEesDsgWYO73RWQz2LWIwjqnMe/uYISQtlbyf2NlT9Q9PoBcBnrO6I5ELoMeyHkNnIXGdv809H/DXNOTeAEc0jWMJFcQxvFnto/5LjEvHrdbmh2Kji9aPL4839TcKPNAa6mlZUyOmZk6lzbPJ3bo56//Cz+Vaqqrat5rY8x7xnzxl3nvo+27jFnz8c/mI9Nmh2XBdMsilrBitsnD9rI8aiN5DI/jSftC9mIf9pMfIB4kHiI+hWfQY5aPAYYYYYwpcyfpMMX0aZzBWZzDeVygchGXcBlX8ApexWt4HW/gLbzNbnfwLt7DJ/p0TX4+Uucji1hCnY/U+cijVB7D46jzkb3Yh/3kB4gHiYeIT+EZ9JjlY4AhRhhjytxJOkwxfRpncBbncB4XqFzEJVzGFbyCV/EaXscbeAtvs9sdvIv3cjmftWavuWs2mg6byt3ooIsFOyx77Kos2kiWsIK/UVPDOjawiQmO4CgdxnAcJzClz2PVbNKsy2ZzvoncjQ66qE2kNpHaRJawgr9RU8M6NrCJCY6gNpFjOI4TmNIn36TNfGSH5RrssKtyN+59b410iF0sUFO0l2UJtY/8jU9rWMcGNjHBEUypf0z8mm7vZLvZaC/LzdhmV2XBvpBF25IlLJOvEFfRI+NjgCFGGGNK5Rs6Z7Ij/45yNzro4m9Ywzo2sIkJjuBj2ZnvLDdjGxntLLWzLGGZfIW4ih4ZHwMMMcIYUyq1s8xkl97bH0y3JkZyM36j/+58rvTQxwBDjDDGNzyVyX35Ccjd6KCLv2EN69jAJiY4go/lfr05F+Ua7CCzGx10sYA9tiWLxCWs2BfyN+Ia1rGBTUxwBEfpMIbjOIEpfdjHvGaTd9LJb0duRp2S1O1I3Y4sYZl8hbiKHhkfAwwxwhhTKt/QOZPfmY3//Ss3Y5tNpTpL9ZQeGR8DDDHCGN/wbCbdfHO5GbW51OZSm8sSlslXiKvokfExwBAjjDGlUpvLTBY0K5KbiDcT672SbXZY6k7lbnTQxQI1h+1FeZTKY3gcT2KvTWUf9pMZIB4kHiI+xcQzxGfpfA7P4wW8yG4eT/kYYIgRxvgb9TWsYwObmOAITlI/xf7TOIOzOIfzuEDlIi7hMq7gFbyK1/A63sBbeJtvdwfv4j28zyaP8QmVL/imL/ENJ5PJHt3RqtyMbbYlPfQxwBAjjPEN9ZksqkMqN6PuV7bZy7LDtuRudNDFwzx1FI/hcTzJp73Yh/3kB4gHiYeIT+EZ9JjlY4AhRhjjb1TWsI4NbGKCIzjJlCmcxhmcxTmcxwVcxCVcxhW8glfxGl7HG3gLbzPxDt7Fe/gY/+egvq0YCAEoCNa1n+KVyTUl3Q0uIhoe+3DnRfV7nXGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOM8XZouTZemS1OAKcAUYAowBZgCTAHm3x31O7p3vNf5c1iXeBkEAQDFcbsJX0IqFBwK7tyEgkPC3R0K7hrXzsIhePPK/7c77jPM1yxSPua0WmuDzNcuNmuLtmq7sbyfsUu7De/xu9fvvvDNfN3ioN9j5pq0ximd1hmd1TmlX7iky7qiq7qmG3pgXYd6pMd6oqd6pud6oZd6pdd6p/f6oI/6pC/KSxvf9F0/1LFl1naRcwwzrAu7AHNarbW6oEu6rCu6qmu6ob9Y7xu+kbfHH1ZopCk25RVrhXKn4LCO6KiOGfvpd+R3is15xXmVWKGRptgaysQKpUwc1hEdVcpEysTI7xTbKHMcKzTSFDtCmVihkab4z0FdI0QQBAEUbRz6XLh3Lc7VcI/WN54IuxXFS97oH58+MBoclE1usbHHW77wlW985wcHHHLEMSecsUuPXMNRqfzib3pcllj5xd+0lSVW5nNIL3nF6389h+Y5NG3Thja0oQ1taEMb2tCGNrQn+QwjrcwxM93gJre4Y89mvsdb3vGeD3zkE5/5wle+8Z0fHHDIEceccMaOX67wNz3747gObCQAQhCKdjlRzBVD5be7rwAmfOMQsUvPLj279OzSYBks49Ibl97In/HCuNDGO+NOW6qlWqqlWqqlWqqlWqqYUkwpphTzifnEfII92IM92IM92IM92IM92IM92I/D4/A4PA6Pw+PwODwOj8M/f7kaaDXQyt7K3mqglcCVwNVAq4FWA60GWglZCVkJWQlZCVkJWQlZDbQyqhpoNdAPh3NAwCAAwwDM+7b2sg8kCjIO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO47AO67AO67AO67AO67AO67AO67AO67AO67AO67AO67AO63AO53AO53AO53AO53AO53AO53AO53AO53AO53AO53AO5xCHOMQhDnGIQxziEIc4xCEOcYhDHOIQhzjEIQ5xiEMd6lCHOtShDnWoQx3qUIc61KEOdahDHepQhzrUoQ6/h+P6RpIjiKEoyOPvCARUoK9LctP5ZqXTop7q/6H/0H+4P9yfPz82bdm2Y9ee/T355bS3/divDW9reFtDb4beDL0ZejP0ZujN0JuhN0Nvht4MvRl6M/Rm6M3w1of3PVnJSlaykpWsZCUrWclKVrKSlaxkJStZySpWsYpVrGIVq1jFKlaxilWsYhWrWMUqVrGa1axmNatZzWpWs5rVrGY1q1nNalazmtWsYQ1rWMMa1rCGNaxhDWtYwxrWsIY1rGENa1nLWtaylrWsZS1rWcta1rKWtaxlLWtZyzrWsY51rGMd61jHOtaxjnWsYx3rWMc61rEeTf1o6kdTP/84rpMqCKAYhmH8Cfy2JjuLCPiYPDH1Y+rH1I+pH1M/pn5M/Zh6FEZhFEZhFEZhFEZhFEZhFFZhFVZhFVZhFVZhFVZhFVbhFE7hFE7hFE7hFE7hFE7hFCKgCChPHQFlc7I52ZxsTgQUAUVAEVAEFAFFQBFQBBQBRUARUAQUAUVAEVAEFAFFQBFQti5bl63L1mXrsnXZuggoAoqAIqAIKAKKgCKgCCgCioAioAgoAoqAIqAIKAKKgCKgCCgCyt5GQBFQBPTlwD7OEIaBKAxSOrmJVZa2TsJcwJ6r0/+9sBOGnTDshOF+DndyXG7k7vfh9+n35fft978Thp2wKuqqqKtarmq58cYbb7zzzjvvfPDBBx988sknn3zxxRdfPHnyVPip8FPhp8JPhZ8KP78czLdxBDAMAMFc/bdAk4AERoMS5CpQOW82uWyPHexkJzvZyU52spOd7GQnu9jFLnaxi13sYhe72MVudrOb3exmN7vZzW52s8EGG2ywwQYbbLDBBnvZy172spe97GUve9nLJptssskmm2yyySabbLHFFltsscUWW2yxxX6+7P+rH/qtf6+2Z3u2Z3u2Z3u2Z3u2Z3s+O66jKoYBGASA/iUFeLO2tqfgvhIgVkOshvj/8f/jF8VqiL8dqyG+d4klllhiiSWWWGKJJY444ogjjjjiiCOO+Pua0gPv7paRAHgBLcEDFOsGAADAurFtJw/bt23btm3btm3btm3btq27UCik/1sq1CH0I9wl/DTSONInsjxyKcpGc0VrRNtGx0dXRF/FpFiV2KbYl3j++Jz4vkTaxKjEgcSXpJzMm6yb3ALkAnoCV0ARLAcOBjdCAJQJqgWNhJZDT2EbbgTPhz8h+ZFJyDbkFSqgVdGh6Br0BhbFFCwHVhNrj43DXuH58V74WcIkahHvyDRkLXIGeY18SxWl+lMHaIVuSc+h3zHpmNbMJOYuy7DF2E7sFvYMJ3Clf+3DHecNvjm/m38g1BYmioxYS5wqbhZ3S0Wl2tJkab50U04pl5CHy9vlmwqlZFJaK4uVnco55YlaUK2kNla7qEPV6epi9aMW01jN0zJohbRZ2mptj3ZWu6e91wE9vT5LX63v0c/q9/UPRiZjprHS2GmcNG4ar8yIOcycZC4yN5mHzMvmE/OrhVq6NcCaYC2wNlgHrAvWQ/t/e6w9115r77XP2fecrE4xp65zwM3lNnZnuBfdZ17E071sXj6vrTfP2+Hd8F74lJ/eL+Hv86/6D/23Qfogf1A+qB10CAYGk4LFwdaf2C+JfQAAAAABAAAA3QCKABYAVgAFAAIAEAAvAFwAAAEOAPgAAwABeAFljgNuBEAUhr/ajBr3AHVY27btds0L7MH3Wysz897PZIAO7mihqbWLJoahiJvpl+Wxc4HRIm6tyrQxwkMRtzNIooj7uSDDMRE+Cdk859Ud50z+TZKAPMaqyjsm+HDGzI37GlqiNTu/tj7E00x5rrBBXDWMWdUJdMrtUveHhCfCHJOeNB4m9CK+d91PWZgY37oBfov/iTvjKgfsss4mR5w7x5kxPZUFNtEoQ3gBbMEDjJYBAADQ9/3nu2zbtm3b5p9t17JdQ7Zt21zmvGXXvJrZe0LA37Cw/3lDEBISIVKUaDFixYmXIJHEkkgqmeRSSCmV1NJIK530Msgok8yyyCqb7HLIKZfc8sgrn/wKKKiwIooqprgSSiqltDLKKqe8CiqqpLIqqqqmuhpqqqW2Ouqqp74GGmqksSaaaqa5FlpqpbU22mqnvQ466qSzLrrqprs9NpthprNWeWeWReZba6ctQYR5QaTplvvhp4VWm+Oyt75bZ5fffvljk71uum6fHnpaopfbervhlvfCHnngof36+Gappx57oq+PPpurv34GGGSgwTYYYpihhhthlJFGG+ODscYbZ4JJJjphoykmm2qaT7445ZkDDnrujRcOOeyY46444qirZtvtnPPOBFG+BtFBTBAbxAXxQYJC7rvjrnv/xpJXmpPDXpqXaWDg6MKZX5ZaVJycX5TK4lpalA8SdnMyMITSRjxp+aVFxaUFqUWZ+UVQQWMobcKUlgYAHQ14sAAAeAFNSzVaxFAQfhP9tprgntWkeR2PGvd1GRwqaiyhxd1bTpGXbm/BPdAbrFaMzy+T75H4YoxiYFN0UaWoDWhP2IGtZtNuNJMW0fS8E3XHLHJEiga66lFTq0cNtR5dXhLRpSbXJTpJB5U00XSrgOqEGqjqwvxA9GsekiJBw2KIekUPdQCSJZAQ86hE8QMVxDoqhgKMQDDaZ6csYH9Msxic9YIOVXgLK2XO01WzXkrLSGFTwp10yq05WdyQxp1ktLG5FgK8rF8/P7PpkbQcLa/J2Mh6Wu42D2sk7GXT657H+Y7nH/NW+Nzz+f9ov/07DXE7QQYAAA==) format("woff")}@font-face{font-family:"Open Sans";font-style:normal;font-weight:700;src:local("Open Sans Bold"),local("OpenSans-Bold"),url(data:font/woff;base64,d09GRgABAAAAAFIkABIAAAAAjFQAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABlAAAABYAAAAWABAA3UdQT1MAAAGsAAAADAAAAAwAFQAKR1NVQgAAAbgAAABZAAAAdN3O3ptPUy8yAAACFAAAAGAAAABgonWhGGNtYXAAAAJ0AAAAmAAAAMyvDbOdY3Z0IAAAAwwAAABdAAAAqhMtGpRmcGdtAAADbAAABKQAAAfgu3OkdWdhc3AAAAgQAAAADAAAAAwACAAbZ2x5ZgAACBwAADiOAABYHAyUF61oZWFkAABArAAAADYAAAA29+HHDmhoZWEAAEDkAAAAHwAAACQOKQeIaG10eAAAQQQAAAICAAADbOuUTaVrZXJuAABDCAAAChcAAB6Qo+uk42xvY2EAAE0gAAABugAAAbyyH8b/bWF4cAAATtwAAAAgAAAAIAJoAh9uYW1lAABO/AAAALcAAAFcGJAzWHBvc3QAAE+0AAABhgAAAiiYDmoRcHJlcAAAUTwAAADnAAAA+MgJ/GsAAQAAAAwAAAAAAAAAAgABAAAA3AABAAAAAQAAAAoACgAKAAB4AR3HNcJBAQDA8d+rLzDatEXOrqDd4S2ayUX1beTyDwEyyrqCbXrY+xPD8ylAsF0tUn/4nlj89Z9A7+tETl5RXdNNZGDm+vXYXWjgLDRzEhoLBAYv0/0NHAAAAAADBQ8CvAAFAAgFmgUzAAABHwWaBTMAAAPRAGYB/AgCAgsIBgMFBAICBOAAAu9AACBbAAAAKAAAAAAxQVNDACAAIP/9Bh/+FACECI0CWCAAAZ8AAAAABF4FtgAAACAAA3gBY2BgYGRgBmIGBh4GFoYDQFqHQYGBBcjzYPBkqGM4zXCe4T+jIWMw0zGmW0x3FEQUpBTkFJQU1BSsFFwUShTWKAn9/w/UpQBU7cWwgOEMwwWg6iCoamEFCQUZsGpLhOr/jxn6/z/6f5CB9//e/z3/c/7++vv877MHGx6sfbDmwcoHyx5MedD9IOGByr39QHeRAABARzfieAFjE2EQZ2Bg3QYkS1m3sZ5lQAEscUDxagaG/29APAT5TwRIgnSJ/pny//W//v8P/u0Bigj9C2MgC3BAqKcM3xgZGLUZLjNsYmQCsoGY4S3DfYZNDAyMIQAKyCHTAAAAeAGNVEd320YQ3oUaqwO66gUpi6wpN9K9V4QEYCquKnxvoTRA7VE5+ZLemEvKyvkvA+tC+eRj6m9Iv0VH5+rMLEiml1XhzPdNn3n0rj6/EKn2/NzszO1bN29cv/bcdOtqGPjNxrPelcuXLl44f+7smdOnjh09crhe279vqrpXPuM+PbmzYj+2rVws5HMT42OjIxZnNQE8DmCkKiphIgOZtOo1EUx2/HotkGEMIhGAH6NTstUykExAxAKmEqSGMFl6aLn6J0svs/SGltwWF9lFSiEFfO1L0eMLMwrlT30ZCdgy8g2S0cMoZVRcFz1MVVStCCB8raOD2Md4abHQlM2VQr3G0kIRxSJKsF/eSfn+y9wI1v7gfGqxXBmDUKdBsgy3Z1TgO64b1WvTsE36hmJNExLGmzBhQoo1Kp2ti7T2QN/t2WwxPlRalsvJCwpGEvTVI4HWH0HlEByQPhx468dJ7HwFatIP4BBFvTY7zHPtt5Qcxqq2FPohw3bk1s9/RJI+Ml61HzISwWoCn1UuPSfEWWsdShHqWCe9R91FKWyp01JJ3wlw3Oy2Ao74/XUHwrsR2HGHn4/6rYez12DHzPMKrGooOgki+HtFumcdtzK0uf1PNMOxwDhN2HVpDOs9jy2iAt0ZlemCLTr3mHfkUARWTMyDAbOrTUx3wAzdY+niaOaUhtHq9LIMcOLrCXQXQSSv0GKkDdt+cVypt1fEuSORsRUwgrZrAsamYJy8fu+Ad0Mu2iYFhexjy9FIVLaLcxLDUJxABnH/97XOJAYQOOjWoewQ5hV4Pgpe0t9YkB49gh5JjAtb880y4Yi8AztlY7hdKitYm1PGpe8GO5vA4qW+FxwJfMosAk2X9n9X2cVVfnA36pzHNHJGbbITj75NTwpn4wQ7ySKfAu9u4kVOBVotr8LTsbMMIl4VynHBizBEJNVKBAfMNA9867j0InNX8+ranLw2s6DOmqIHBIbDfQR/CiOVk4XBY4VcNSeU5YxEaGgjIEIUZOMi/oeJag4mEB3PUOweCaG4wwbWWAYcEMGKn9mR/segY3R6zdYg2jipGKfZctzINQ/vxkJa9BOjR44W0OpTKAskcnjLTcKyuU/SVIWSKzKSHQHebYW9mfGYjfSHYfbT3+v877XhsIwGzEUaleEwITyE2u/0q0Yfqq0/0dMDWuicvDanKbjsB2RY+TQwOnfvbMUhiNPFyDCRwhZhdjE69Ty6FjoOoeX0spZz6qKxxu+ed523KNd2do1fm2/Ua6nFGqnkH8+kHv94bkFt2oyJj+fVPYtbzbgRpXuRU5uCMc+gFqEIGkWQQpFmUckZe2fTY6xr2FEDGH2px5nBcgOMs6WelWF2lmiKEiFjITOaMd7AehSxXIZ1DWZeymhkXmHMy3l5r2SVLSflBN1D5D5nLM/ZRomXuZOi16yBe7yb5j0ns+iihRdlFbd/S91eUBslhm7mPyZq0MNzmezgspUUgVimQ3kn6ug48mntu3E1+MuBy8u4JnkZCxkvQUGuNKAoG4RfIfxKho8TPoEnyndzdO/i7m8Dpwt4XrnSBvH45462t2hTEX4Bafun+q8jIzK/AAEAAgAIAAr//wAPeAF8egd8lFXW9zn3PmX6PNMnPZNJMRRDMkzmDYgZMRRDCEmMMUPJIgZEepHlRYyIiNhRUdYuS4ksy9reLDYsdOmLLC/Ly7L2CgKrrCJkLt+9T2YyYPl+D8804J5zT/n/zznPBQKbACSTvAEoqJAdtUhUJpQYjBJVAUrKSkIOJ1ZUOEKOUGkfV8ARiPB7E72m87WJZF58ibzhXPVE6QsAAnMufI4H9XXsUBh1UpOJSJLmQNWqNsasLkKhsrKnA/T1HCF9PQzSAPYtD5V5PW4lmFeIK86EcCRbObLp2lGjGxpH4+f0wLkjjU3NDSNGxYSMxbSdDkzomhE1SypQalCISvniob1lDuTL7injC1O+Mr/xmeJtxeRt/iJviJ8mmrjFOr0BJCZ3QAbkQFu0ypCZ45HcRqNJQkiT/LKsOO02s2Ryudze7CxVUnw+v9+tmKTcgEEymzPRlgN2e5rHaeOXyeeiisnJFagMOSsqSkr45kL8Tr450SfM5/y1V66pGvBwTV1BcYcDEX67QjQkbo8cigTplyVI2OHh/6zdXHO4+iR6SjoxMPzo8O21h2tPx7O2lmylNV/tY5Nwubj3fXUA/8BuFveBr74CoNB84V6pSnFCLhRCL7g7OijfR7Oy3FalR49AcXYRFBnsQUcgkAYO6H15j6wiAGu+I+Ao6pleFDAWKJZMX+aImNunWOpiskIVH796ewAqEzvV9gqX9nQ4Qd8S/1V/ScSM/rmsTP9FfNUNIvzuVlRPMFxY5PB6fY6iwsJw3/JIOOTx+lT+WzaR+xYWecrR7fWFFanqi/33nnn9+v+MvXr7mk933/v5Gy3PrN6yZjg7WFV1D5s2oGoh7nx+k2vvTrkeDT0HKlieXvvakkfecj/5uKnhm6iNHRk27a6bevTL+clH3ulVkX3cBTJUXjip/CDvBiO4wQ95PB6qo/len0+WTRpofo8nLa04mB3UgpeX5PbMLEzzKz4/tapOlXt5a1llpXhN7FF7r8zJ37o/iN15Q2XhvsE8RdajOqwFyrwFGETXr/0F9u9dNnZsWW9869X1azow9qe/kpc7D52mPRf//HcJFrR1npvf9sWX336EO7/9x7lqeUMn6frt8y+//ZD/JjzecOGEAnxvWdzjpTAzWtHbGjRhlhdMXqvLVZSWnl5kpSoChLJVtcwXSPea8vNLSrT0dEnTegyPaZIUqIlJLnSKhAV/pfBuhb9EbE53bYVIM/3S45hfiZ+7th8IFPHN5QuXcscms1vF8kiAZ2qBsEEEFQX7FnJDeNy+8nIF2JLZ7/77DPtk3rJhVV9vefPD+57CzCF98cr82+s631s4/vbxrKPf1XjT0Iqrh/+uafTMxR+9e++mxqZnxzzx5l8embstxo7PeX0Ju3DjoqYJA7C611hyd3hAtH/zpD5jAAVm4DM6Zjj5C5WIAIu9DuxCIB0kuvEBAKGBbSTz+L+3Qm7UZjaZqCSBqtrN+VQgmAMTua3joeaMhBTicTt9wULS8PSj5x58eNk9Z5c9RUrRiPte3MTKzvyHRd5Yh9vFygP4yq3JlfmyfHG+so1LyP/5yqgRNVjuDPclRSGvk7Q+/ejZJY89/OA5sTT7ifVb+zru/OEM7tv0EisFhErSJGUpbrBBOOo3ms0ypVZUVc0umUyqilarYrDxpN1aJrKQuykJwvwz/yPMUOCTXSqlRa6CiEzJy8U4J8DWf/jpM/eeOMZeLMKpxYqbPTyx088Oz8MKtnMuFqefm4gzAKEZPpUqpG1g5qivGRSjkSKAxWo2giJRKOFCysqS4vjNhQXCAa4Bxz1HEI+yNlx0FBextqOk9SjezW49yhaIHbGzuBtOggKe1wgFWVapDCXbdSNt5ghfoNCgMxLA3X1v++dV+eg/vIsdR9MJYWVcS5rISqDg+CuVQQLkSiTc7QoHPANIGq49dw6wi7GwgmvujZoUrrSRNsaMLqjsmfjnkYu4aU6SlJZ28xECNyqt0mMrM2pBricBidueiNS5iDcRA0ir4h+y4yQgGJP/DwLVF05IQ+W9XLoPLou6LYoTFPCnGT0jYkaV2kfEaBok8y+1kkYCeeDQnIEyQI2nUrlDE3kkDT3PzsfZhXMoxZHGw2OmTRl7w+SpLeQoW8gexttwNi7C6ewO9hD7/usTaELr8eOAMA+A1nJtTNAj6jJKAAZEs8WgqihJRgX9wJHOkYoXkf8iwR2RiKKqRRiitWw3lYdnr30cDzNae/8Tw/1L3sS5gFALINXpKDQgmp1pQxW86M3O8aoqMTlNtTGnSjATM2tjXEgCYfS3hKyuCkFHkzBeScI6WKhFVxLuD+EQLt4TkOo6CU5f1drrhvrrVly/dspDayfe+8EtQx7fuJG0HcbZLyyc1r+5qXbojtE1xa0dt4x/5c31r9hA6MYtP5DrVgijoiV5Po6KKs3MBOCVStFlgez8bG57v8/vq4tZ/Gilfr8pX7VqJm1EzJQGeg3j5/xX8ruWMbrG4oduFyXxMEFyQlkpkMeJTvhKbCMY1j/o2ykPlEmSr335KxvYPvbZydev29P65KNrX58+c92zfxv6+Kil76PnU1Sl6fe+l694//zIweMjUO1ZPnH2TU3fxqa09+l/6OHXAQgEAaSZuhddMDiaZ1epkRAzpTKAxyVzrnGh7JLreGi7qF1VqO5WvoGQ0DwF584uo3cpz4sCBzc9T9SAQPKgoqI082X2QfxhshCzXmZ5Jmoo6MvOYAk7gCWH6cudN5+98oSroZZNBoRWbuEw1ygDmqI9OZ36aJrbbTPYqIFmZrldRpdFA27ONADF4/HXxjyKYhkRU9LgYsIJ6e+pgHAkGUjkgUhLSBg2N9w3IMwpylMaKScT/n6efcC+PLN8xActmMGOhu+4bH6EpsV/yAgOoO0n9/+HnR2B5h7hr455LAPJ1+wc+1i1AYGhXOs6eQf4IR+uigYUp8WSlweZTnAWFNpz6mJ2u4d60kbEPGnUwENEvUTbVJbqTCjIAQJlPo8IXEUNdQEJcCAhMvd/gvy8Q3E6TmsbErv++Z2tRuuN/7f1X+zsNyv/vYhoN066sbVlcRuZiq/iWvuP7rEb/7LuhyPfsFPLMffdxfMnz7+1fu5qEc0RPdM6QIHLo14FgCDKRFYNMiWU1MaoAsLfupYpQwobhpDby4OfkoJ4iZQWPyy9jNLm8wLSdEtUyzvBB3lwOVwbLXYqnl6U+o3+Qo/Hnp1ttBtL+ihOZyBQXGwBS0Z9zJIGwfoYXGwTYYlLnVeWdKFwoCSqAj0/LqoW8qk7kShFiku3kK9cfCPVHyDedt/qpeyLL06zk4uXtU1DyfXfE2fPmrng0Ccjbhg+flxtq7zz3ZUzXhrU/O6sjqN73mrbXD2iY/Kzm89vbBp7Y/3VcwaOI3vqq674XdnlYysH1Ym8GajvcgekQQFURnOzZJfFEgyCCwqLtNy6mKZRrzd9RMyrUkMdR+Nfdbfu7DIBzCIaw0J5kS16edcXuNOdBXwbyU1J1ewxtvTOqxtHP/3+JIOl3xOz3v0nmr9Y+f2d8VNjp4xrbbm7jQ5mdazJdtYzasufW2r+83/H0fEE+3DTXbdNum1+Hfd4stOSZuvMURh1OXnyAPjtnsaYXeumMPAnaOwXTOb4NVYT72PqU+xG7xcf6mPNQAQX6/IUcHKmcllV1UUlBRXFZdIaYyZNUjgzJ6Rpm8u6mKrApzM0vUgYbrTrbF2SFHbS18Xa5GhSmF5P7JYqZODSiqKajIK/VYNEqQIEZRigFxShVFwJURhGD6JU0ZlDP443kvW7ccNSPH2abWFfCns140peoYDeNeZHHSqlRgkMcp00ViJSV30QKhkjagSue7JMQH4304/FkrTgKC9Tjh69VLueUScBrhFPNVAUJJTKEur6Ce0u1dCFuorNZH28UayJb2IaDjjNtKWsWmioXPicrpB365FYFc3LTU9PA+B2dlqdhUV2QCMFCAazGmNBl900ImaXkg7mVCR4KJVkyfpRJFR5F86oRckaXOFoe0m/7W6YevPVY5uWvzf1w3P7vm99YGyIHU4139VjH6ob1tLvqqpxR9u2r5m2onVI9RVXsHUX9eMTLkxQdnCc6AuVEIv2VCsq3G5XOGzt77rMZaWBtEDvNOgN0au8hkhEMg3QTPzqkVUq5feAklS7rOucMleiPU7ivc6kQtuiYCqrfNTdlVF8fxLxCKgtj3iUQC44+jrzOa06UfyDSESH3x2j106vnpWmTXnhlT1o+UfT/qt9NdGau79/Zhf73+exCP2T2Pz/ZefZXez6I/gIyv/EkRs7Yf3IFpM1FG27n5x++NQ9Q/otPPTGQSQBH/Pd/9Yf/vjjne1sx152gh0p6f3eKHwYW3/EZZ93sA627uCCpcfMzwj7AIC8WN4IKljh6miAWKkBQZHNZgqip6CSZLOSmpjVSs0yBZocIpTouZRiZWGortKL8gsDiITjI5Uik+LHJ7FXiYTziRJnywoMgWdwNFstbzxXRcbikdvy72CqiPvXAaQznI/t4Idczsm9VLdbktKzzeY83vfZ7QGDlqalDY9ZNLRSTbODPb0mZneCvyYG9BLcSxY9KQVDSTe5ArmSp7voCQYwWfE4HPqnwOu4AyOYNn/C/fPZh2fjx7C84/aZ8xev2nXHraxT3vDKpkVrHaacdQ++/xGdXTuy8Zr4NrZo3PgNgDCXI/UBnh9eKI36VZeLN+NWnxscUBNzSKpskmtiJleyNBOvSfVEKuQRD2+0Iw4l2BUdoTI+ZiikBS+9h9OfOtrxL7aJvdiOkQOHDrc2tEs72U/HmW846xyGi3DSZ3j9azd1FvUDImwoz+E2NIBd1OtGAIdVkjTZUhOTqWTlLbMzaamUcEELnGVzAbVA0BHKleew8ew2Ng534wR8gL3Dxq5ZjO/xGuQP7A55A7ubrcHDnUMBdY8RLs0Mg6L5BgnAqphMiBbFWBOzKNxLAnII3zehaKqJofOXXkp5iCsitPAkbol0bqDV8RN4ijmIm4tl7zK2BLqkUsalGqFvNN1AqVkBQDQJoSl5QlZS0MVSLhaCX7P9dHD8OHKMEwKWxLu8KBdxL6ZDTbQo3e8nNquVEFemy2DIsGlmjQdbOr9BNkt+r+zlsmTu1FB3wd0z5VlnstgW8BBwKLpv9YJL5RlPdMKNOALkU1L14E93sr+yVfg43vTxgZtW/GXnd1vevKGVHafhuOnyAlyMU3AcPjDybB377rOT591Y2mUHeYJu/Ug004jIzW+QJFm2GGhNrMaABoNsUijK3QmbMnfKFN2XPIHtjr/NdmE5uRrDZG78Xj5t2EIGAOCFiawBT+ozgRw+bSAGXiPLwM0MRsr79e4NCw4Rxa5IJL6kRnJurq0bOKEZy79hDV4k7gVL5JHn1l4AdgYS+tfxVS0wMJpjIcRkNiOAzUBl2cq/UrNZoXwP3VtwpgBXF1eWAOXEQAdVfSMRDKBcx1awhYvEZm7FB7CZETKxJf4D39CN6/Hf8XkJ6VIlly6LPUkqBVCQArccJKJUl6GXoPq6r3PD1MsbzldfSPxvRcyR3dAvmukGo9nI1bbxUPHKisdJjEQxq9QGilBcN36X0mUp6hA6Y9DpEYujXuXykscVRBpkK4wudhzbcaSC07GdfUgtRrZEms9Wzok3cw1WSi3nqklH6R3oPr8kYcedOm6WR9NMYETFagVwUFlRVM1MVW5RVLtHv11adI/EnAKwL1KEcM/JO9nv43fpSiwh81U7+qQGdrQtXseFv4FZvycdQPQ8+VKfDHgE0jgAfBZF8RpdNTGjRO01Mer6daQROSBexQQy16Hxpkj+kj3BXubXE3gz1vNr/PlDb76Bs9nSNzaSY+xxdivejVP5tZCj0mP/OYvf4smfoAvtpHU62rkEFkhGowdsNrvdbQXBV3ZNM9TENGr/TSzoRn/ZLXHoEyAo4ckJSx+au+BBspEdYacX8yA6iCb0UGXmlKkTd504Fz8rb/gchAXYat0CdkjjEZynUFmSCDVIJg9AhmYypVOVEwBXRFK5UWSV22N7Ev4uHU92T9OQe+LX7PPaKziWzWZnfL9pJMZW1bO5OPS3LSUP1S3lg9poocvnk0ySppm8njQw8cTzu4wWMA6PAZgtFm40C/WaRcikzJbSWfPzuXKqQ0sxKLdfgl3BF0A82brsgaXLW7gB12EPzH7oTqxuZWvZKtp73M0Tm+Pz4vvlDUeOLdxZwVwPk1KRVS2cQX0ce4s4n+RlpKcHICC7LeCGy4rdAbAELNlGX3ZNzCdRYyq+uhvwVHHWrRpn+IvGGoVFl/MhDadWMcJP9LZen9cr+din7JuOx/ZeN2FqnzFL7767DtWvZu2f2TrnyermlsJrn977BC7f/lkz5g4srx3e8+orqypveeqmzf8qL/13n8KGgcUDKqrHbRP6FwNIYiqrimdLCgBFNBhVKlHOuxSdv3y2lARgcoLtYrOlOn53IGEMEF7k+dXC13JCQdThQHSbDQaX08hRhsdSYuuXVBAOtyLx4BHI6+6CYLnlEXbyLfYFex/D9zz7BAf0ztqVZ+7EwHn6YufCPz33/DraBqjXfyHBI2K+RonRKAOiVZYkC3BDJ+q9VNpUJOaj+sXtVx6h57CC2dmLTMMKdPlKFXO0a4DY+dTwvZeN/qJLhrqRy8gSsx+T0e52yQh+v2ynlszMrKwci9mcnemSzdRvt6NJiOSi+EtCbgo1UyM3WkiKOMKJUtMlGvCIi78nPihD2fPbzWFJ6WPdxqngfix9q9Sr9HQdwoJDth5mUy/nm1hKoRixV/mpUJxwVT85trLi1EAa6twb+aS+9uuhNBsStmnSbVMVzTXLnPpUo6oYTYpJ0C2VLGYDkWXJqFCUkhDL9evG+ooUZ3VpjZj8Izex59h6fnXg56wfNmF/DGMtC5Pi+GHyHdka/47Y4j27dJCYyF2B7wZVlZEQEERvNFFF4QqiSgVDdslOjEH5Z65AarLLowIDZAGWchEZbA/LwDo6mozsXBTfQUqoXleVJiZ0RugfzTJISFUVEExmlYuSRP1I0IAGUcZdOgxNpl1qFqqPbALSzPPvkbfjTVJ6vIrs30m/RXi/0ykkLWUbyWw9T7KjVgXRIIFRJlTBfN2EuvH0BNZX4iUpmc0y8bOPPmIblXMHz60Xa1gA6MDkVFt/ZIKYnGpfnBa6sUmAHY9/mJhqI4S4fJ+QL55xoKIY+VYNoOZTiaaCvQtCfCFHMMy1CH34IX7GMmfKjQd/UoR8AzFIA+R3QIHeUTdBWVYkSTznFd6SVJko0DW+xLKLeyTRZYcwiGjADQ/jqVO8uP6KGOiGzmqyKN4maq1OtpHWXhja9SRIRonoRhEaJZ5K0NrOFyl//vMAAGKNdIQ+qATAwK1gBjVKRVTIdwCUpB/rioP0XWLww7EvHPD6PGRL5ZkqbKpcLx3ptW2gZ/z7GYIdmjju9pfm6E8Zq6OFTovBQvLy/P78LIMhaEkbFrNYZLfbPjjm5jWdnDM4JnvBk0Az/y+ZVYSeXlcUJWdMvMcN9+1u8h0omny9N6YT+huGr1r0xzd+Or/5xbv/On7T8Y9PswO/X3znY5MWPHHDsNfXvfono1K6rn7f+K3vx32E27h55MJbxwOBFVznDsUNTsjh7BvIojRg1Mw2n89szrWA2WPUFFDSh8QUL7iGxEC7mCz83SHi7H5mUeZ0aISzRVANCgTlw1AfH9d2D8WobftHX+7YNsMT+hpLLZbJM2ZOJJNvaZk+Q5rNdrPv2XH2t6XzFTdbPuiJ9jP3rwh0PPOXNWvWAMLoCyfoMWk2eDi6esRYymclxCubh8RkDexcM++lZZJuOTk32SdwmnJoYkjgUBQyIf4DZqJx81Mjh9525cmTzcuHVf/BTQZgFvauOZFVwBH49ZIydr4kH4iQK81M2CcaDRi9Gi+obTZhqFy7xwIOIyi6fTTdPt5ft4+oT4Q+ecShOXlPGioU/BLkji3iOnVPiAnZ9vHnOw9ON/mw7Jv+1omT5kyVp7dNmDnLjWVoRx7zq9vG4YSfTjyy5vt7ViWNk9BynD61y+DMEKROSUpzOLKcJlOm3+OkzuoYFVUUVMesmuoZHFNTel5aloiry3bI3RbgrbNeR4XKwOMJ6AVAxMMtOP2GaQZcT2aVs+/Y3zDt7LdoiJfID985vmNc3Qb61PyZM+d3NmAPdGAahth3Jx+789Eel5+4rCjB7nSOkgMeuCKa7SZElSn1+qwAPhndyHVz283akJgZqJ4bgp8v7QVDiRwWFgxH9KfOeieocBWpiZ1l+9eu3bj/ufm1o2uv6ocGOq9zCZ23rKHh3ZdLPsoafsVgoKAwtzSV26sYyiEKd0SrzFlZAwZIfRwOUqzmSkGUpIHpPXr4fJFg8Kp0K1jRqlj7qv2GxYy5Eke5wr7FpDpWXFxYWDksVqi5e1fH3BkXz+n4pxIOWz79gRHv0LneqJs2FQ76ewKfPao+pSsqEvmsj+ykQFfCF6ZeRcGFyUQK8v26El/4WGzqS33OfxjpXbL2ndc3sTfYvm9+vP3WksHVg5tvOnmsZKGTFc2buvrNabOfa5w5/drrmura10otT/ceNqZjJ5Xzew187smt/1i1bPw9We5Roeh1xYVrZ732vkM6L1UOHVlb2WcEHT5q0qRRuwBhBYC0lmeDB8LRdATw2Y0Wg8Fo9Nolp1MaEnNqJkCjR6D/JfU5336yUOPaKqJJEuCQeFQirWX7O+6YxfZjqapqE/61bQ958LsXt8S/40CwpeDekav/vh0ILAPAD7lsA1jEZFcyGsFksprtJg9Rr4kR6DJ/ZWoO7uobKtNnnyJUlrW3X3ttO14phMgLHn98yIjzPqkFgFxoY259XSt4oSTqd/L0JgaDT/NcE9PAaBctOk/sjOTEKYEwCRGJxwB6tajQpMDBcxoHXzN8CJbum6GLZe60066mRmnd+eJXN6mThXRIWPMH/Un+NdGgxLmTUKrIsmYzWa0Gg8lkN4P41WCzUcXkofbu2oTf3cjSZdpuokXRuGOyi1dx22KswGZWhYd5AffOIrF9jYxdh40sI74Et93MVivueDXr0gYPcG0ouF4DRIkAevQioLvExgPivyvuhO7qQJ5BQRgeLXS7XPrsKDMzI6PAajSaTPkuq9WRKzu46XwOzWzPRJNH7+G7krl7+OC8ePqbjJDCRIiEfKFykdziVfBd8q+ke9n++uvnTGL7vy529F437Xwso/dL097ZwvbVXz9jOnlw3rz12+LfSS1Lh1+/urZpy+F4kfhtxYuQjGCut1tMFxHAq6vrscoOoatQFU0Xx29SyV/XLRG8TS0ierkyof+ZtWWXEPbn7boC9dce3JHE5yf0pzhpostXLJYMcLnSvcYhMa9mp0Nidu8vu/xUrvPeVQMOCCQs6MzrxGVT5986ecr8W6dQmX3ELvzxh7swGyl/I6Xt6/70Qnv7mhfYKbbnQTS8jE7s8wA7B4LrOep1cC1ckMMn1Hl+RVFNlKpZmqrlcuQEq9U9hBOEwa5mQEaKzBKmSBWoSQVlTvPepDFCnPndRKFJtuemosq2GZrG9p/taZv8wfaPbt58TGf7vePdSx/wsv5K9SPtbB87/T/s7H10mU722JDgM67pTN1euaIq8dIsyh+TpOUZ+fg6PcNnz/ZanE5V4I0FhsQsv8m6iSfIBUmS5S2dL8HBXl8ook+LIkFBaLdMkafPPzxZ2v7R5zsmPXeFIQMJ22e1lq48uri9oOMZ9uLa9lNYiho3Z9+6xqU/bcBDAybXN3ZFFJ3LddVEh0mcejw5BCxZZVnUS7wGFxqlMrTMRy+JIqpdWewrCD+6iu3/sre97yvSbCP7xLR8SXyH1LKxZTYkqp/1XIZ4dpmjpLktAEU5bnchWNw5lhxTli9rcMynUdPgGPX+vJ2/2BgiqPTHK2HB5clePsGgXCkPt082oetPnbx1/bDrDtW395oycuG8yJd/3/Xu6MZHa5Zcv2zRrf2wZn1HILfzsvKx+b0rCstHz73+8VXN/8y//JriK/qHR/+30LeE6xuRa8AjToRYDHa7y2UyEIfB4fWZnHbn4JjVYrfL3HVyQt3QpktOVnRhgnBcxKOXvoLpIyFPwCO6cjK3bsas9tdeeHRt8xasYDuu+TD4aeiNN0jGwgknTn4e//yqK4UOT/Gc4zM+cENZ1E8cDrfby3t/j9NoJ7JNtumyPcmJ1sVDgItr7tQYgH+grxdrpR2zt72PpSLjsXRp7XUHt5Mj8dki4Ynt/EpI9JkPcrlm6BV1m0GWiYgIK0G0GNEuC5llKWndDU1X/x0SbTfiOtaElf/INyryZYexkjVJLfFF86aMXUzaumS4AZRtXEaWOMsoSyaOIVng81ETVTMyMjNzVEXJ9plMVLbbMxQ7yDqidR3RdPz2LIDSIO1WQ8wBsin/pGskRZpuUfew19lm7LMwJ1eRcrT7sG6R5NCsqBgvN92NPdk7uARPdt4vtTDH4m9q1lxH/PGvvE03jMkcer4XnuKKI5gApOW6bWqi+YoMaKSUSAQlGWWzQVWtfIZmMSoUAA1mj4T2S2cBqaROkYZeq3KlhdkClOu/mD2BI48cxZHsMWxja46fYO2kPwmyZ7A1fiy+DRewhcJLzK17ycs1KTC73ZrXK0koahm/Jgob/pNT8no0p9XJMTHDAFyVskQJkKKvhBlTUzxHyokifvTqgNsSaw9mmBRz7n4cwoqu+vcfR9RErqqfl+fkfr2/YcZNo8ic866XXnR8Z72xNZI450HXce2MIn+oKqkIYDYgmvQhAm8c7YR/MwyOoefSIULSSMJGySlCWEwR6LrOB4nC0uhAZiCmDrLp6+3xekDI4T38Id7D54ipCHUbcnIcfn+uNTMzIFGXy8qjKd9qSbTzYosp2hbbF7bnuBrm+REWRw08Coc18VTQ4xFQ6+EJhDmL2m6/c/OZG4cpn31T3XpmM9quH32qucGAVz7Z9jEdXMUObcyzBF8xskNVg+knbU8BIO5gJWSlYgMK7tcIpZJMAaCyhONDYlbqCOKOo0cV29lA1ylOauB7yBN7yOHlOmgGQ75bkoI52TabW3Z7qCzl/3/2IIuHzuFynuSi2BZnlftyiBSnzxyCyzwcrImh4e0Xbhz2+9mfKtWtL7xTP39x26LeM2aFPyFVQ7CnuWmyw5K3EXsOrqIfh2dPY5tNjY2nGm7QTxGQIqmCtoEHIlG/Ag4zmKnd7qNeu82mSJSaHQ5QoCRU1lYi9ElBdqqp5pwa1sv/RAMmELwQB0baym968pqFwxaOC99ePv7pgf89chFZcXX5l1NzcyPRii+nphf8lzhBwpbiQanl0rP6Dg26zurbad4v56mukCugE0Wi7Vh7JsTasSV5lIO0dJbKBcljHAhLOdJqfN6cwad7QYchPV3OyCA+n4mYMrPSXCNiBtuIGMiGNH4pGWmKygXqpwH4S8+ePzvOII575nOCTh4R15lS69q26gmSEBt94OCr7YtF6z7vlm8b7mpdcN+rL/fHcyhjZk77c8arjmflv/Bn9kZObzbAuFFEB4A0ST+d2BztZXeaidFqTfd6iV/zO51ado7Fn+avjxnT0sDFqcleG3P6QR7xs+NNXUfUIJTSVqjbjT+pBpRfbpXXFSKawsFwiBuQbNyyZcyzs2sbcS679w9k3/mvbhr+6qufy7sbvojGrt10dOm6WtZ5ttes1keObtl5BAjMBCYFpHXcnkW8R87TLC6j7EsnBrDZ8jIhM/OyYp9LSycWo2xQPZ4ctYBHz/YyHc11H2qb9S+iA4oURXyC3SM+0WGqPrVIoJJaFCmMXFRdbixfuGzBqEk3j1qwfGE43Pbogt+Nn93Y9siC8v1T6+qnzxxRO50cnPC7BcsWhCMLly6MTZs8uu2RtlBo/iNtYyYOnz6ttm7aDBHpCoDEp+PghZnR/7I53U6Plce2UaYyMYkJqxeRED/HBp/idDkbYkCRuuwmm93WEFPtdgt6FMsl5xX9mtiW3kNfypcpEhAfkgPKkCfoEXdAGF7cGCBD0YAVbOGWH374gX38448/vsOW4BViZBv3vHrfq8eO8RdyHMhFiKNCMGoniiKGmUaJSlTVsUcEbCpFdAhyJGBIAFHnAbag8wAAgUm89lnw/0o5D7g2jvTvPzOzu9KCJNSFaAKEBMYHAokSuQpiY04OODjYsWxCcjbkNaluuPdyiXuaS0jHpPfeE0N68fVO/ObSe+8uy39mVlqEzr76oeyi+bG7U3bK83yfkUZBGZwCMyKlaRaXRRTLC6E4JyfkAld4DKmpsbkrK0ttpSafxzc15nHqTVNjepQycUvmivi5NiuyMYtA0qyNo3NOVr9OFfZJmt75WUW7VMhOWtE4fsubj9zRP33SzuaW6LxFB3rWTJj4xSuvXdHyYsOAb/bpj257c+OS5s4tvmrim7appHXPputbn8kPlVdURssit194/xklXdGr7p3261Hh7uKKUGH0uu2nzi8Pxya1V5qmAUYu4UfygiRwVi0/YrQaWIvIdGcQ4pBB7dzU9snCdpLZJF/SOXJNjdRPPa0uMhVd2TKurqk5Mq5FXFPXEB0/7ucNExvqGieOb6wDIIw7lSbR99oBPqhmvm9ikm0mm7/c7yzPc+bV1IrpYEmnX1mlhbZglpActKMVbEo36zBrHWyifBGnSASrw44ZvIhr6bwgFCxiuH4R45HIul+c91p4c3j55tf/fvilPddGFx5b8zJqf5X9DCi9v/m10vvcrj6U09uHsg/0Ke/29invHSBfX7VJ+TAv99nwkcNvfNd82xjlI/4/Su+rLyi3/ObXaPaLTJb0b6xlBfCX+DHKMLqgAOoieZk65HLlmXXU56PLK/RmGI2e9HQbys4GEGweShSEA0F1mAtak3BQbR1SPGxVVo3K6irbp3YM1ToJV3pGr452r7n58XnrWi6tr79h3tY9yqTy/KbYvMvxsYvGRLrPu/BCWegef0l+cNcmpeGP/qIz6oqkNPas06Fd6BEEkMAIbZHRaUaDTKd2RMKCgERqGDdkGNkrBpBGCE4XBIMoIpOMsR4lWko4kLBqJI+K5j8Faab66Q897w8yR4ALIR3yqYfpaPGg8hFyDSo70RG06A12/oayC49HL1E/s9K3DL2QNXzKGb8fhTCZCCJkRZgzSkcQkogAAdYJoQTf6LXQWZQQHjx2hLz1I7pgEIaGErEHWAIzAAhaezTEW+S5kUqBYFHUgcViJEbamxB9uT/ROLFE8QLBIegdsp5+naSN8spKbara53ErgY4FlFnoIwadmhP5X7VaYcvuz5QHAu8h/cO3K+s89eFTJuceP+dft9utd0xUFqDpyj3kqh3K1+H6uhrlzX/ZctHQEckuSNLhJG8MjPTGCNLRbwWDZH+Fr/6Jm7D5hAmyIDMiQ0ZGTrbVkMkqRQ3FUq17vL06HSowmDyctbXd2N5201ln3XjW5a88G6uvnz2nLjJHWMg+7W0766bZL10emd02YWJ7G+NFAYSwiCGdcx+ZGTqdRB35BoSomd9sMRrSZYQkAYOKeoYC8S5MM5WnxriwyfZwnAs9I2/h3kG0RVlFY12UNylYiiCAo/gZTriVRKwOA5LAgiyuTNnkwQ4Hyucer4lJXb96j39EPHUF+JnjK/5+briipGXeqiuf3np9+4YudA6O3jbYEQv6S2bt37Cle8be7rMBwVgcxo+Ir4APJkRy7enY7QbIl/LTzVK65C8mdrvDIed4PSa5IIE5pbQ8dlABTRX6S6xu1DgHrezj3QjuuaN9/n1P7N541ards5oXtJ3REgwFWsOdE/b9v3W9wlu7a432i6at2N7wzOzzq6tvrAr76ePuDExYn+qLI0JEDyCnCdwXdyjui3uFjR/VNMjMIUk6ao6YiGZWHZ0i/DX75U5H1aEgAOK2LmrkhkxmMUmXJFnOsjrBQR/drXNlOGl7yiCq4Y2Z+zTTkbYwT8qwtv73xo0CxS6XhZtDZ7WvpVaAD0ZnlC6fNWF+vigy+yj67YoVdz/PrAF7Z8wo/9mM65SDUhQQLFSOCbslO2RAIOJINwsiAoTMFr0emUykKWYSWc8XiHtk4gMlbe5qgAb7UsMIa0IFwu6bbumd0PqX1/72IW5Tjkmn/3QfCVmPHEWCwiKd8Cj0e7KGEUURmUU6Ebk1RiCQCHSypSLhfEr/+2Eqe2hQsaNeALBCVcRlNjI7Fh1Y7Gaz0W60ySYW9pXNXt9QQI0EXB1/3PjAIiZPQYprQ3RWgnr3Xd88KXuOu/GW5v7s6Kwj6xc5btOZJpzh7hmf2cktXDiKGxPRSYI8MjopD+WfMDoJeePRSb4QbvyciNkVzReismdxFD2z4Oyi0vHr6MwOwnTUfEt8ic9KPBFjIvYqgzhkDw/xTGK3kxc9YlKPgt969IarH3/wwP4nFG9dY+PEiY2NdULbnf0v3Hr7wAu3dHR2dnTMm5cy6s2OlKZTy49OL2AW1Ib01FNiGh70BD7YIdHEB79/Oej1B9UBL+6NL0aoFonqQehRdg4ip/LxIFqsSMPn2KuMXYbaUNsyJZw1fMrGrnIA6Qpa2n5Y+TuAYvg1fgUA6eAP5Nrjj4L8IMFW+uJUVye0D51Au5h8T7W6B7CZSZlyNlXeJ75ClUs8XEnM8as+Eb9qmXpVwDBeWUH+LLTzNU5DpKiQug4YJk0jh0pMoyDbnI1lQp0JPk9rzJdhoRy8xZvKwaN4g9Cm5HHsnddbrUub3bCVWHLF4ldiF1wYPjM27aFzzp37w3lvHP3F7rOrUcnw6jY6d1dT86yJ4eiY0sOnTO6//YLru+j0cyyamXhHhoZU2lu3GPuhiOexHiQ0HfQPYqfoh9HVJ1B0w2//heIgzFQV2SMV52iKgYTCOlIxU1N0cUXaQwR7uWRYkxbXSNDfPYvXhpfEa4MpdD7OPtrg4sg4yUbMNmIRLCjNZEJsvgbgEETRbiYUvqb4syENGQkj/JFkkzkxTAQrMmlscsKiQLvUAAeUNb8G7yQ062PCs0QKkEYsI9rR6nzH9imOvcoLeLew9/ghbKIUT+hoLlq5jiPvcYqZDnXNrC6WKXZGjNP8+VlGYAXOBfY556p5+ZaodTT0KC89ZE+UXqqiG9pSFPdShT1JcXDoO1XhHnmNmZqia+gnXgMYFag1wGbucZ7cAJnQGCmivUCW3ep0GlBamtthAIqVWwGovcRJi9eKLYy8TgmP0+BgddahWmkscQqUlpiPo4MhBwPPA1tV5FzFz7cKwm9+d+CzzzahATIdd1Du/G5GoOPWnR9+ofQoyl1qHsRXeDuriLez36eUA+dUeTlUxtt7N1fgvJMpulHDv1AchOdUhXek4hxNMZBQZI1UzNQUXVzB2vvoeGkj2IAMglnogXTIjaRLBGTZYORGZXcgqMUn8260FqnLBlSM7lL+uB+Vocqr6Rhetkf5tfL7vfj3qKxH+SMavZf++VuaSiUAhD7DLeIHkgA2yIZCCEdyXJ4cuz0tB9LAW+TMK3Ab3QxXJQWpdOWImbyK8arGGFaJqpEG2V2IO/yqihEFV1Wm94Xts3tnv8iA1RevaL1x1sDRP56CjrR2UWL1/ZBiOG0+WqzyvXWXXHDpANrEwNWGNfM3DSi/fHYJ/rbsp+8e6j5uKR4aUmlIXgO18Vocrdaz1uOkKrqR6V8oDkKPqsgfqZipKbq4gr0RJcl9kqDwq4yNv3kb1KtYuCSJSmbrqZpIDiOjjbIoSpJTMDbFZEdTTJAFWdIRyZowKGrdjOZBjePIDroW0tZGwh2UUz1yNcPaH1CQ4fikjst3rbt0NcHv/agMUij5c2Vc18rz5/NZJM3JfMkD1dAaGU3tegXFxQDlWSZTbXkgUGPKKtBBcbEui2SWhkqnxEIQcFgyozFLwnGq7ZUx0g03TH/aTYLqcnOkuuX8iaFL8zhXsVAn4a3SSDRSWl1/RVfoo3fmXTau+ubIbfnTo2vnNjQ0TVjXsWQjbb4+hL9FfuGvkV+cNqai1JldVTJn7srmu+7JLfy6KLhqVGhcaeOylsh5lbWnl49r6TrnKPVMv/LO/azH5ASbVEBr5VQ+UtQfAPb2jbbEazY1vfvCE6Xna+kHfxhi6RUj001a+kAasPTikemClt4lAX+3T+GCYcUDmqJ/lKrwqwogTCEpQjeUQBBOgS2RydU1JDM/P2g3GoNBuabG7/GMKZPlsC/fW50fjVVXsyDp7OxQNJZtNo6aSoF3p+S0NFDHPHgbYiBJgQZGv/ERLZmZ0t5q6wkJKnqMhzBz8MufZG0ZXsZRzHYYrWJk1TDShwoZfiVWbn2rce4L19/03NdfPRtr2nHzvKc/emdx/d3LDyM4XkaJq+cfm/bY8bqFq1fv6FyOvX+1oHvwefbOru7Y0zcz5q91cn3Tq52bInXKZx9RCGvWp8UlOEsQzpxD6T/05acLVrNap952xtZhP0xWx0+0iY+fnCrjtT1FbQ2389oqStRWanr34n+eflDP00eNTBe09C6rWpeVidoeugYAvcGv8LTaXynTgF0DGRLXuBwA/y5J0T00eaRi6JdU8UmS4qDyuqqwJBTvUMXlkqApuriC9Vdu9UkSBIfk5fPVpZGx4MYuV46oJ+kEY0tOTnr6qEKLpcQNmZh+SJ2ImdjppB56CnnSKS02+RpiJifBU2MEnYC8izsQ2clwI9I+1YYLf3Gtkw8SVgdtm4XAwyNdtX46hDAvXCL2GCmnN3ZetuitjjuuvUr5/0PfKX9DwuFDDfpT17zfga0rz19x8fIFq84TXdXF99Wdtr1n/m5lz4fKh8pLyPrJR8gyV+hdtuva4/Mv2Lj1ih27+lg74MwMf2tPV9/aEPAZUHI97ucl3KK2k5t4PReeOJ319ZfAyRW8pRiS+gUt3aSlD6jpeSPTBS29y6C2pIDWK8yCw0JYeIl7wbKhNGJ1pqWZBQEIyYUcNwVKAXHz0vPBYdBQiw8WTxJRTWOGj2+K1tf/PFpXNzVaf2ojO+KOwcEvTpva/POG6c1EmNrUMqWhpRkIfcaHKAN0OZ81eEfOGnzxWQOjb0jBFAZx/C+zhmCNsJ9hQWsvOLVn0n5GBm1eUrt/zK5jR21o/OiJKy9AhwzKa/6alefjSoYJlXV2dVyL7IwUqpp+Qes1ytH2RjTouvnWlnFKMOP2oSGVpeD1c2ZST4ByefGmpvMavgVOruA1XMnTC0emC1p6V0B9A0u1np977PkV5qi9zXh+BQ8XJOgmziYWsLhqD+1vHQZzli2Dxi8VWsCcbXDIRM6dEpOdxEnL+CQocxLLTDtnDWdWTT4Wyh0nAU7ot8Herhf//uZLf5xv0ulUfvGjOONEDrXMYEgzK+CtE9qVsXpQVixvbB7mnLQ8CVqeut5Qc/0zNdcJKk9oH6byMk5M5VGJGk2mO108BE7wQmekxuJwGFF+vs6WAeDL0umKLHa6drMgI7HQX0YznaWSNBddcwhCLotpRQ5tBcd+ThplmiAy+BMMx2M6XcOLuERnVGvx+3WnH9vn31Wm9Cv3oTPQhPGbvaRDW9Q9dstdd/XVrfR7t8jpaBvqQuejTSZZXeCR145+8+1PDivZbnPyN+hT3SphMXhgNARhQWRMoMKEHQ6/X19RkWu3V+Xr9aEchzvgiMYCATCbfxaNmc3YJNDOmfLEZnDT4VwQvFNiQupwHj45Cp00iOdT56kG4bniI7dDo6KTeT2fSk+Ltyhf7dl5pPfHLSgb4QUvT7nsi2+R+bhTt2fL+U90tDx99FwN5Pu4fbWMBnC3/ZprdiD9/ciByqY1XcvYaf26naXlbOCeHGf7BhavuJhFHD0h/FXwSAVgZP0Zi5ozAMh6jE0ZWF4vsh39sg5pyx2NKqQzEZ2XGU+dFNAgrdc1Ne977elTUafn6kbhr2ed0XJ29tMLqh5sYBENqFX4M4lKD8Q9ehmS1eqmkUWyR8ay7CDxvRTYHVKNZ7qk8YhEdy1YcOklCy+67Pqa0tKaiorSGvGlCzavv+iCDZu7ykKhsrKqKkDwa+HPgkEygQuqIm4KNEUEQjLdBhvobPTrYvM6MzavFyCQ9fpZmoNENQebXw6qkISXvbF5mNVHiE23yjF6xRM27knfvXTUtKZoET+/fAk7F+uray7vKyjOr+KHAr4bGHqI3IN7+G5S+AS7SU0nbeih999Xlbp/qtQllG7Sj/p4jIw7kiaIOqTTySBou5KZB5gLq7jGWhvCumKTs7N6sN5L+p1zkG2h8t3HkHQFCVwRmQhIknSCRC8wvD8WUrffQHtNwbWDkz3iI84XlPdRySFI3luLeVIwEfnuWhIEtNuffHstwOzeZBl/+gzwRczUIGsiggSSZNFlkHRtI0Z+oT8E+bOoWSnwxY/oUzVPdILhSZyRP8ezp2Vz+E4SGJn/ndpNDXwrMFMaMYjsRi+qN9Luoz60qB5QH885cqO31JNM8Ua1DBJFgVlJkOt5SRihMGIaeQcIpN7Ap91gROGgt0eWkkvbi2wunXrfKIyCdLA9wszuRplAgHssUq3uc6/avnXvvku37cGf9hzou3r/LbcAELbTizQXhfm75mXsYF6m6kEvys4gbKuXAofMQuS5LUhtbJnmP9AJy8gdX3yp56m7v+Aps89kZzPacGPqPmctKUf+VkA7vpHbtCsijrgDV9RLQAg9pa0JI9VZmsxW0W/VN5vqlE12xKZeO24nRzp2bfoHPRPEf7z2SBs4vvHEBm8ApCxj83oe25YVSSeAEcaCFtqW8B8j5EX48mN//IKMjge2AeK7BW0S+6EYdkQaJaL3+XI8RW5ntmywWIrSafaLika5cnP12dklBpdLzpRy83Knx0heRt66PJxOMvMy82yFPiiEabFCndlkMzXHbNp2YiNNoxZenyxzKUghO/CtQOhvro/H5DgKdA420DrVfS4oWELdb/7qWvq7BuL7XXhXXu9CVyrtGKN5yj0hZNq9ecn93ynPj9q6VMBLtvjQpG+e6ps7ebnwys5f3ucNFDzwTXgIxqK0Tx5wFVff9zVyT//Q4+XsWgfzjp+0n6MTYDbdHRriMbs/Sh7wQyNfQ04lboD45x8nfd7MPgcMBhzF34tPQRpYGbthFXUmWnBEBixim90k62TJikTRaiW6PJLPDTwBLSYu4RpNwn+8DhpfWI1CfA+zWrZnHP5+zefKBrTh0zXKHkmuzliH39q3rwfXHT/UN3Nu1gWuZ9Wn05u0pyuGRuJWn14KAMTT4QTpzcPp0q6k3PF0dS8BvtMDAcsjIIiIQGKXQLYPAt8FgTU2uvZ8EQDruB3sL/EV7krVDmZIWNNupYoPkxTdQ3NGKoYYgS4mKQ4q76sKS0JxHADfqZupKbq4gq9wuaT6/wCVeR0IAAAAAQAAAAEZmiehT9dfDzz1AAkIAAAAAADJQhegAAAAAMnoSqH7DP2oCo0IjQABAAkAAgAAAAAAAHgBY2BkYODo/buCgYGr9zfPv0quXqAIKrgJAJZXBsIAeAFtkQOsGEEQhv/bnd272rZtG0Ft27ZtW1G9dYMiamrbZlgrqN17M89K8uVfTna/oRs4AwCUGVBCU0zQl7DAlEIZWoPOfhXUs0BbVQAL1CG0ZepQd9STPdUW9dQ61FGN+U5LpOW1pswUpmU0hZj+TGOmWnQ2lPNyV2rEoO/A+mUw0CwATG8cNjkwyXzEYZrG9Of5NUyy+XBY7Q4Hm9a8tgCH/WU4bOcwPfmsjc7GvDcYPWk7StjU2G8qAf5xwHQE6D+zHRXUbqzi96bmrEQNEeim4V965jWnB+ho0sNRHnTn7E5H0V3nQAlaAGsawqkxWKfGhDPoO2Ts/Gdwsk5fIecd011vh9O/OaegHO9toBWAfYLM5JBSxvoNquliyEeDvUucbeXvMd55vIqRtTGMJTnzAkP5bdnsXvTX6VGOPkbfYe+yRgh/6xHoLms6QDmmlvyFPThTB2PEtbczfMbr3XUu1JD7fmqUjaYre68jzpPD3wJIH6QH0RyQ5L6Ui/GeGFqDOZLiPj7iXnpkDsKJ5+TwO3LmEe8JYecb2fcazoXMC/Ed4z0J7EFS3MdH3EuPJJX07gom+ff4/DMcpS1ee85bBLQNGO84cgiqPerpVcghUBEeK/S1jzBBfUZbwUv5X/7bkOlslqCEwJ5TBw4lBFsBJdRuHA4vYk/own8RLYvLrQAAeAEc0jWMJFcQxvFnto/5LjEvHrdbmh2Kji9aPL4839TcKPNAa6mlZUyOmZk6lzbPJ3bo56//Cz+Vaqqrat5rY8x7xnzxl3nvo+27jFnz8c/mI9Nmh2XBdMsilrBitsnD9rI8aiN5DI/jSftC9mIf9pMfIB4kHiI+hWfQY5aPAYYYYYwpcyfpMMX0aZzBWZzDeVygchGXcBlX8ApexWt4HW/gLbzNbnfwLt7DJ/p0TX4+Uucji1hCnY/U+cijVB7D46jzkb3Yh/3kB4gHiYeIT+EZ9JjlY4AhRhhjytxJOkwxfRpncBbncB4XqFzEJVzGFbyCV/EaXscbeAtvs9sdvIv3cjmftWavuWs2mg6byt3ooIsFOyx77Kos2kiWsIK/UVPDOjawiQmO4CgdxnAcJzClz2PVbNKsy2ZzvoncjQ66qE2kNpHaRJawgr9RU8M6NrCJCY6gNpFjOI4TmNIn36TNfGSH5RrssKtyN+59b410iF0sUFO0l2UJtY/8jU9rWMcGNjHBEUypf0z8mm7vZLvZaC/LzdhmV2XBvpBF25IlLJOvEFfRI+NjgCFGGGNK5Rs6Z7Ij/45yNzro4m9Ywzo2sIkJjuBj2ZnvLDdjGxntLLWzLGGZfIW4ih4ZHwMMMcIYUyq1s8xkl97bH0y3JkZyM36j/+58rvTQxwBDjDDGNzyVyX35Ccjd6KCLv2EN69jAJiY4go/lfr05F+Ua7CCzGx10sYA9tiWLxCWs2BfyN+Ia1rGBTUxwBEfpMIbjOIEpfdjHvGaTd9LJb0duRp2S1O1I3Y4sYZl8hbiKHhkfAwwxwhhTKt/QOZPfmY3//Ss3Y5tNpTpL9ZQeGR8DDDHCGN/wbCbdfHO5GbW51OZSm8sSlslXiKvokfExwBAjjDGlUpvLTBY0K5KbiDcT672SbXZY6k7lbnTQxQI1h+1FeZTKY3gcT2KvTWUf9pMZIB4kHiI+xcQzxGfpfA7P4wW8yG4eT/kYYIgRxvgb9TWsYwObmOAITlI/xf7TOIOzOIfzuEDlIi7hMq7gFbyK1/A63sBbeJtvdwfv4j28zyaP8QmVL/imL/ENJ5PJHt3RqtyMbbYlPfQxwBAjjPEN9ZksqkMqN6PuV7bZy7LDtuRudNDFwzx1FI/hcTzJp73Yh/3kB4gHiYeIT+EZ9JjlY4AhRhjjb1TWsI4NbGKCIzjJlCmcxhmcxTmcxwVcxCVcxhW8glfxGl7HG3gLbzPxDt7Fe/gY/+egvq0YCAEoCNa1n+KVyTUl3Q0uIhoe+3DnRfV7nXGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOMc5zjHOc5xjnOc4xznOMc5znGOc5zjHOc4xznOcY5znOM8XZouTZemS1OAKcAUYAowBZgCTAHm3x31O7p3vNf5c1iXeBkEAQDFcbsJX0IqFBwK7tyEgkPC3R0K7hrXzsIhePPK/7c77jPM1yxSPua0WmuDzNcuNmuLtmq7sbyfsUu7De/xu9fvvvDNfN3ioN9j5pq0ximd1hmd1TmlX7iky7qiq7qmG3pgXYd6pMd6oqd6pud6oZd6pdd6p/f6oI/6pC/KSxvf9F0/1LFl1naRcwwzrAu7AHNarbW6oEu6rCu6qmu6ob9Y7xu+kbfHH1ZopCk25RVrhXKn4LCO6KiOGfvpd+R3is15xXmVWKGRptgaysQKpUwc1hEdVcpEysTI7xTbKHMcKzTSFDtCmVihkab4z0FdI0QQBAEUbRz6XLh3Lc7VcI/WN54IuxXFS97oH58+MBoclE1usbHHW77wlW985wcHHHLEMSecsUuPXMNRqfzib3pcllj5xd+0lSVW5nNIL3nF6389h+Y5NG3Thja0oQ1taEMb2tCGNrQn+QwjrcwxM93gJre4Y89mvsdb3vGeD3zkE5/5wle+8Z0fHHDIEceccMaOX67wNz3747gObCQAQhCKdjlRzBVD5be7rwAmfOMQsUvPLj279OzSYBks49Ibl97In/HCuNDGO+NOW6qlWqqlWqqlWqqlWqqYUkwpphTzifnEfII92IM92IM92IM92IM92IM92I/D4/A4PA6Pw+PwODwOj8M/f7kaaDXQyt7K3mqglcCVwNVAq4FWA60GWglZCVkJWQlZCVkJWQlZDbQyqhpoNdAPh3NAwCAAwwDM+7b2sg8kCjIO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO4zAO47AO67AO67AO67AO67AO67AO67AO67AO67AO67AO67AO63AO53AO53AO53AO53AO53AO53AO53AO53AO53AO53AO5xCHOMQhDnGIQxziEIc4xCEOcYhDHOIQhzjEIQ5xiEMd6lCHOtShDnWoQx3qUIc61KEOdahDHepQhzrUoQ6/h+P6RpIjiKEoyOPvCARUoK9LctP5ZqXTop7q/6H/0H+4P9yfPz82bdm2Y9ee/T355bS3/divDW9reFtDb4beDL0ZejP0ZujN0JuhN0Nvht4MvRl6M/Rm6M3w1of3PVnJSlaykpWsZCUrWclKVrKSlaxkJStZySpWsYpVrGIVq1jFKlaxilWsYhWrWMUqVrGa1axmNatZzWpWs5rVrGY1q1nNalazmtWsYQ1rWMMa1rCGNaxhDWtYwxrWsIY1rGENa1nLWtaylrWsZS1rWcta1rKWtaxlLWtZyzrWsY51rGMd61jHOtaxjnWsYx3rWMc61rEeTf1o6kdTP/84rpMqCKAYhmH8Cfy2JjuLCPiYPDH1Y+rH1I+pH1M/pn5M/Zh6FEZhFEZhFEZhFEZhFEZhFFZhFVZhFVZhFVZhFVZhFVbhFE7hFE7hFE7hFE7hFE7hFCKgCChPHQFlc7I52ZxsTgQUAUVAEVAEFAFFQBFQBBQBRUARUAQUAUVAEVAEFAFFQBFQti5bl63L1mXrsnXZuggoAoqAIqAIKAKKgCKgCCgCioAioAgoAoqAIqAIKAKKgCKgCCgCyt5GQBFQBPTlwD7OEIaBKAxSOrmJVZa2TsJcwJ6r0/+9sBOGnTDshOF+DndyXG7k7vfh9+n35fft978Thp2wKuqqqKtarmq58cYbb7zzzjvvfPDBBx988sknn3zxxRdfPHnyVPip8FPhp8JPhZ8KP78czLdxBDAMAMFc/bdAk4AERoMS5CpQOW82uWyPHexkJzvZyU52spOd7GQnu9jFLnaxi13sYhe72MVudrOb3exmN7vZzW52s8EGG2ywwQYbbLDBBnvZy172spe97GUve9nLJptssskmm2yyySabbLHFFltsscUWW2yxxX6+7P+rH/qtf6+2Z3u2Z3u2Z3u2Z3u2Z3s+O66jKoYBGASA/iUFeLO2tqfgvhIgVkOshvj/8f/jF8VqiL8dqyG+d4klllhiiSWWWGKJJY444ogjjjjiiCOO+Pua0gPv7paRAHgBLcEDlNxQAADArI3Ydv7Vtm3btm3btm3btm3bD7VvBoIgLXVVqCf0ztXT9dzd3j3cvcX90CN5Snmae/p45np2e356gbeH94HP8Q3x3feH/X38NwJwoHigQ2Ba4GBQCK4NfgxVDE0OnQr7w1nCI8P7wi8jdqR4ZGzkRDQSLRmdH/0UqxTrEVsbux/PHe8b3xh/lgglzESJRJfE6MS6ZChZJzkj+RouCA9GJKQuMhI5hsZRHR2A7kZ/YZWxldhtPDPeFd+IPybyE0OIy2SIrEy2IneSX8mvFKB6UpfodPQYeiOTjmnK3GOzsCPYpexaLjdXiRvBHeJ+8BX5Lvxe/qOACmWEnsJ60SsyYjqxiLhE3CoeE6+LL8RvUlRqJXWThkszpJXSbjkq83JaOZ9cXm4gd5IXKZACK4qSSSmiVFWmq0lVUtOr+dXyagO1oxbRSM3UsmnFtOpaC62nNkqbo7M60HPppfXaemu9j77X4IwUI49RxqhrtDWOGzeM92Y985lFWWWtcdZia4d10/piU3YZu6+91j7rME5xp5szGVAgDcgBioDhYDpYDjaDE+AmeAW+p8R/A5ajfCcAAAABAAAA3QCKABYAWAAFAAIAEAAvAFwAAAEAAQsAAwABeAF9jgNuRAEYhL/aDGoc4DluVNtug5pr8xh7jj3jTpK18pszwBDP9NHTP0IPs1DOexlmtpz3sc9iOe9nmddyPsA8+XI+qI1COZ/kliIXhPkiyDo3vCnG2CaEn0+2lH+gmfIvotowZa3769ULZST4K+cujqTb/j36S4w/QmgDF0tWvalemNWLX+KSMBvYkhQSLG2FZR+afmERIsqPpn7+yvxjfMlsTjlihz3OuZE38bTtlAAa/TAFAHgBbMEDjJYBAADQ9/3nu2zbtm3b5p9t17JdQ7Zt21zmvGXXvJrZe0LA37Cw/3lDEBISIVKUaDFixYmXIJHEkkgqmeRSSCmV1NJIK530Msgok8yyyCqb7HLIKZfc8sgrn/wKKKiwIooqprgSSiqltDLKKqe8CiqqpLIqqqqmuhpqqqW2Ouqqp74GGmqksSaaaqa5FlpqpbU22mqnvQ466qSzLrrqprs9NpthprNWeWeWReZba6ctQYR5QaTplvvhp4VWm+Oyt75bZ5fffvljk71uum6fHnpaopfbervhlvfCHnngof36+Gappx57oq+PPpurv34GGGSgwTYYYpihhhthlJFGG+ODscYbZ4JJJjphoykmm2qaT7445ZkDDnrujRcOOeyY46444qirZtvtnPPOBFG+BtFBTBAbxAXxQYJC7rvjrnv/xpJXmpPDXpqXaWDg6MKZX5ZaVJycX5TK4lpalA8SdnMyMITSRjxp+aVFxaUFqUWZ+UVQQWMobcKUlgYAHQ14sAAAeAFFSzVCLEEQ7fpjH113V1ybGPd1KRyiibEhxt1vsj3ZngE9AIfgBmMR5fVk8qElsRjHOHAYW+Qwyumxct4bKxXkWDEvx7JjdszQNAZcekzi9Zho8oV8NCbnIT/fEXNRJwqmlaemnQMbN8E1OE7Mzb/P/8xzKZrEMA2hl3rQATa0Uxs2bN+2f8M2AEpwj5yQBvklvJ3AqRcEaMKrWq/19eWakl7NsZbyJoNblqlZc7KywcRbRnBjc00FeF6/enoi05EcG62tsXhkPcdk87BHVC+ZXleUPrOsUHaUI2tb4y/8OwbsTEAJAA==) format("woff")}*{box-sizing:border-box}body{padding:0;margin:0;font-family:"Open Sans","Helvetica Neue",Helvetica,Arial,sans-serif;font-size:16px;line-height:1.5;color:#606c71}a{color:#1e6bb8;text-decoration:none}a:hover{text-decoration:underline}.page-header{color:#fff;text-align:center;background-color:#159957;background-image:linear-gradient(120deg,#155799,#159957);padding:1.5rem 2rem}.page-header :last-child{margin-bottom:.5rem}@media screen and (max-width:42em){.page-header{padding:1rem 1rem}}.project-name{margin-top:0;margin-bottom:.1rem;font-size:2rem}@media screen and (max-width:42em){.project-name{font-size:1.75rem}}.project-tagline{margin-bottom:2rem;font-weight:400;opacity:.7;font-size:1.5rem}@media screen and (max-width:42em){.project-tagline{font-size:1.2rem}}.project-author,.project-date{font-weight:400;opacity:.7;font-size:1.2rem}@media screen and (max-width:42em){.project-author,.project-date{font-size:1rem}}.main-content,.toc{max-width:64rem;padding:2rem 4rem;margin:0 auto;font-size:1.1rem}.toc{padding-bottom:0}.toc .toc-box{padding:1.5rem;background-color:#f3f6fa;border:solid 1px #dce6f0;border-radius:.3rem}.toc .toc-box .toc-title{margin:0 0 .5rem;text-align:center}.toc .toc-box>ul{margin:0;padding-left:1.5rem}@media screen and (min-width:42em) and (max-width:64em){.toc{padding:2rem 2rem 0}}@media screen and (max-width:42em){.toc{padding:2rem 1rem 0;font-size:1rem}}.main-content :first-child{margin-top:0}@media screen and (min-width:42em) and (max-width:64em){.main-content{padding:2rem}}@media screen and (max-width:42em){.main-content{padding:2rem 1rem;font-size:1rem}}.main-content img{max-width:100%}.main-content h1,.main-content h2,.main-content h3,.main-content h4,.main-content h5,.main-content h6{margin-top:2rem;margin-bottom:1rem;font-weight:400;color:#159957}.main-content p{margin-bottom:1em}.main-content code{padding:2px 4px;font-family:Consolas,"Liberation Mono",Menlo,Courier,monospace;color:#567482;background-color:#f3f6fa;border-radius:.3rem}.main-content pre{padding:.8rem;margin-top:0;margin-bottom:1rem;font:1rem Consolas,"Liberation Mono",Menlo,Courier,monospace;color:#567482;word-wrap:normal;background-color:#f3f6fa;border:solid 1px #dce6f0;border-radius:.3rem;line-height:1.45;overflow:auto}@media screen and (max-width:42em){.main-content pre{font-size:.9rem}}.main-content pre>code{padding:0;margin:0;color:#567482;word-break:normal;white-space:pre;background:0 0;border:0}@media screen and (max-width:42em){.main-content pre>code{font-size:.9rem}}.main-content pre code,.main-content pre tt{display:inline;max-width:initial;padding:0;margin:0;overflow:initial;line-height:inherit;word-wrap:normal;background-color:transparent;border:0}.main-content pre code:after,.main-content pre code:before,.main-content pre tt:after,.main-content pre tt:before{content:normal}.main-content ol,.main-content ul{margin-top:0}.main-content blockquote{padding:0 1rem;margin-left:0;color:#819198;border-left:.3rem solid #dce6f0;font-size:1.2rem}.main-content blockquote>:first-child{margin-top:0}.main-content blockquote>:last-child{margin-bottom:0}@media screen and (max-width:42em){.main-content blockquote{font-size:1.1rem}}.main-content table{width:100%;overflow:auto;word-break:normal;word-break:keep-all;-webkit-overflow-scrolling:touch;border-collapse:collapse;border-spacing:0;margin:1rem 0}.main-content table th{font-weight:700;background-color:#159957;color:#fff}.main-content table td,.main-content table th{padding:.5rem 1rem;border-bottom:1px solid #e9ebec;text-align:left}.main-content table tr:nth-child(odd){background-color:#f2f2f2}.main-content dl{padding:0}.main-content dl dt{padding:0;margin-top:1rem;font-size:1rem;font-weight:700}.main-content dl dd{padding:0;margin-bottom:1rem}.main-content hr{height:2px;padding:0;margin:1rem 0;background-color:#eff0f1;border:0}code span.kw { color: #a71d5d; font-weight: normal; } 
code span.dt { color: #795da3; } 
code span.dv { color: #0086b3; } 
code span.bn { color: #0086b3; } 
code span.fl { color: #0086b3; } 
code span.ch { color: #4070a0; } 
code span.st { color: #183691; } 
code span.co { color: #969896; font-style: italic; } 
code span.ot { color: #007020; } 
</style>





</head>

<body>




<section class="page-header">
<h1 class="title toc-ignore project-name">Pop Gen Lectures</h1>
<h4 class="author project-author">Maryam Nouri-Aiin</h4>
<h4 class="date project-date">Sep 11, 2023</h4>
</section>



<section class="main-content">
<div id="table-of-contents" class="section level1">
<h1>Table of Contents:</h1>
<ul>
<li><a href="#id-section1">Entry 1: 2023-09-11</a></li>
<li><a href="#id-section2">Entry 2: 2023-09-13</a></li>
<li><a href="#id-section3">Entry 3: 2023-09-18</a></li>
<li><a href="#id-section4">Entry 4: 2023-09-20</a></li>
<li><a href="#id-section5">Entry 5: 2023-09-25</a></li>
<li><a href="#id-section6">Entry 6: 2023-09-27</a></li>
<li><a href="#id-section7">Entry 7: 2023-10-02</a></li>
<li><a href="#id-section8">Entry 8: 2023-10-04</a></li>
<li><a href="#id-section9">Entry 9: 2023-10-09</a></li>
<li><a href="#id-section10">Entry 10: 2023-10-11</a></li>
<li><a href="#id-section11">Entry 11: 2023-10-16</a></li>
<li><a href="#id-section12">Entry 12: 2023-10-18</a></li>
<li><a href="#id-section13">Entry 13: 2023-10-23</a></li>
<li><a href="#id-section14">Entry 14: 2023-10-26</a></li>
<li><a href="#id-section15">Entry 15: 2023-11-30</a></li>
<li><a href="#id-section16">Entry 16: 2023-11-01</a></li>
<li><a href="#id-section17">Entry 17: 2023-11-06</a></li>
<li><a href="#id-section18">Entry 18: 2023-11-08</a></li>
<li><a href="#id-section19">Entry 19: 2023-11-13</a></li>
<li><a href="#id-section20">Entry 20: 2023-11-15</a></li>
<li><a href="#id-section21">Entry 21: 2023-11-20</a></li>
<li><a href="#id-section22">Entry 22: 2023-11-22</a></li>
<li><a href="#id-section23">Entry 23: 2023-11-27</a></li>
<li><a href="#id-section24">Entry 24: 2023-11-29</a></li>
<li><a href="#id-section25">Entry 25: 2023-12-04</a></li>
<li><a href="#id-section26">Entry 26: 2023-12-06</a></li>
</ul>
<div id="id-section1">

</div>
<div id="entry-1-2023-09-11." class="section level2">
<h2>Entry 1: 2023-09-11.</h2>
<div id="ecological-genomics-tutorials-population-landscape-genomics-1" class="section level3">
<h3>Ecological Genomics Tutorials: Population &amp; Landscape Genomics
1</h3>
<p>September 11, 2023</p>
</div>
<div id="heres-our-pipeline" class="section level3">
<h3>2. Here’s our “pipeline”</h3>
<ul>
<li>Visualize the quality of raw data (Program: FastQC)</li>
<li>Clean raw data (Program: Trimmomatic)</li>
<li>Visualize the quality of cleaned data (Program: FastQC)</li>
<li>Calculate #’s of cleaned, high quality reads going into mapping</li>
</ul>
<div id="then-use-these-cleaned-reads-to-align-to-the-reference-genome-next-time-so-that-we-can-start-estimating-genomic-diversity-and-population-structure." class="section level5">
<h5>Then use these cleaned reads to align to the reference genome next
time so that we can start estimating genomic diversity and population
structure.</h5>
</div>
</div>
<div id="visualize-clean-and-visualize-again" class="section level3">
<h3>3.-5. Visualize, Clean, and Visualize again</h3>
<div id="whenever-you-get-a-new-batch-of-ngs-data-the-first-step-is-to-look-at-the-data-quality-of-coming-off-the-sequencer-and-see-if-we-notice-any-problems-with-base-quality-sequence-length-pcr-duplicates-or-adapter-contamination.-a-lot-of-this-info-is-stored-in-the-raw-data-files-you-get-from-the-core-lab-after-sequencing-which-are-in-fastq-format." class="section level6">
<h6>Whenever you get a new batch of NGS data, the first step is to look
at the data quality of coming off the sequencer and see if we notice any
problems with base quality, sequence length, PCR duplicates, or adapter
contamination. A lot of this info is stored in the raw data files you
get from the core lab after sequencing, which are in “fastq”
format.</h6>
</div>
</div>
<div id="the-fastq-files-for-our-project-are-stored-in-this-path-netfilesecogenpopulationgenomicsfastqred_spruce" class="section level3">
<h3>The fastq files for our project are stored in this path:
/netfiles/ecogen/PopulationGenomics/fastq/red_spruce</h3>
<p><strong><em>cd over there and ll to see the files.</em></strong></p>
<div id="there-should-be-190-fastq-files-2-for-each-of-the-95-samples-2-filessample-because-these-are-paired-end-reads-and-each-sample-gets-a-file-with-the-forward-reads-r1-and-another-with-the-reverse-reads-r2." class="section level6">
<h6>There should be 190 fastq files – 2 for each of the 95 samples (2
files/sample because these are paired-end reads, and each sample gets a
file with the forward reads (R1) and another with the reverse reads
(R2)).</h6>
</div>
<div id="the-naming-convention-for-our-data-is-_.fast.gz" class="section level6">
<h6>The naming convention for our data is:
<PopCode><em><RowID></em><ColumnID>_<ReadDirection>.fast.gz</h6>
</div>
<div id="together-define-the-unique-individual-id-for-each-dna-sample-and-there-should-be-2-files-per-sample-and-r1-and-an-r2" class="section level5">
<h5>Together, <PopCode><em><RowID></em><ColumnID> define the unique
individual ID for each DNA sample, and there should be 2 files per
sample (and R1 and an R2)</h5>
</div>
</div>
<div id="sowhat-is-a-.fastq-file-anyway" class="section level3">
<h3>So…what is a <strong><em>.fastq file</em></strong> anyway?</h3>
<p><strong><em>A fastq file is the standard sequence data format for
NGS. It contains the sequence of the read itself, the corresponding
quality scores for each base, and some meta-data about the
read.</em></strong></p>
<div id="the-files-are-big-typically-many-gb-compressed-so-we-cant-open-them-completely.-instead-we-can-peek-inside-the-file-using-head.-but-size-these-files-are-compressed-note-the-.gz-ending-in-the-filenames-and-we-want-them-to-stay-compressed-while-we-peek.-bash-has-a-solution-to-that-called-zcat.-this-lets-us-look-at-the-.gz-file-without-uncompressing-it-all-the-way.-lets-peek-inside-a-file" class="section level6">
<h6>The files are big (typically many Gb compressed), so we can’t open
them completely. Instead, we can peek inside the file using head. But
<em>size these files are compressed (note the <strong>.gz</strong>
ending in the filenames)</em>, and we want them to stay compressed while
we peek. Bash has a solution to that called <strong>zcat</strong>. This
lets us look at the .gz file without uncompressing it all the way. Let’s
peek inside a file:</h6>
</div>
</div>
<div id="zcat-2505_9_c_r2.fastq.gz-head--n-4" class="section level3">
<h3>zcat 2505_9_C_R2.fastq.gz | head -n 4</h3>
<p><span class="citation">@A00354:455:HYG3FDSXY:1:1101:3893:1031</span>
2:N:0:CATCAAGT+TACTCCTT
GTGGAAAATCAAAACCCTAATGCTGAAAGGAATCCAAATCAAATAAATATTTTCACCGACCTGTTTCGATGCCAGAATTGTCTGCGCAGAAGACTCGTGAAATTTCGCCAGCAGGTAAAATTAAAAGGCTAGAATTAACCGCTGAAATGGA
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:F:F
Note: zcat lets us open a .gz (gzipped) file; we then “pipe” | this
output from zcat to the head command and print just the top 4 lines
-n4</p>
</div>
<div id="the-fastq-file-format-has-4-lines-for-each-read" class="section level3">
<h3>The fastq file format has 4 lines for each read:</h3>
<div id="line-description" class="section level4">
<h4>Line Description</h4>
</div>
<div id="always-begins-with-and-then-information-about-the-read" class="section level4">
<h4>1 Always begins with ‘@’ and then information about the read</h4>
</div>
<div id="the-actual-dna-sequence" class="section level4">
<h4>2 The actual DNA sequence</h4>
</div>
<div id="always-begins-with-a-and-sometimes-the-same-info-in-line-1" class="section level4">
<h4>3 Always begins with a ‘+’ and sometimes the same info in line
1</h4>
</div>
<div id="a-string-of-characters-which-represent-the-quality-scores-always-has-same-number-of-characters-as-line-2" class="section level4">
<h4>4 A string of characters which represent the quality scores; always
has same number of characters as line 2</h4>
<p>Here’s a useful reference for understanding Quality (Phred) scores.
If P is the probability that a base call is an error, then: Q =
-10*log10(P)</p>
<p>So: Phred Quality Score Probability of incorrect base call Base call
accuracy 10 1 in 10 90% 20 1 in 100 99% 30 1 in 1000 99.9% 40 1 in
10,000 99.99% The Phred Q score is translated to ASCII characters so
that a two digit number can be represented by a single character.</p>
</div>
</div>
<div id="quality-encoding--.0123456789abcdefghi" class="section level3">
<h3>Quality encoding: !“#$%&amp;’()*+,-./0123456789:;&lt;=&gt;?<span class="citation">@ABCDEFGHI</span></h3>
<pre><code>               |         |         |         |         |
Quality score: 0........10........20........30........40   </code></pre>
<p>What kind of characters do you want to see in your quality score?</p>
</div>
<div id="visualize-using-fastqc" class="section level3">
<h3>Visualize using FastQC</h3>
<div id="were-going-to-use-the-program-fastqc-already-installed-on-our-server.-fastqc-looks-at-the-quality-collectively-across-all-reads-in-a-sample." class="section level4">
<h4>We’re going to use the program <strong><em>FastQC</em></strong>
(already installed on our server). <strong><em>FastQC looks at the
quality collectively across all reads in a sample.</em></strong></h4>
<p>First, let’s <em>cd back to our home directories ~/</em> and
<strong>set up some new folders</strong> to store our work. We’ll make 3
directories to store our data, scripts, and results:</p>
<p>mkdir mydata/</p>
<p>mkdir myscripts/</p>
<p>mkdir myresults/</p>
<div id="then-lets-cd-into-the-myresults-folder-then-use-pwd-to-prove-to-yourself-that-youre-in-the-myresults-folder-within-your-home-directory.-it-should-look-like-this-but-with-your-home-directory-info-instead-of-mine" class="section level5">
<h5>Then let’s cd into the myresults/ folder then use pwd to prove to
yourself that you’re in the myresults/ folder within your home
directory. It should look like this (but with your home directory info
instead of mine):</h5>
<p>[<a href="mailto:kellrlab@ecogen" class="email">kellrlab@ecogen</a>
myresults]$ pwd</p>
<p>/users/k/e/kellrlab/myresults</p>
</div>
</div>
<div id="now-within-myresults" class="section level4">
<h4>Now within myresults/</h4>
<div id="lets-make-another-folder-called-fastqc-to-hold-the-outputs-from-our-qc-analysis.-do-that-on-your-own-just-like-we-did-above-then-cd-into-the-fastqc-folder-and-type-pwd-again-to-prove-to-yourself-you-did-it-right." class="section level5">
<h5>let’s make another folder called fastqc/ to hold the outputs from
our QC analysis. Do that on your own, just like we did above, then cd
into the fastqc/ folder and type pwd again to prove to yourself you did
it right.</h5>
</div>
</div>
<div id="now-were-ready-to-run-fastqc-to-look-at-the-quality-of-our-sequencing.-the-basic-command-is-like-this" class="section level4">
<h4>Now, we’re ready to run FastQC to look at the quality of our
sequencing. The basic command is like this:</h4>
</div>
</div>
<div id="fastqc-filename.fastq.gz--o-outputdirectory" class="section level3">
<h3>fastqc filename.fastq.gz -o outputdirectory/</h3>
<div id="this-will-generate-an-.html-output-file-for-each-input-file-youve-run." class="section level4">
<h4>This will generate an .html output file for each input file you’ve
run.</h4>
<div id="once-youve-got-results-lets-use-filezilla-to-transfer-the-folder-myresultsfastqc-over-to-your-laptop-and-into-the-results-folder-in-your-github-repo.-once-the-file-transfer-is-complete-go-to-where-you-saved-your-files-on-your-laptop-and-try-double-clicking-one-of-the-html-outputs.-it-should-open-with-a-web-browser." class="section level6">
<h6>Once you’ve got results, let’s use <strong><em>Filezilla to transfer
the folder ~/myresults/fastqc/ over to your laptop and into the results
folder in your github repo</em></strong>. Once the file transfer is
complete, go to where you saved your <strong><em>files on your laptop
and try double-clicking one of the html outputs</em></strong>. It should
open with a web browser.</h6>
</div>
<div id="how-does-the-quality-look" class="section level6">
<h6>How does the quality look?</h6>
</div>
<div id="since-we-made-some-changes-added-files-to-our-github-repo-we-should-practice-committing-these-and-then-pushing-to-github" class="section level6">
<h6>Since we made some changes (added files) to our github repo, we
should practice committing these and then pushing to GitHub!</h6>
<hr />
<div id="id-section2">

</div>
</div>
</div>
</div>
</div>
<div id="entry-2-2023-09-13." class="section level2">
<h2>Entry 2: 2023-09-13.</h2>
<div id="ecological-genomics-tutorials-population-landscape-genomics-2" class="section level3">
<h3>Ecological Genomics Tutorials: Population &amp; Landscape Genomics
2</h3>
<p>September 13, 2023</p>
</div>
<div id="learning-objectives-for-091323" class="section level3">
<h3>Learning Objectives for 09/13/23</h3>
<div id="review-visualizing-qc-on-our-sequence-reads" class="section level4">
<h4>Review visualizing QC on our sequence reads</h4>
</div>
<div id="trim-reads-using-fastp-and-assess-pre--and-post-trimming" class="section level4">
<h4>Trim reads using fastp and assess pre- and post trimming</h4>
</div>
<div id="start-mapping-a.k.a.-aligning-each-set-of-cleaned-reads-to-a-reference-genome" class="section level4">
<h4>Start mapping (a.k.a. aligning) each set of cleaned reads to a
reference genome</h4>
</div>
<div id="visualize-sequence-alignment-files" class="section level4">
<h4>Visualize sequence alignment files</h4>
</div>
<div id="calculate-mapping-statistics-to-assess-quality-of-the-result" class="section level4">
<h4>Calculate mapping statistics to assess quality of the result</h4>
</div>
</div>
<div id="clean-using-fastp" class="section level3">
<h3>Clean using fastp</h3>
<div id="now-that-we-have-a-sense-of-the-quality-of-our-data-and-where-we-might-need-some-trimming-our-next-step-is-to-clean-the-reads-by-removing-low-quality-bases-and-leftover-illumina-sequence-adapters-from-the-sequences." class="section level5">
<h5>Now that we have a sense of the quality of our data, and where we
might need some trimming, our next step is to clean the reads by
removing low quality bases and leftover Illumina sequence adapters from
the sequences.</h5>
</div>
<div id="well-use-the-fastp-program-to-.-see-also-the-paper-published-by-chen-in-may-2023.-another-very-similar-program-for-trimming-is-trimmomatic.-it-works-great-but-fastp-does-everything-trimmomatic-does-but-in-only-a-faction-of-the-time-the-program-is-already-installed-on-our-server" class="section level4">
<h4>We’ll use the <strong><em>fastp program</em></strong> to : . See
also the paper published by Chen in May 2023. Another very similar
<strong>program for trimming is Trimmomatic</strong>. It works great,
but fastp does everything Trimmomatic does but in only a faction of the
time! The program is already installed on our server:</h4>
</div>
</div>
</div>
<div id="datapopgenfastp" class="section level2">
<h2>/data/popgen/fastp</h2>
<div id="this-job-will-be-bit-more-complicated-since-we-want-you-each-to-process-all-the-samples-for-a-given-population." class="section level6">
<h6>This job will be bit more complicated, since we want you each to
process all the samples for a given population.</h6>
</div>
<div id="since-were-doing-multiple-samples-from-the-same-population-we-want-to-be-clever-and-process-in-a-batch-instead-of-manually-one-at-a-time.-we-can-do-this-by-writing-a-bash-script-that-contains-a-loop.-this-is-the-first-taste-of-how-bash-scripting-is-powerful" class="section level5">
<h5>Since we’re doing multiple samples from the same population, we want
to be clever and process in a batch instead of manually one at a time.
We can do this by writing a bash script that contains a
<strong>loop</strong>. This is the first taste of how bash scripting is
powerful!</h5>
<p>The basic syntax of a bash loop is like this:</p>
<p>cd <path to the input data></p>
<p>for FILE in somelist</p>
<p>do</p>
<p>command1 -options ${FILE} -moreOptions command2 -options ${FILE}
-moreOptions</p>
<p>done</p>
</div>
<div id="note-the-use-of-variable-assignment-using-." class="section level4">
<h4>Note the use of variable assignment using ${}.</h4>
</div>
<div id="we-define-the-word-file-in-the-for-loop-as-the-variable-of-interest-and-then-call-the-iterations-of-it-using-file.-for-example-we-could-use-the-wildcard-character-in-a-loop-to-call-all-files-that-include-the-id-code-for-your-population-and-then-pass-those-filenames-in-a-loop-to-the-commands.-for-example-if-your-pop-was-2xxx-then-you-could-write-a-loop-that-would-process-all-the-r1-fastq-files-in-that-population-like-this" class="section level4">
<h4>We define the word FILE in the “for loop” as the variable of
interest, and then call the iterations of it using ${FILE}. For example,
we could use the wildcard character (*) in a loop to call all files that
include the ID code for your population and then pass those filenames in
a loop to the commands. For example, if your pop was “2XXX”, then you
could write a loop that would process all the R1 fastq files in that
population like this:</h4>
<p>MYPOP=“2XXX”</p>
<p>cd <path to the input data></p>
<p>for FILE in ${MYPOP}*R1.fastq.gz</p>
<p>do</p>
<p>command1 -options ${FILE} -moreOptions command2 -options ${FILE}
-moreOptions</p>
<p>done</p>
<div id="thats-the-basic-idea-lets-do-it" class="section level6">
<h6>That’s the basic idea! Let’s do it…</h6>
<p>We’ve provided a partially completed example script here:</p>
<p>/netfiles/ecogen/PopulationGenomics/scripts/fastp.sh</p>
<ol style="list-style-type: decimal">
<li>Make a copy of the script cp and put it into your ~/myscripts
directory</li>
<li>Open and edit your copied script using the program vim vim
~/myscripts/fastp.sh.</li>
<li>Edit the file so that you’re trimming the fastq files for the
population code assigned to you; you can add annotations as notes to
yourself using the hashtag (#)</li>
<li>Save the file after you’re done making changes. To do this, hit to
get out of “Insert” mode, then :wq This will write (w) the changes to
your file and then quit (q) vim. NOTE: fastp needs both read pairs
(i.e., the R1 and R2 files) given to it at the same time. This makes the
use of name variables in the loop a bit tougher since we can’t rely just
on the population name (example, 2505) but also need each individual
name too (example, 2505_9_C).</li>
</ol>
</div>
</div>
<div id="well-use-a-couple-of-tricks-to-recycle-the-r1-file-name-to-quickly-create-a-mathcing-r2-file-name-within-the-loop.-well-do-a-simialr-trick-to-quickly-name-the-output-files-adding-_clean-to-the-end.-well-talk-about-this-in-class-together-but-there-are-also-annotations-in-the-script-itself-explaining-whats-happening." class="section level3">
<h3>We’ll use a couple of tricks to recycle the R1 file name to quickly
create a mathcing R2 file name within the loop. We’ll do a simialr trick
to quickly name the output files adding “_clean” to the end. We’ll talk
about this in class together, but there are also annotations in the
script itself explaining what’s happening.</h3>
<div id="once-youre-ready-go-ahead-and" class="section level5">
<h5>Once you’re ready, go ahead and</h5>
<p>that you saved into your ~/myscripts/ folder:</p>
<p>cd ~/myscripts/ bash fastp.sh Note that the output (the trimmed and
cleaned reads) are going to get saved to a folder on the network drive
that you’ve been given write access to:
/netfiles/ecogen/PopulationGenomics/fastq_red_spruce/cleanreads</p>
</div>
</div>
</div>
<div id="visualize-pre--and-post-trimming-read-quality-with-the-fastp-html-output" class="section level2">
<h2>Visualize pre- and post-trimming read quality with the fastp html
output</h2>
<p>Conveniently, fastp is not only fast, but also produces a summary of
the change in quality pre- and post-trimming. Just like FastQC, the
output is in an html file that we can use FileZilla to transfer back to
our laptops and look at in a browser. Pretty cool!</p>
<p>Mapping cleaned and trimmed reads against the reference genome Now
that we have cleaned and trimmed read pairs, we’re ready to map them
against the reference genome.</p>
<p><strong><em>The first step of read mapping is downloading the
reference genome, freely available from the developers at
plantgenie.org: The Picea abies reference genome is based on Norway
spruce (P. abies) and published by Nystedt et
al. (2013).</em></strong></p>
<div id="you-dont-actually-need-to-download-the-genome-because-we-already-have-the-file-in-the-directory-below.-but-for-future-reference-wget-is-a-useful-command-to-ownload-files-from-the-web." class="section level4">
<h4>You don’t actually need to download the genome because we already
have the file in the directory below. But for future reference
<strong><em>wget is a useful command to ownload files from the
web</em></strong>.</h4>
<p>cd /netfiles/ecogen/PopulationGenomics/ref_genome wget “<a href="ftp://plantgenie.org:980/Data/PlantGenIE/Picea_abies/v1.0/fasta/GenomeAssemblies/Pabies01-genome.fa.gz" class="uri">ftp://plantgenie.org:980/Data/PlantGenIE/Picea_abies/v1.0/fasta/GenomeAssemblies/Pabies01-genome.fa.gz</a>”</p>
</div>
<div id="rather-than-trying-to-map-to-the-entire-19.6-gbp-reference-yikes-we-first-subsetted-the-p.-abies-reference-to-include-just-the-contigs-that-contain-one-or-more-probes-from-our-exome-capture-experiment.-for-this-we-did-a-blast-search-of-each-probe-against-the-p.-abies-reference-genome-and-then-retained-all-scaffolds-that-had-a-best-hit." class="section level4">
<h4>Rather than trying to map to the entire 19.6 Gbp reference (yikes!),
we first subsetted the P. abies reference to include just the contigs
that contain one or more probes from our exome capture experiment. For
this, we did a BLAST search of each probe against the P. abies reference
genome, and then retained all scaffolds that had a best hit.</h4>
<p>This reduced reference contains: 1,376,182,454 bp (~1.38 Gbp) in
33,679 contigs The mean (median) contig size is 10.5 (12.9) kbp The N50
of the reduced reference is 101,375 bp The indexed reduced reference
genome to use for your mapping is on our server here:
/netfiles/ecogen/PopulationGenomics/ref_genome/Pabies1.0-genome_reduced.fa</p>
</div>
<div id="to-help-make-our-scripting-approach-efficient-were-going-to-write-several-short-scripts-optimizing-each-one-at-a-time-then-put-them-together-at-the-end" class="section level4">
<h4>To help make our scripting approach efficient, we’re going to write
several short scripts, optimizing each one at a time, then put them
together at the end</h4>
<p>First, we want to <strong><em>specify the population of interest and
the paths to the input and output directories</em></strong>. We can do
this by defining variables in bash, like so:</p>
<p>Each student gets assigned a population to work with:</p>
<p>MYPOP=“XXXX”</p>
<p>Directory with the cleaned fastq files</p>
<p>INPUT=“/netfiles/ecogen/PopulationGenomics/fastq/red_spruce/cleanreads”</p>
<p>Output dir to store mapping files (bam)</p>
<p>OUT=“/netfiles/ecogen/PopulationGenomics/fastq/red_spruce/cleanreads/bam”</p>
</div>
<div id="for-mapping-well-use-the-program-bwa-which-is-a-very-efficient-and-very-well-vetted-read-mapper.-lots-of-others-exist-and-can-be-useful-to-explore-for-future-datasets.-we-tried-several-and-for-our-exome-data-bwa-seems-to-be-the-best." class="section level3">
<h3>For mapping, we’ll use the program <strong>bwa</strong>, which is a
very efficient and very well <strong>vetted read mapper</strong>. Lots
of others exist and can be useful to explore for future datasets. We
tried several, and for our exome data, bwa seems to be the best.</h3>
</div>
<div id="lets-write-a-bash-script-called-mapping.sh-that-calls-the-r1-and-r2-reads-for-each-individual-in-your-population-and-uses-the-bwa-mem2-algorithm-to-map-reads-to-the-reference-genome.-we-can-test-this-out-using-one-sample-individual-at-a-time-and-then-once-the-syntax-is-good-and-the-bugs-all-worked-out-we-can-scale-this-up-to-all-the-inds-in-our-populations." class="section level3">
<h3>Let’s write a <strong><em>bash script called
mapping.sh</em></strong> that calls the R1 and R2 reads for each
individual in your population, and uses the <strong><em>bwa-mem2
algorithm to map reads to the reference genome</em></strong>. We can
test this out using one sample (individual) at a time, and then once the
syntax is good and the bugs all worked out, we can scale this up to all
the inds in our populations.</h3>
<p>The basic bwa-mem2 command we’ll use is below. Think about how we
should write this into a loop to call all the fastq files for our
population of interest…(hint, look back at the fastp.sh script)</p>
<p>/data/popgen/bwa-mem2/bwa-mem2 mem -t 1 ${REF} ${READ1} ${READ2} &gt;
<span class="math inline">\({OUT}\)</span>{NAME}.sam where</p>
<p>-t 1 is the number of threads, or computer cpus to use (in this case,
just 1) -${REF} specifies the path and filename for the reference genome
${READ1} specifies the path and filename for the cleaned and trimmed R1
reads <span class="math inline">\({READ2} specifies the path and
filename for the cleaned and trimmed R2 reads
&gt;\)</span>{OUT}/${NAME}.sam specifies the path and filenam for the
.sam file to be saved into a new directory Other bwa options detailed
here: bwa manual page Because you’re each mapping sequences from
multiple samples (N=8/pop), it’s going to take a little while.</p>
</div>
<div id="whenever-you-have-a-job-that-will-take-a-long-time-youre-going-to-want-to-start-a-screen-session-using-the-tmux-command.-tmux-initiates-a-new-shell-window-that-wont-interrupt-or-stop-your-work-if-you-close-your-computer-log-off-the-server-or-leave-the-uvm-network.-anytime-youre-running-long-jobs-you-definitely-want-to-use-tmux." class="section level3">
<h3>Whenever you have a job that will take a long time, you’re going to
want to start a “screen session” using the tmux command.
<strong><em>tmux initiates a new shell window that won’t interrupt or
stop your work if you close your computer, log off the server, or leave
the UVM network</em></strong>. Anytime you’re running long jobs, you
definitely want to use tmux.</h3>
<p>Using it is easy. Here’s what you need to do:</p>
<p>Type tmux and hit <return></p>
<p>You are now in a “screen” — meaning you can start a job and once you
exit the screen it will keep running.</p>
<p>Start your program or run your bash script like so: bash
mapping.sh</p>
</div>
<div id="youll-see-the-program-start-running.---assuming-everything-is-working-and-youre-not-seeing-errors-you-want-to-now-detach-from-the-screen.-to-do-so-hold-the-key-down-while-you-hit-the-b-key.-then-release-the-key-and-hit-just-the-d-key." class="section level3">
<h3><strong><em>You’ll see the program start running. - Assuming
everything is working and you’re not seeing errors, you want to now
detach from the screen. To do so, hold the <control> key down while you
hit the b key. Then release the <control> key and hit just the d
key.</em></strong></h3>
<div id="if-you-do-it-right-then-you-should-exit-your-screen-session-and-go-back-to-a-regular-terminal.-if-you-dont-do-this-youll-lose-your-work" class="section level5">
<h5>If you do it right, then you should exit your screen session and go
back to a “regular” terminal. If you don’t do this, you’ll lose your
work!</h5>
</div>
<div id="you-can-recover-your-screen-by-typing-tmux-attach.-thatll-re-attach-you-back-to-your-program" class="section level4">
<h4>You can recover your screen by typing <strong><em>tmux
attach</em></strong>. That’ll re-attach you back to your program!</h4>
<p>When your script is ready, do:</p>
<p>tmux</p>
<p>bash mapping.sh Don’t forget to detach from your screen!</p>
<p>While that’s running, let’s take a look at a *** Sequence AlignMent
(SAM) file*** already available in
/netfiles/ecogen/PopulationGenomics/fastq/red_spruce/cleanreads/bam/
First, try looking at a SAM file using head and tail. tail -n 100
FILENAME.sam</p>
</div>
</div>
<div id="a-sam-file-is-a-tab-delimited-text-file-that-stores-information-about-the-alignment-of-reads-in-a-fastq-file-to-a-reference-genome-or-transcriptome.-for-each-read-in-a-fastq-file-theres-a-line-in-the-sam-file-that-includes" class="section level3">
<h3>A SAM file is a tab delimited text file that stores information
about the alignment of reads in a FASTQ file to a reference genome or
transcriptome. For each read in a FASTQ file, there’s a line in the SAM
file that includes</h3>
<ul>
<li><p>the read, aka. query, name,</p></li>
<li><p>a FLAG (number with information about mapping success and
orientation and whether the read is the left or right read),</p></li>
<li><p>the reference sequence name to which the read mapped</p></li>
<li><p>the leftmost position in the reference where the read
mapped</p></li>
<li><p>the mapping quality (Phred-scaled)</p></li>
<li><p>a CIGAR string that gives alignment information (how many bases
Match (M), where there’s an Insertion (I) or Deletion (D))</p></li>
<li><p>an ‘=’, mate position, inferred insert size (columns
7,8,9),</p></li>
<li><p>the query sequence and Phred-scaled quality from the FASTQ file
(columns 10 and 11),</p></li>
<li><p>then Lots of good information in TAGS at the end, if the read
mapped, including whether it is a unique read (XT:A:U), - the number of
best hits (X0:i:1), the number of suboptimal hits (X1:i:0).</p></li>
<li><p>The left (R1) and right (R2) reads alternate through the file.
SAM files usually have a header section with general information where
each line starts with the ‘@’ symbol. SAM and BAM files contain the same
information; SAM is human readable and BAM is in binary code and
therefore has a smaller file size.</p></li>
</ul>
<p>Find the official Sequence AlignMent file documentation can be found
here or more officially.</p>
<p>Some useful FLAGs to know - for example what do the numbers in the
second column of data mean?</p>
<p>Here’s a SAM FLAG decoder by the Broad Institute.</p>
</div>
<div id="how-can-we-get-a-summary-of-how-well-our-reads-mapped-to-the-reference" class="section level3">
<h3>How can we get a summary of how well our reads mapped to the
reference?</h3>
<div id="we-can-use-the-program-sambamba-for-manipulating-sambam-files.-sambamba.-sambamba-is-closely-related-to-its-progenitor-program-samtools-which-is-written-by-the-same-scientist-who-develop-bwa-heng-li.-sambamba-has-been-re-coded-to-increase-efficiency-speed." class="section level4">
<h4>We can use the program <strong><em>sambamba for manipulating sam/bam
files</em></strong>. sambamba. Sambamba is closely related to its
progenitor program samtools which is written by the same scientist who
develop bwa, Heng Li. sambamba has been re-coded to increase efficiency
(speed).</h4>
<p>First we have to convert the sam file to bam format:</p>
<p>sambamba view -S -f bam IN.sam -o OUT.bam</p>
<p>Then we can use the command flagstats gets us some basic info on how
well the mapping worked:</p>
<p>sambamba flagstat FILENAME.bam</p>
<hr />
<div id="id-section3">

</div>
</div>
</div>
</div>
<div id="entry-3-2023-09-18." class="section level2">
<h2>Entry 3: 2023-09-18.</h2>
<p>Learning Objectives for 09/18/23 Introduce lab notebooks Visualize
sequence alignment files (<em>.sam) Process the mapping file </em>sam to
binary (*.bam), sort, and remove duplicate reads Calculate mapping
statistics to assess quality of the result 1. Lab Notebooks: Take notes
on your workflow today so you can remember what you did for the future!
An important step is taking good notes on your workflow so you can
remember what you did down the road (your “future self”) and share your
process with others (reproducible science!). It’s also really important
once you start making detailed decisions that will affect the analysis
outcome of your data, so you can recreate the results and explore the
effect of different assumptions/decisions.</p>
<p>We want you to keep such a notebook for each module in the course
(kind of like you would for each experiment, or each thesis chapter you
will work on). We’ve provided you with a notebook template based on the
markdown (md) language.</p>
<p>Notebook template in markdown (md) format</p>
<p>You should save a copy of this template to your github repo and
rename it PopulationGenomics_Notebook.md in your main repo folder. You
can then open it up and edit this file directly within RStudio, taking
notes using either the Source interface if you know the markdown
language (or want to learn) or you can use RStudio’s built-in markdown
GUI editor under the Visual tab. This works very similar to Word.</p>
<p>Here’s a cheatsheet for markdown language if you want to write using
Source code:</p>
<p>Markdown cheatsheet</p>
<p>When you’re done your entries, save and quit RStudio, then commit
your changes and push to Github. You can check out your notebook on your
repo’s site, and Github will automatically render the markdown code into
a nice format!</p>
<ol start="2" style="list-style-type: decimal">
<li>Visualize the mapping: By now, you should all have Sequence
AlignMent (SAM) files for the inds in your populations! Let’s take a
peek at one of the non-binary (sam) alignment files
/netfiles/ecogen/PopulationGenomics/fastq/red_spruce/cleanreads/bam/</li>
</ol>
<p>First, try looking at a SAM file using head and tail. Pick one of
your files (just one) to play with below: tail -n 2 YOURFILENAME.sam A
SAM file is a tab delimited text file that stores information about the
alignment of reads in a FASTQ file to a reference genome or
transcriptome. For each read in a FASTQ file, there’s a line in the SAM
file that includes</p>
<p>the read, aka. query, name, a FLAG (number with information about
mapping success and orientation and whether the read is the left or
right read), the reference sequence name to which the read mapped the
leftmost position in the reference where the read mapped the mapping
quality (Phred-scaled) a CIGAR string that gives alignment information
(how many bases Match (M), where there’s an Insertion (I) or Deletion
(D)) an ‘=’, mate position, inferred insert size (columns 7,8,9), the
query sequence and Phred-scaled quality from the FASTQ file (columns 10
and 11), then Lots of good information in TAGS at the end, if the read
mapped, including whether it is a unique read (XT:A:U), the number of
best hits (X0:i:1), the number of suboptimal hits (X1:i:0). The left
(R1) and right (R2) reads alternate through the file. SAM files usually
have a header section with general information where each line starts
with the ‘@’ symbol. SAM and BAM files contain the same information; SAM
is human readable and BAM is in binary code and therefore has a smaller
file size.</p>
<p>Find the official Sequence AlignMent file documentation can be found
here or more officially.</p>
<p>Here’s a SAM FLAG decoder by the Broad Institute.</p>
<p>You can look up TAGs specific to bwa mem mapping here</p>
<ol start="3" style="list-style-type: decimal">
<li>Process our mapping files using samtools and sambamba We can use the
program sambamba for manipulating alignment (sam/bam) files. sambamba.
Sambamba is closely related to its progenitor program samtools which is
written by the same scientist who develop bwa, Heng Li. sambamba has
been re-coded to increase efficiency (speed).</li>
</ol>
<p>There are several steps we need to do:</p>
<p>convert sam alignment file to (binary) bam format sort the bam file
by its read coordinates mark and remove PCR duplicate reads index the
sorted, duplicate removed alignment for quick lookup Here’s a script
that we can customize for the above jobs:
/netfiles/ecogen/PopulationGenomics/scripts/process_bams.sh</p>
<p>Copy that script into your ~/myscripts folder and open it in vim to
edit</p>
<p>When you’re ready, enter a screen session using tmux then execute
your script bash process_bams.sh. If you get an error that the script
doesn’t exist, then either cd into the ~/myscripts directory before
running your script, or incorporate the path into the file name when you
give your bash command.</p>
<p>Detach from the screen using <CTRL>+b then d. You can always reattach
by tmux attach-session</p>
<ol start="4" style="list-style-type: decimal">
<li>Calculate mapping stats: How can we get a summary of how well our
reads mapped to the reference? We can use the program samtools Written
by Heng Li, the same person who wrote bwa. It is a powerful tool for
manipulating sam/bam files.</li>
</ol>
<p>The samtools command flagstat gets us some basic info on how well the
mapping worked</p>
<p>We can also estimate depth of coverage (avg. number of reads/site)
using the samtools command depth</p>
<p>We’ll use both of these commands in loops to assess the mapping stats
on each sample in our population.</p>
<p>We’ll also use the awk tool to help format the output.</p>
<p>Here’s a script to get us started:
/netfiles/ecogen/PopulationGenomics/scripts/bam_stats.sh</p>
<p>If there’s time while that’s running, we can take a look at one of
our alignment files (sam or bam) using an integrated viewed in samtools
called tview. To use it, simply call the program and command, followed
by the sam/bam file you want to view and the path to the reference
genome. For example: samtools tview
/netfiles/ecogen/PopulationGenomics/fastq/red_spruce/cleanreads/bam/XXXX_XX_X.sorted.rmdup.bam
/netfiles/ecogen/PopulationGenomics/ref_genome/Pabies1.0-genome_reduced.fa</p>
<hr />
<div id="id-section4">

</div>
</div>
<div id="entry-4-2023-09-20" class="section level2">
<h2>Entry 4: 2023-09-20</h2>
<p>Learning Objectives for 09/20/23</p>
<p>Finish calculating mapping statistics to assess quality of the result
Introduce use of genotype-likelihoods for analyzing diversity in low
coverage sequences Use the ‘ANGSD’ program to calculate nucleotide
diversity (thetas) and neutrality stats 1. Calculate mapping stats: How
can we get a summary of how well our reads mapped to the reference? We
can use the program samtools Written by Heng Li, the same person who
wrote bwa. It is a powerful tool for manipulating sam/bam files.</p>
<p>The samtools command flagstat gets us some basic info on how well the
mapping worked</p>
<p>We can also estimate depth of coverage (avg. number of reads/site)
using the samtools command depth</p>
<p>We’ll use both of these commands in loops to assess the mapping stats
on each sample in our population.</p>
<p>We’ll also use the awk tool to help format the output. (awk
cheatsheet here)</p>
<p>Here’s a script to get us started:
/netfiles/ecogen/PopulationGenomics/scripts/bam_stats.sh</p>
<ol start="2" style="list-style-type: decimal">
<li>Inference of population genomics from the aligned sequence data:
should we call genotypes? Many of the papers you’ll read that do popgen
on NGS data have a SNP calling step that results in a specific gneotype
being called for each SNP site for each individual. For example,</li>
</ol>
<p>SNP Ind1 Ind 2 1 CC CT 2 AG AA 3 GT TT But how do we know that Ind1
is homozygous at SNP-1 (CC) – couldn’t it be CT and we just didn’t have
enough coverage to observe the second allele?</p>
<p>The basic problem is that read data are counts that produce a
binomial distribution of allele calls at a given site, and if you have
few reads, you might by chance not observe the true genotype. So, what’s
the right thing to do?</p>
<p>As with almost anything in statistics, the right thing to do is not
throw away that uncertainty, but instead incorporate it into your
analysis. That’s what we’re going to do…</p>
<p>Genotype-free population genetics using genotype likelihoods A
growing movement in popgen analysis of NGS data is embracing the use of
genotype likelihoods to calculate stats based on each individual having
a likelihood (probability) of being each genotype.</p>
<p>A genotype likelihood (GL) is essentially the probability of
observing the sequencing data (reads containing a particular base),
given the genotype of the individual at that site.</p>
<p>These probabilities are modeled explicitly in the calculation of
population diversty stats like Theta-pi, Tajima’s D, Fst, PCA, etc…;
thus not throwing out any precious data, but also making fewer
assumptions about the true (unknown) genotype at each locus</p>
<p>We’re going to use this approach with the program ‘ANGSD’, which
stands for ‘Analysis of Next Generation Sequence Data’</p>
<p>This approach was pioneered by Rasmus Nielsen, published originally
in Korneliussen et al. 2014.</p>
<p>ANGSD has a user’s manual (it’s a work in progress…)</p>
<p>The basic work flow of ANGSD goes like this: Create a list of bam
files for the samples you want to analyze Estimate genotype likelihoods
(GL’s) and allele frequencies after filtering to minimize noise Use GL’s
to: estimate the site frequency spectrum (SFS)</p>
<p>estimate nucleotide diversities and neutrality stats (Thetas,
Tajima’s D, …)</p>
<ol style="list-style-type: decimal">
<li>In your ~/myscripts folder, enter vim and create a file called
ANGSD.sh Create the header for your inputs and outputs</li>
</ol>
<p>mkdir ~/myresults/ANGSD</p>
<p>INPUT=“”</p>
<p>OUTPUT=~/myresults/ANGSD</p>
<p>REF=“/netfiles/ecogen/PopulationGenomics/ref_genome/Pabies1.0-genome_reduced.fa”</p>
<p>MYPOP=“”</p>
<p>ls <span class="math inline">\({INPUT}/\)</span>{MYPOP}*sorted.rmdup.bam
&gt;<span class="math inline">\({OUTPUT}/\)</span>{MYPOP}_bam.list Write
(w) and quit(q) your file and try running it at the command line.</p>
<p>Check your output bamlist to see it was made properly!</p>
<p>Where would you look for this file? (Hint, refer back to the ls
command that makes it). How would you verify its contents? (hint: use
head or cat or even vim) 2. Open your ANGSD.sh script back up in vim
Estimate your GL’s and allele freqs, optionally filtering for base and
mapping quality, sequencing depth, SNP probability, minor allele
frequency, etc.</p>
<p>Add the following code chunk at the bottom of your script:</p>
</div>
</div>
<div id="file-suffix-to-distinguish-analysis-choices" class="section level1">
<h1>File suffix to distinguish analysis choices</h1>
<p>SUFFIX=“”</p>
</div>
<div id="estimating-gls-and-allele-frequencies-for-all-sites-with-angsd" class="section level1">
<h1>Estimating GL’s and allele frequencies for all sites with ANGSD</h1>
<div id="section" class="section level22">
<p class="heading"></p>
<p>ANGSD -b <span class="math inline">\({OUTPUT}/\)</span>{MYPOP}<em>bam.list<br />
-ref ${REF} -anc ${REF}<br />
-out <span class="math inline">\({OUTPUT}/\)</span>{MYPOP}</em>${SUFFIX}<br />
-nThreads 1<br />
-remove_bads 1<br />
-C 50<br />
-baq 1<br />
-minMapQ 20<br />
-minQ 20<br />
-GL 1<br />
-doSaf 1<br />
##### below filters require <code>do-Counts</code> #-doCounts 1<br />
#-minInd 4<br />
#-setMinDepthInd 1<br />
#-setMaxDepthInd 40<br />
#-setMinDepth 10<br />
#-skipTriallelic 1<br />
#-doMajorMinor 1<br />
##### below filters require <code>doMaf</code> #-doMaf 1<br />
#-SNP_pval 1e-6<br />
#-minMaf 0.01 What do all these options mean?</p>
<p>Option Description -nThreads 1 how many cpus to use – be conservative
-remove_bads 1 remove reads flagged as ‘bad’ by samtools -C 50 enforce
downgrading of map quality if contains excessive mismatches -baq 1
estimates base alignment qualities for bases around indels -minMapQ 20
threshold for minimum read mapping quality (Phred) -minQ 20 threshold
for minimum base quality (Phred) -GL 1 calculate genotype likelihoods
(GL) using the Samtools formula -doSaf 1 output allele frequency
likelihoods for each site -doCounts 1 output allele counts for each site
-minInd 4 min number of individuals to keep a site (see also ext 2
filters) -setMinDepthInd 1 min read depth for an individual to count
towards a site -setMaxDepthInd 40 max read depth for an individual to
count towards a site -setMinDepth 10 min read depth across ALL
individual to keep a site -skipTriallelic 1 don’t use sites with &gt;2
alleles -doMajorMinor 1 fix major and minor alleles the same across all
samples -doMaf 1 calculate minor allele frequencies -SNP_pval 1e-6 Keep
only site highly likely to be polymorphic (SNPs) -minMaf 0.01 Keep only
sites with minor allele freq &gt; some proportion. NOTES</p>
<p>If you want to restrict the estimation of the genotype likelihoods to
a particular set of sites you’re interested in, add the option -sites
selected_sites.txt (tab delimited file with the position of the site in
column 1 and the chromosome in column 2) or use -rf
selected_chromosome.chrs (if listing just the unique “chromosomes” or
contigs you want to anlayze) Some popgen stats you want to estimate only
the polymorphic sites; for this you should include the -SNP_pval 1e-6
option to eliminate monomorphic sites when calculating your GL’s There
are good reasons to do it BOTH ways, with and without the -SNP_pval 1e-6
option. Keeping the monomorphic sites is essential for getting proper
estimates of nucleotide diversity and Tajima’s D. But other analyses
such as PCA or GWAS want only the SNPs. Write (q) and quit (q) your
script. Then run at the command line</p>
<p>3a. In your ~/myscripts folder, create ANGSD_doTheta.sh to estimate
the SFS and nucleotide diversity stats for your pop Based on the saf.idx
files from ANGSD GL calls, you can estimate the Site Frequency Spectrum
(SFS), which is the precursor to many other analyses such as nucleotide
diversities (as well as Fst, demographic history analysis, etc.)</p>
<p>REF=“/netfiles/ecogen/PopulationGenomics/ref_genome/Pabies1.0-genome_reduced.fa”</p>
<p>OUT=~/myresults/ANGSD</p>
<p>MYPOP=“”</p>
<p>SUFFIX=“”</p>
<p>#Estimation of the SFS for all sites using the FOLDED SFS realSFS
<span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.saf.idx<br />
-maxIter 1000<br />
-tole 1e-6<br />
-P 1<br />
&gt; <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.sfs 3b. After
the SFS, estimate the theta diversity stats: Once you have the SFS, you
can estimate the thetas and neutrality stats by adding the following
code chunk at the end of your ANGSD_doTheta.sh script:</p>
</div>
</div>
<div id="estimate-thetas-and-stats-using-the-sfs" class="section level1">
<h1>Estimate thetas and stats using the SFS</h1>
<p>realSFS saf2theta <span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.saf.idx<br />
-sfs <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.sfs<br />
-outname <span class="math inline">\({OUT}/\)</span>{MYPOP}_${SUFFIX}</p>
<p>thetaStat do_stat <span class="math inline">\({OUT}/\)</span>{MYPOP}_${SUFFIX}.thetas.idx If we
wanted to analyze this on sliding windows, we could instead replace the
above code chunk with the following:</p>
</div>
<div id="for-sliding-window-analysis" class="section level1">
<h1>For sliding window analysis:</h1>
<p>thetaStat do_stat <span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.thetas.idx<br />
-win 50000<br />
-step 10000<br />
-outnames <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.thetasWindow.gz
For either of the results files above, the first column of the results
file (*.thetas.idx.pestPG) is formatted a bit funny and we don’t really
need it. We can use the cut command to get rid of it:</p>
<p>cut -f2- <span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.thetas.idx.pestPG
&gt; <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.thetas This
is now ready to bring into R to look at the mean and variability in
nucleotide diversity for our pop. How does it compare to others?</p>
<p>We can compare the diversity in our different pops by entering your
diversity stats in this google doc:</p>
<p><a href="https://docs.google.com/spreadsheets/d/1y3GMvnGP65fYfBoJsdrixGLkGCe_TcDlo_7GFyPe1QM/edit?usp=sharing" class="uri">https://docs.google.com/spreadsheets/d/1y3GMvnGP65fYfBoJsdrixGLkGCe_TcDlo_7GFyPe1QM/edit?usp=sharing</a></p>
<hr />
<div id="id-section5">

</div>
<div id="entry-5-2023-09-25." class="section level2">
<h2>Entry 5: 2023-09-25.</h2>
<p>Learning Objectives for 09/25/23 Calculate diversity stats for our
focal pops (SFS, theta-W, theta-Pi, Tajima’s D) Summarize the results in
R and share to google doc Introduce Fst in ANGSD using genotype
probabilities 1. Calculate SFS and diversity stats At the end of our
last session, we used ANGSD to estimate genotype likelihoods for our red
spruce populations. We wrote the script ANGSD.sh to work on these, which
should have the following output files in your ~/myresults/ANGSD
directory:</p>
<p>-rw-r–r–. 1 kellrlab users 1417463962 Sep 20 11:44 9999_.saf.gz
-rw-r–r–. 1 kellrlab users 1086161 Sep 20 11:44 9999_.saf.idx -rw-r–r–.
1 kellrlab users 79501025 Sep 20 11:44 9999_.saf.pos.gz These “saf”
files contain “site allele frequency” likelihoods, and are the info
needed to estimate stats that depend on population allele frequencies,
like nucleotide diversities, neutrality stats like Tajima’s D, and
population divergence stats like Fst. Each of these stats depends on the
SFS – the Site Frequency Spectrum So, our workflow will be to use the
.saf files to estimate the SFS, and then use the SFS to estimate our
diversity stats and Fst.</p>
<p>In your ~/myscripts folder, create ANGSD_doTheta.sh to estimate the
SFS and nucleotide diversity stats for your pop Based on the saf.idx
files from ANGSD GL calls, first estimate the Site Frequency Spectrum
(SFS)</p>
<p>REF=“/netfiles/ecogen/PopulationGenomics/ref_genome/Pabies1.0-genome_reduced.fa”</p>
<p>OUT=~/myresults/ANGSD</p>
<p>MYPOP=“”</p>
<p>SUFFIX=“”</p>
<p>#Estimation of the SFS for all sites using the FOLDED SFS realSFS
<span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.saf.idx<br />
-maxIter 1000<br />
-tole 1e-6<br />
-P 1<br />
-fold 1<br />
&gt; <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.sfs After the
SFS, estimate the theta diversity stats: Once you have the SFS, you can
estimate the thetas and neutrality stats by adding the following code
chunk at the end of your ANGSD_doTheta.sh script:</p>
</div>
</div>
<div id="estimate-thetas-and-stats-using-the-sfs-1" class="section level1">
<h1>Estimate thetas and stats using the SFS</h1>
<p>realSFS saf2theta <span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.saf.idx<br />
-sfs <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.sfs<br />
-outname <span class="math inline">\({OUT}/\)</span>{MYPOP}_${SUFFIX}</p>
<p>thetaStat do_stat <span class="math inline">\({OUT}/\)</span>{MYPOP}_${SUFFIX}.thetas.idx If we
wanted to analyze this on sliding windows, we could instead replace the
above code chunk with the following:</p>
</div>
<div id="for-sliding-window-analysis-1" class="section level1">
<h1>For sliding window analysis:</h1>
<p>thetaStat do_stat <span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.thetas.idx<br />
-win 50000<br />
-step 10000<br />
-outnames <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.thetasWindow.gz
An important distinction! The unfolded vs. folded SFS The big difference
here is whether we are confident in the ancestral state of each variable
site (SNP) in our dataset</p>
<p>If we know the ancestral state, then the best info is contained in
the unfolded SFS, which shows the frequency histogram of how many
derived loci are rare vs. common</p>
<p>bins in the unfolded SFS go from 0 to 2N – why? When you don’t know
the ancestral state confidently, you can make the SFS based on the minor
allele (the less frequent allele; always &lt; 0.5 in the
population).</p>
<p>bins in the folded SFS go from 0 to 1N – why? Essentially, the folded
spectra wraps the SFS around such that high frequency “derived” alleles
are put in the small bins (low minor allele freq).</p>
<p>Now we have some diversity results! For either of the results files
above, the first column of the results file (*.thetas.idx.pestPG) is
formatted a bit funny and we don’t really need it. There are also a
bunch of other neutrality stats that we don’t need right now. We can use
the cut command to get rid of these extra columns. The below cut command
retains columns 2-5, 9 and 14</p>
<p>cut -f2-5,9,14 <span class="math inline">\({OUT}/\)</span>{MYPOP}<em>${SUFFIX}.thetas.idx.pestPG
&gt; <span class="math inline">\({OUT}/\)</span>{MYPOP}</em>${SUFFIX}.thetas The
columns correspond to the following stats:</p>
<p>Col Statistic Description 2 Chr The chromosome or (more
appropriately) contig being analyzed 3 WinCenter The center of the
contig, in bp 4 tW Watterson’s Theta – an estimate of nucleotide
diversity based on segregating sites 5 tP Theta Pi – estimate of
nucleotide diversity based on pairwise divergence 9 Tajima Tajima’s D –
a neutrality stat that tests for the difference in tW-tP 14 nSites The
number of base pairs being analyzed along this stretch of cont. 2.
Summarize diversity stats in R We’re now ready to use Filezilla to
download these 2 files:</p>
<p>.thetas.idx.pestPG .sfs Save these to your results folder in your
githuib repo so we can import into R to look at the mean and variability
in nucleotide diversity for our pop. Here’s some basic R code to help
you along:</p>
<p>setwd(““) # set your path to your results folder in your repo where
you saved your diversity stats file</p>
<p>list.files() # list out the files in this folder to make sure you’re
in the right spot.</p>
</div>
<div id="first-lets-read-in-the-diversity-stats" class="section level1">
<h1>First let’s read in the diversity stats</h1>
<p>theta &lt;- read.table(“_.thetas”,sep=“,header=T)</p>
<p>theta<span class="math inline">\(tWsite = theta\)</span>tW/theta<span class="math inline">\(nSites #scales the theta-W by the number of sites
theta\)</span>tPsite = theta<span class="math inline">\(tP/theta\)</span>nSites #scales the theta-Pi by
the number of sites</p>
<p>summary(theta)</p>
</div>
<div id="you-can-order-the-contig-list-to-show-you-the-contigs-with-the-highest-values-of-tajimas-d-or-the-lowest" class="section level1">
<h1>You can order the contig list to show you the contigs with the
highest values of Tajima’s D, or the lowest</h1>
<p>head(theta[order(theta$Tajima, decreasing = TRUE),]) # top 10
Tajima’s D values</p>
<p>head(theta[order(theta$Tajima, decreasing = FALSE),]) # bottom 10
Tajima’s D values</p>
<p>#You can also look for contigs that have combinations of high
Tajima’s D and low diversity – these could represent outliers for
selection #theta[which(theta<span class="math inline">\(Tajima&gt;1.5
&amp; theta\)</span>tPsite&lt;0.001),]</p>
<p>sfs&lt;-scan(‘9999_.sfs’) sfs&lt;-sfs[-c(1,which(sfs==0))]
sfs&lt;-sfs/sum(sfs)</p>
</div>
<div id="be-sure-to-replace-9999-with-your-pop-code-in-the-main-legend-below" class="section level1">
<h1>Be sure to replace “9999” with your pop code in the “main” legend
below</h1>
<p>barplot(sfs,xlab=“Chromosomes”, names=1:length(sfs),
ylab=“Proportions”, main=“Pop 9999 Site Frequency Spectrum”,
col=‘blue’)</p>
</div>
<div id="put-the-nucleotide-diversities-tajimas-d-and-sfs-into-a-4-panel-figure" class="section level1">
<h1>Put the nucleotide diversities, Tajima’s D, and SFS into a 4-panel
figure</h1>
<p>par(mfrow=c(2,2)) hist(theta<span class="math inline">\(tWsite,
xlab=&quot;theta-W&quot;, main=&quot;Watterson&#39;s theta&quot;)
hist(theta\)</span>tPsite, xlab=“theta-Pi”, main=“Pairwise Nucleotide
Diversity”) hist(theta$Tajima, xlab=“D”, main=“Tajima’s D”)
barplot(sfs,names=1:length(sfs),main=‘Site Frequency Spectrum’)</p>
</div>
<div id="to-reset-the-panel-plotting-execute-the-line-below" class="section level1">
<h1>To reset the panel plotting, execute the line below:</h1>
<p>dev.off() We can compare the diversity in our different pops by
entering your diversity stats in this google doc:</p>
<p><a href="https://docs.google.com/spreadsheets/d/1y3GMvnGP65fYfBoJsdrixGLkGCe_TcDlo_7GFyPe1QM/edit?usp=sharing" class="uri">https://docs.google.com/spreadsheets/d/1y3GMvnGP65fYfBoJsdrixGLkGCe_TcDlo_7GFyPe1QM/edit?usp=sharing</a></p>
<ol start="3" style="list-style-type: decimal">
<li>Use ANGSD and the SFS for multiple pops to calculate genetic
divergence between pops (Fst) We can calculate Fst between any pair of
populations by comparing their SFS to each other. For this, we’ll need
to estimate the SFS for pairs of populations; we can each contribute to
the overall analysis by looking at how our focal pop is divergent from
the others.</li>
</ol>
<p>For this analysis, let’s calculate Fst between our focal red spruce
population (MYPOP) and the black spruce samples…this could tell us which
of our pops might be hybridizing. What would we expect for Fst in this
case?</p>
<p>Let’s write a bash script called ANGSD_Fst.sh that includes the
following code:</p>
</div>
<div id="start-with-the-usual-bash-script-header" class="section level1">
<h1>Start with the usual bash script header</h1>
</div>
<div id="give-yourself-some-notes" class="section level1">
<h1>Give yourself some notes</h1>
</div>
<div id="path-to-black-spruce-bs-input-saf.idx-data" class="section level1">
<h1>Path to Black Spruce (BS) input saf.idx data:</h1>
<p>BLKSPR=“/netfiles/ecogen/PopulationGenomics/fastq/black_spruce/cleanreads/bam/ANGSD”</p>
<p>#Path to save your output:</p>
<p>OUTPUT=</p>
<p>MYPOP=“”</p>
<p>SUFFIX=“”</p>
</div>
<div id="estimate-fst-between-my-red-spruce-pop-and-black-spruce" class="section level1">
<h1>Estimate Fst between my red spruce pop and black spruce:</h1>
<p>realSFS ${MYPOP}_.saf.idx <span class="math inline">\({BLKSPR}/BS_all.saf.idx -P 1
&gt;\)</span>{MYPOP}<em>BS.sfs realSFS fst index ${MYPOP}</em>.saf.idx
${BLKSPR}/BS_all.saf.idx -sfs ${MYPOP}_BS.sfs -fstout ${MYPOP}_BS
-whichFst 1 realSFS fst stats ${MYPOP}_BS.fst.idx</p>
<table style="width:14%;">
<colgroup>
<col width="13%" />
</colgroup>
<tbody>
<tr class="odd">
<td></td>
</tr>
<tr class="even">
<td>## Entry 7: 2020-10-02.</td>
</tr>
<tr class="odd">
<td>Learning Objectives for 10/02/23 Part 1 - Review the population
structure results - Perform a scan for selection and identify contigs
with outlier loci - Identify and visualize outlier loci</td>
</tr>
<tr class="even">
<td>1. What have we learned about the genetic structure from the PCA and
admixture plots? Let’s take a look at the PCA and admixture figures
posted to the Slack #coding channel.</td>
</tr>
<tr class="odd">
<td>What does the PCA seem to be telling us?</td>
</tr>
<tr class="even">
<td>What different picture does the admixture plot reveal? How does it
relate to the PCA?</td>
</tr>
<tr class="odd">
<td>Would we want to look for higher levels of K in the admixture
analysis? How do we do that??</td>
</tr>
<tr class="even">
<td>2. How has selection acted to drive divergence of individual loci in
excess of the population structure we’ve observed? When selection acts
in response to local environmental conditions, we observe an excess of
population structure at certain loci. This can be thought of as Fst at a
single locus exceeding some background level of divergence that exists
across the genome as a whole. We call these “Fst outliers” and
identifying these outliers is a major goal in ecological genomic
studies. But, a major challenge is how best to control for background
population structure – in other words, how can we sift out the
“outliers” from the rest of the population structure in the genome?</td>
</tr>
<tr class="odd">
<td>One approach to identifying Fst outliers is using genetic PCA to (a)
identify the major axes of population structure, and then (b) find loci
that “load” very strongly on these axes, indicating their exceptional
divergence is probably driven by the action of selection in excess of
genetic drift.</td>
</tr>
<tr class="even">
<td>We can run a scan for Fst outliers using pcANGSD, just like we did
for the genetic PCA, but here we initiate a selection scan for loci that
are exceptionally divergent along one or more of the inferred genetic PC
axes. This method is especially helpful when it’s hard to define what
are genetic “populations”, since it uses the genetic PCA to infer the
genetic structure directly. The method employed in pcANGSD follows the
“FastPCA” selection scan method described by Galinsky et al. 2015</td>
</tr>
<tr class="odd">
<td>The approach essentially uses the SNP weights or “loadings” on a
given genetic PC axis to determine the strength of association, and if
it exceeds the expectation due to neutral drift.</td>
</tr>
<tr class="even">
<td>Let’s write a bash script called pcANGSD_selection.sh that includes
the following code. Recall that pcANGSD uses genotype likelihoods in
“beagle” format, which have been calculated for you (see last tutorial)
and can be found here: /netfiles/ecogen/PopulationGenomics/ANGSD/</td>
</tr>
<tr class="odd">
<td>Here’s a start to our script:</td>
</tr>
<tr class="even">
<td>INPUT=“”</td>
</tr>
<tr class="odd">
<td>OUTPUT=</td>
</tr>
<tr class="even">
<td>SUFFIX=“allRS_poly”</td>
</tr>
<tr class="odd">
<td>cp ${INPUT}/allRS_bam.list ${OUTPUT}</td>
</tr>
<tr class="even">
<td># Activates the pcANGSD environment and run the selection scan:</td>
</tr>
<tr class="odd">
<td>source /data/popgen/pcangsd/venv/bin/activate</td>
</tr>
<tr class="even">
<td>pcangsd -b <span class="math inline">\({INPUT}/\)</span>{SUFFIX}.beagle.gz<br />
-o <span class="math inline">\({OUTPUT}/\)</span>{SUFFIX}<br />
-e &lt;what number should we use here?&gt; –selection<br />
–minMaf 0.05<br />
–sites_save<br />
–snp_weights<br />
–threads 1</td>
</tr>
<tr class="odd">
<td>Note the minMaf 0.05 option tests only those loci that have a minor
allele frequency of 0.05 of greater. That’s because there’s little
statistical power for testing association of alleles that are very rare
(&lt;5% in the sample).</td>
</tr>
<tr class="even">
<td>It should run quickly! Once finished, you’ll the see a new set of
files, including:</td>
</tr>
<tr class="odd">
<td>File name contents allRS_poly.selection.npy selection scores for
each locus on each tested PC axis allRS_poly.weights.npy weights that
show how strongly each locus “loads” on each PC axis allRS_poly.sites
for each locus shows whether it got tested (1) or not (0) You’ll also
want to get a file with minor allele frequencies (“maf”) that ANGSD
makes when first estimating genotype likelihoods. You have one of these
files from when you worked on your focal pops, but we need one for the
“allRS_poly” run of all red spruce individuals for just polymorphic
sites. I’ve provided this here:</td>
</tr>
<tr class="even">
<td>/netfiles/ecogen/PopulationGenomics/ANGSD/allRS_poly.mafs.gz</td>
</tr>
<tr class="odd">
<td>We need to combine the “mafs” file and the “sites” file to create a
file that has the contig and position info for each locus that got
tested for selection. Here’s a few lines of bash code to create this
file:</td>
</tr>
<tr class="even">
<td>First, we need to add a header to the “sites” file: sed -i ‘1i
kept_sites’ ~/myresults/ANGSD/allRS_poly.sites</td>
</tr>
<tr class="odd">
<td>Next, we need to get the contig and position info out of the “mafs”
file and combine it with the “sites” file: zcat
/netfiles/ecogen/PopulationGenomics/ANGSD/allRS_poly.mafs.gz | cut -f1-7
| paste - ~/myresults/ANGSD/allRS_poly.sites
&gt;allRS_poly_mafs.sites</td>
</tr>
<tr class="even">
<td>Now we have all the files we need! Be sure to transfer the 3 listed
above, plus the new allRS_poly_mafs.sites you just made, using FileZilla
and save into your repo on your laptop. When you have these files
transferred (don’t forget where you saved them to on your laptop!), open
up RStudio and let’s start making some figures!</td>
</tr>
<tr class="odd">
<td>3. Idenitfy and visualize outlier loci We’ll use R to import the
selection scan results and associated meta-data, assign p-values for
each tested locus, and visualize the results!</td>
</tr>
<tr class="even">
<td>Just a reminder, the following is R code, not bash. ;)</td>
</tr>
<tr class="odd">
<td>################################### # Selection scans for red spruce
# ###################################</td>
</tr>
<tr class="even">
<td>library(RcppCNPy) # for reading python numpy (.npy) files</td>
</tr>
<tr class="odd">
<td>setwd(“~/Documents/Github/Ecological_Genomics/Fall_2023/pcangsd/”)</td>
</tr>
<tr class="even">
<td>list.files()</td>
</tr>
<tr class="odd">
<td>### read in selection statistics (these are chi^2 distributed)</td>
</tr>
<tr class="even">
<td>s&lt;-npyLoad(“allRS_poly.selection.npy”)</td>
</tr>
<tr class="odd">
<td># convert test statistic to p-value pval &lt;-
as.data.frame(1-pchisq(s,1)) names(pval) = “p_PC1”</td>
</tr>
<tr class="even">
<td>## read positions p &lt;-
read.table(“allRS_poly_mafs.sites”,sep=“,header=T, stringsAsFactors=T)
dim(p)</td>
</tr>
<tr class="odd">
<td>p_filtered = p[which(p$kept_sites==1),] dim(p_filtered)</td>
</tr>
<tr class="even">
<td># How many sites got filtered out when testing for selection?
Why?</td>
</tr>
<tr class="odd">
<td>## make manhattan plot plot(-log10(pval<span class="math inline">\(p_PC1), col=p_filtered\)</span>chromo,
xlab=“Position”, ylab=“-log10(p-value)”, main=“Selection outliers:
pcANGSD e=1 (K2)”)</td>
</tr>
<tr class="even">
<td># We can zoom in if there’s something interesting near a
position…</td>
</tr>
<tr class="odd">
<td>plot(-log10(pval<span class="math inline">\(p_PC1[2e05:2.01e05]),
col=p_filtered\)</span>chromo, xlab=“Position”, ylab=“-log10(p-value)”,
main=“Selection outliers: pcANGSD e=1 (K2)”)</td>
</tr>
<tr class="even">
<td># get the contig with the lowest p-value for selection sel_contig
&lt;- p_filtered[which(pval==min(pval$p_PC1)),c(“chromo”,“position”)]
sel_contig</td>
</tr>
<tr class="odd">
<td># get all the outliers with p-values below some cutoff cutoff=1e-3 #
equals a 1 in 5,000 probability outlier_contigs &lt;-
p_filtered[which(pval&lt;cutoff),c(“chromo”,“position”)]
outlier_contigs</td>
</tr>
<tr class="even">
<td># how many outlier loci &lt; the cutoff?
dim(outlier_contigs)[1]</td>
</tr>
<tr class="odd">
<td># how many unique contigs harbor outlier loci?
length(unique(outlier_contigs$chromo)) How do we find out what
functional genes are contained on the unique contigs harboring
outliers?</td>
</tr>
<tr class="even">
<td>We can use the Picea abies reference genome annotation to get the
genes based on the outlier contigs, then test for enrichment of gene
function using available public databases.</td>
</tr>
<tr class="odd">
<td>First, export your unique outlier contigs in R:</td>
</tr>
<tr class="even">
<td>write.table(unique(outlier_contigs$chromo),
“allRS_poly_PC1_outlier_contigs.txt”, sep=“, quote=F, row.names=F,
col.names=F) Transfer back to the server and then grep out the gene IDs.
We can do this with a series of piped commands like so:</td>
</tr>
<tr class="odd">
<td>Define your paths to the ref genome annotation on the server and
your newly made outlier loci file</td>
</tr>
<tr class="even">
<td>Use zcat to open the annotation without unzipping it, then pipe to
grep and use the -f flag to take the contig names from the outliers file
– this is a handy trick, and saves us a lot of time from having to
manually input each contig one at a time and search!</td>
</tr>
<tr class="odd">
<td>Pipe to a new grep to get just the contigs containing genes</td>
</tr>
<tr class="even">
<td>Pipe the list of gene IDs to take just the unique ones, cut out the
9th column (containing the gene IDs), and get rid of the annoying “ID=”
portion of each gene ID</td>
</tr>
<tr class="odd">
<td>Just a reminder, the following is bach code, not R ;)</td>
</tr>
<tr class="even">
<td>ANNOT=“/netfiles/ecogen/PopulationGenomics/ref_genome/annotation/Pabies1.0-all-cds.gff3.gz”</td>
</tr>
<tr class="odd">
<td>OUTLIERS=~/myresults/ANGSD/allRS_poly_PC1_outlier_contigs.txt</td>
</tr>
<tr class="even">
<td>zcat ${ANNOT} | grep -f ${OUTLIERS} | grep “gene” | uniq | cut -f9 |
sed “s/ID=//g” Voila! You should now have a list of gene IDs printed to
your screen. You can either copy these by highlighting with your mouse
and clicking or you can save to an external file for later (using the
&gt;outfile.txt command)</td>
</tr>
<tr class="odd">
<td>Take your gene IDs and go to the plantgenie website. Here, you can
create a gene list using your copied IDs, determine which genes they
correspond to (not all wil be annotated!), and test for functional
enrichment:</td>
</tr>
<tr class="even">
<td>In the upper left corner, see a circle with 3 red bars –&gt; click
it, and create a new gene list</td>
</tr>
<tr class="odd">
<td>Paste your gene IDs into the list, and plantgenie will search the P.
abies annotation and return all the known genes</td>
</tr>
<tr class="even">
<td>Click the “+” button to save these to the current gene list, then
close out</td>
</tr>
<tr class="odd">
<td>Go to the “Analysis Tools” tab on the top bar of the page, and
choose “Enrichment”</td>
</tr>
<tr class="even">
<td>Plantgenie will now test for whether the annotation Gene Ontology
(GO) and Protein Family (Pfam) functional descriptions assigned to each
gene ID are over-represented compared to all other genes containing
those GO and Pfam categories in the P. abies genome.</td>
</tr>
<tr class="odd">
<td>Relate back to the biology!</td>
</tr>
<tr class="even">
<td>What sort of functional genes did you find are represented in your
outliers? Are any enriched for certain GO or Pfam functions? (focus on
p-values &lt;&lt; 0.01, or q-values &lt; 0.05) Remember the population
structure axis (genetic PC) these loci are outliers along…are any of the
enriched functions interesting in light of this structure? Any
surprising?</td>
</tr>
</tbody>
</table>
<div id="id-section7">

</div>
<div id="entry-8-2020-10-04." class="section level2">
<h2>Entry 8: 2020-10-04.</h2>
<p>Learning Objectives for 10/02/23 Part 2 Extract climate data for red
spruce localities and summarize with PCA Start genotype-environment
association (GEA) analysis running We’ve now learned about outlier loci
that show extreme population structure along genetic PC axes. These are
loci that tend to have high Fst (population structure) suggestive of
selection and local adaptation. But how does this relate to different
aspects of the environment? How do we know what environmental gradients
might be responsible for driving this population differentiation?
Where’s the “landscape” in all this genomics stuff?</p>
<p>That’s where GEA comes in – short for “Genotype-Environment
Association”. The whole idea behind GEA is to look for loci that show
strong associations between allele frequencies and environmental
gradients, like climate. There’s a substantial literature on GEA methods
in landscape genomics, but a good background paper is Rellstab et
al. (2015).</p>
<p>For any GEA analysis, you essentially need two things:</p>
<p>Genotype data from individuals sampled across one or more
environmental gradients. Check! Environmental data (like cimate) you
want to test as drivers of the selection on allele frequencies. Let’s go
get some! 1. Getting Bioclim climate data for our red spruce samples On
our laptops, we’ll use R to query the Worldclim climate database and
pull out the bioclim variables for each of our samples based on their
lat/longs.</p>
<p>Remember, the following code is in R, not bash ;)</p>
</div>
</div>
<div id="you-made-need-to-use-install.packages-if-you-dont-have-some-of-the-below-libraries-already" class="section level1">
<h1>You made need to use “install.packages” if you don’t have some of
the below libraries already</h1>
<p>library(raster) library(FactoMineR) library(factoextra)
library(corrplot)</p>
<p>setwd(““)</p>
<p>bio &lt;- getData(“worldclim”,var=“bio”,res=10)</p>
<p>coords &lt;- read.csv(“<a href="https://www.uvm.edu/~kellrlab/forClass/colebrookSampleMetaData.csv" class="uri">https://www.uvm.edu/~kellrlab/forClass/colebrookSampleMetaData.csv</a>”,
header=T)</p>
<p>The chunk below refers to your bamlist file that you transferred
during last week’s PCA/admixture analysis. It should be the same one you
want to use here – if your sample list for analysis changes in the
future, you’ll need a different bamlist!</p>
<p>names &lt;- read.table(“pcangsd/allRS_bam.list”) names &lt;-
unlist(strsplit(basename(as.character(names[,1])), split =
“.sorted.rmdup.bam”)) split = strsplit(names, “_“) pops &lt;-
data.frame(names[1:95], do.call(rbind, split[1:95])) names(pops) =
c(”Ind”, “Pop”, “Row”, “Col”)</p>
<p>angsd_coords &lt;- merge(pops, coords, by.x=“Ind”, by.y=“Tree”)</p>
<p>points &lt;-
SpatialPoints(angsd_coords[c(“Longitude”,“Latitude”)])</p>
<p>clim &lt;- extract(bio,points)</p>
<p>angsd_coords_clim &lt;- cbind.data.frame(angsd_coords,clim)
str(angsd_coords_clim) And just like that, we’ve got Bioclim data for
our samples! Now, we don’t want to test all 19 variables in our GEA, but
rather just one or a few that seem to capture the climate gradients in
our samples. For this, we can use PCA on the climate data (so, an
environmental, not a genetic PCA in this case). Using the cliamte PCA
can then let us see which variables most strongly define the climate
space of red spruce, and we can then use these for our GEA test.</p>
<p>In R: clim_PCA = PCA(angsd_coords_clim[,15:33], graph=F)</p>
</div>
<div id="screeplot-of-pca-eigenvalues" class="section level1">
<h1>screeplot of PCA eigenvalues</h1>
<p>fviz_eig(clim_PCA)</p>
</div>
<div id="plot-the-variable-loadings-on-the-1st-2-pcs" class="section level1">
<h1>plot the variable loadings on the 1st 2 PCs</h1>
<p>fviz_pca_var(clim_PCA)</p>
</div>
<div id="which-variables-show-the-strongest-loading-on-each-axis" class="section level1">
<h1>Which variables show the strongest loading on each axis?</h1>
<p>var &lt;- get_pca_var(clim_PCA) corrplot(var$cos2, is.corr=FALSE)</p>
</div>
<div id="we-can-ask-the-same-question-using-correlation-tests-of-each-variable-with-each-pc-axis" class="section level1">
<h1>We can ask the same question using correlation tests of each
variable with each PC axis:</h1>
<p>dimdesc(clim_PCA)</p>
</div>
<div id="lastly-how-do-our-samples-fall-out-in-climate-pca-space" class="section level1">
<h1>Lastly, how do our samples fall out in climate PCA space?</h1>
<p>fviz_pca_ind(clim_PCA, geom.ind=“point”, col.ind =
angsd_coords_clim$Latitude, gradient.cols = c(“#00AFBB”, “#E7B800”,
“#FC4E07”), legend.title=“Latitude”) What do you think? Let’s pick 1
Bioclim variabe for each PC1 and 2 axis and export the values:</p>
<p>write.table(scale(angsd_coords_clim[,c(“bioXX”,“bioYY”)]),
“allRS_bioXX_YY.txt”, sep=“, quote=F, row.names = F, col.names=F) After
you’ve written this file out, use FileZilla to transfer it to the server
and into your ~/myresults/ANGSD folder.</p>
<p>You’ll then need to use the cut command to write out separate
environmental files for each variable (ANGSD only tests one variable at
a time).</p>
<p>Can you think of a cut statement to take your allRS_bioXX_YY.txt file
as input and write out a new output called allRS_bioXX.txt that has just
the first variable in it? Then try it again for the 2nd variable.</p>
<p>cut options someinfile &gt;someoutfile</p>
<ol start="2" style="list-style-type: decimal">
<li>Getting our GEA run on the server started! The GEA analysis will
likely take some time – maybe over a day to run. So let’s get it started
now so we can have the results to look at on Wednesday.</li>
</ol>
<p>We’ll use ANGSD again to test for associations between allele
frequencies and the bioclim variables while using genotype likelihoods
to account for uncertainty. The ANGSD routine we’ll use is called doAsso
which stands for “do Association”. It has it’s own manual page on
ANGSD’s website</p>
<p>AGNSD doAsso requires genotype probabilities (yes, these are slightly
different mathematical than genotype likelihoods), which means we’ll
need to call the initial set of ANGSD commands to import the bam’s and
filter them before sending the probabilities to the doAsso function.</p>
<p>bash header</p>
<p>notes</p>
<p>REF=“/netfiles/ecogen/PopulationGenomics/ref_genome/Pabies1.0-genome_reduced.fa”</p>
<p>SUFFIX=“”</p>
<p>INPUT=“”</p>
<p>OUTPUT=</p>
</div>
<div id="make-a-list-of-all-the-sorted-and-pcr-dup-removed-bam-files" class="section level1">
<h1>Make a list of all the sorted and PCR dup removed bam files:</h1>
<p>ls
/netfiles/ecogen/PopulationGenomics/fastq/red_spruce/cleanreads/bam/2*sorted.rmdup.bam
&gt; ${OUTPUT}/allRS_bam.list</p>
</div>
<div id="run-angsd-to-estimate-the-geno-probs-and-then-perform-the-gea-using-doasso" class="section level1">
<h1>Run ANGSD to estimate the Geno Probs and then perform the GEA using
doAsso:</h1>
<p>ANGSD -b ${OUTPUT}/allRS_bam.list<br />
-ref ${REF} -anc ${REF}<br />
-out <span class="math inline">\({OUTPUT}/\)</span>{SUFFIX}<br />
-nThreads 1<br />
-remove_bads 1<br />
-C 50<br />
-baq 1<br />
-minMapQ 20<br />
-minQ 20<br />
-GL 1<br />
-doSaf 1<br />
-doCounts 1<br />
-minInd 47<br />
-setMinDepthInd 1<br />
-setMaxDepthInd 40<br />
-skipTriallelic 1<br />
-doMajorMinor 1<br />
-doMaf 1<br />
-SNP_pval 1e-6<br />
-minMaf 0.05<br />
-doPost 1<br />
-doAsso 5<br />
-yQuant ${OUTPUT}/allRS_bioXX.txt That’s a lot of options for ANGSD!
Here’s the breakdown (after the obvious in/out and reference genome at
the beginning):</p>
<p>Code option Meaning -nthreads how many CPU’s you want to use
-remove_bads discards reads that don’t map as a pair to a unique
location in the ref genome -C 50 downgrades the quality of reads with
excessive mismatches to the ref genome -baq 1 performs re-calibration of
base Q-scores around regions of structural variation -minMapQ 20
discards reads with mapping Q-score &lt;20 -minQ 20 sets bases with
Q-scores &lt; 20 to missing -GL 1 estimate genotype likelihoods using
the Samtools algorithm -doSAF 1 estimate site-allele frequencies
-doCounts 1 count mapped bases at each site – needed for filters below
-minInd 47 skip loci that have &lt; minInd number of individuals with
data; set to ~50% of your full sample size -setMinDepthInd 1 set
individuals with &lt;1 read per site to missing for that site
-setMaxDepthInd 40 set individuals with &gt; 40 reads per site to
missing for that site -skipTriallelic 1 if there are &gt;2 alleles
present at a site – discard! -doMajorMinor 1 calculate which allele is
the most common (major) and which is rare (minor) -doMaf 1 compute the
minor allele frequencies -SNP_pval 1e-6 discard sites unlikely to be
polymorphic, based on a p-value of &lt;1e-6 -minMaf 0.05 discard sites
with minor allele frequencies &lt; 5% Most of these are identical to
what I used for calling the genotype likelihoods (beagle files) that we
used as input into pcANGSD. But I wanted to give you all the options and
their meaning (see also the ANGSD manual online so you could have them
in case you need to change anything.</p>
<p>Those last 3 lines are the important part for the GEA:</p>
<p>Code option Meaning doPost 1 compute genotype probabilities doAsso 5
perform the GEA association test -yQuant path and filename containing
the input environmental variable to test You’ll also notice that we do
this just on loci that are present in at least 5% frequeny across the
entire set of samples (-minMaf 0.05)</p>
<p>Save your script as ANGSD_GEA.sh</p>
<p>Then you know what to do….tmux, bash, dettach from screen, top to see
if you’re running!</p>
</div>
</section>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
